{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e45e54b2-abf9-464e-b786-53d5fcf911ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "import primer3\n",
    "import openpyxl\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import itertools\n",
    "import re\n",
    "from scipy.spatial import distance\n",
    "import json\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "import random\n",
    "import pprint\n",
    "import string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f2a7c101-e5bd-4633-8eb3-cecfcf342541",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import primers\n",
    "orthogonal_F = pd.read_excel('RR_orthogonalFv1_plate.xlsx')\n",
    "orthogonal_R = pd.read_excel('RR_orthogonalRv1_plate.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "71bfde6a-aae7-4377-b0ce-86604d3c4e56",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Isolate 20-nt variable primer binding region and only use that to align to sequences\n",
    "orthogonal_F['PrimerEnd'] = orthogonal_F.Sequence.apply(lambda x: x[-20:])\n",
    "orthogonal_R['PrimerEnd'] = orthogonal_R.Sequence.apply(lambda x: x[-20:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c77b6ae3-f45a-4253-8420-ec06a1887c09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exclude poor primers determined empirically\n",
    "bad_primers_F = [\"A9\", \"C3\", \"D3\", \"E3\", \"G4\", \"G6\", \"H4\"]\n",
    "bad_primers_R = [\"A9\", \"C6\", \"D3\", \"D7\", \"E10\", \"E3\", \"E6\", \"H4\"]\n",
    "red_primers = [\"A8\",\"A9\",\"C3\",\"C4\",\"C6\",\"C8\",\"C9\",\"H2\",\"D3\",\"D4\",\"D8\",\"D9\",\"E3\",\"E6\",\"F3\",\"F4\",\"G3\",\"G4\",\"G6\",\"H3\",\"H4\",\"H5\",\"H7\"]\n",
    "\n",
    "orthogonal_F['Exclude'] = orthogonal_F['Well Position'].map(lambda x: True if x in bad_primers_F else False)\n",
    "orthogonal_R['Exclude'] = orthogonal_R['Well Position'].map(lambda x: True if x in bad_primers_R else False)\n",
    "\n",
    "orthogonal_F['Worse'] = orthogonal_F['Well Position'].map(lambda x: True if x in red_primers else False)\n",
    "orthogonal_R['Worse'] = orthogonal_R['Well Position'].map(lambda x: True if x in red_primers else False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6abbe5db-8fb0-4c3d-a3cb-94ec88d56e88",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check for BsaI sites and remove any primers with them\n",
    "bsaI_seq = Seq('GGTCTC')\n",
    "orthogonal_F['BsaI_Site_Present'] = orthogonal_F.Sequence.apply(lambda x: ( (str(bsaI_seq) in x) | (str(bsaI_seq.reverse_complement()) in x) ))\n",
    "orthogonal_R['BsaI_Site_Present'] = orthogonal_R.Sequence.apply(lambda x: ( (str(bsaI_seq) in x) | (str(bsaI_seq.reverse_complement()) in x) ))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e07a3329-7ed0-49d1-8cf0-35d5995cdd34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adapted from Willow Coyote-Maestas' DIMPLE paper, cite doi: 10.1186/s13059-023-02880-6\n",
    "def check_nonspecific(primer, fragment, Tm_verb = 20, Tm_rem = 28, primer3_shift = 4, verbose=True):\n",
    "    non = []\n",
    "    # Forward\n",
    "    for i in range(len(fragment) - len(primer)):  # Scan each position\n",
    "        match = [\n",
    "            primer[j].lower() == fragment[i + j].lower() for j in range(len(primer))\n",
    "        ]\n",
    "        first = 10\n",
    "        for k in range(len(match) - 3):\n",
    "            if (match[k] and match[k + 1] and match[k + 3]) or (\n",
    "                match[k] and match[k + 1] and match[k + 2]\n",
    "            ):\n",
    "                first = k\n",
    "                break\n",
    "        if (\n",
    "            sum(match[first:]) > len(primer[first:]) * 0.8\n",
    "            and sum(match[first:]) > 6\n",
    "            and match[-1]\n",
    "        ):  # string compare - sum of matched nt is greater than 80%\n",
    "            try:\n",
    "                melt = mt.Tm_NN(\n",
    "                    primer[first:],\n",
    "                    c_seq=fragment[i + first : i + len(primer)].complement(),\n",
    "                    nn_table=mt.DNA_NN2,\n",
    "                    de_table=mt.DNA_DE1,\n",
    "                    imm_table=mt.DNA_IMM1,\n",
    "                )\n",
    "                if verbose == True:\n",
    "                    if melt > Tm_verb:\n",
    "                        print(\"Found non-specific match at \" + str(i + 1) + \"bp:\")\n",
    "                        print(\" match:\" + fragment[i : i + len(primer)])\n",
    "                        print(\"primer:\" + primer + \" Tm:\" + str(round(melt, 1)))\n",
    "                if melt > Tm_rem:\n",
    "                    non.append(True)\n",
    "            except ValueError as valerr:\n",
    "                # use primer3 instead, as mt.DNA_NN2 table does not have enough information to compute Tm\n",
    "                result = primer3.calcHeterodimer(str(primer[first:]), \n",
    "                                                 str(fragment[i + first : i + len(primer)].complement())\n",
    "                                                )\n",
    "                melt = result.tm\n",
    "                if verbose == True:\n",
    "                    if melt > Tm_verb-primer3_shift:\n",
    "                        print(\"Found non-specific match using Primer3 at \" + str(i + 1) + \"bp:\")\n",
    "                        print(\" match:\" + fragment[i : i + len(primer)])\n",
    "                        print(\"primer:\" + primer + \" Tm:\" + str(round(melt, 1)))\n",
    "                if melt > Tm_rem-primer3_shift:\n",
    "                    non.append(True)\n",
    "                    \n",
    "    # Reverse\n",
    "    fragment = fragment.reverse_complement()\n",
    "    for i in range(len(fragment) - len(primer)):\n",
    "        match = [\n",
    "            primer[j].lower() == fragment[i + j].lower() for j in range(len(primer))\n",
    "        ]\n",
    "        first = 10\n",
    "        for k in range(0, len(match) - 3, 1):\n",
    "            if match[k] and match[k + 1] and match[k + 3]:\n",
    "                first = k\n",
    "                break\n",
    "        if (\n",
    "            sum(match[first:]) > len(primer[first:]) * 0.8\n",
    "            and sum(match[first:]) > 6\n",
    "            and match[-1]\n",
    "        ):  # string compare - sum of matched nt is greater than 80%\n",
    "            try:\n",
    "                melt = mt.Tm_NN(\n",
    "                    primer[first:],\n",
    "                    c_seq=fragment[i + first : i + len(primer)].complement(),\n",
    "                    nn_table=mt.DNA_NN2,\n",
    "                    de_table=mt.DNA_DE1,\n",
    "                    imm_table=mt.DNA_IMM1,\n",
    "                )\n",
    "                if verbose == True:\n",
    "                    if melt > Tm_verb:\n",
    "                        print(\"Found non-specific match at \" + str(i + 1) + \"bp:\")\n",
    "                        print(\" match:\" + fragment[i : i + len(primer)])\n",
    "                        print(\"primer:\" + primer + \" Tm:\" + str(melt))\n",
    "                if melt > Tm_rem:\n",
    "                    non.append(True)\n",
    "            except ValueError as valerr:\n",
    "                # use primer3 instead, as mt.DNA_NN2 table does not have enough information to compute Tm\n",
    "                result = primer3.calcHeterodimer(str(primer[first:]), \n",
    "                                                 str(fragment[i + first : i + len(primer)].complement())\n",
    "                                                )\n",
    "                melt = result.tm\n",
    "                if verbose == True:\n",
    "                    if melt > Tm_verb-primer3_shift:\n",
    "                        print(\"Found non-specific match using Primer3 at \" + str(i + 1) + \"bp:\")\n",
    "                        print(\" match:\" + fragment[i : i + len(primer)])\n",
    "                        print(\"primer:\" + primer + \" Tm:\" + str(round(melt, 1)))\n",
    "                if melt > Tm_rem-primer3_shift:\n",
    "                    non.append(True)\n",
    "    return sum(non)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "64071724-4c10-483e-8e6c-b2f4cae6e30b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing Homo sapiens von Willebrand factor (VWF), mRNA\n"
     ]
    }
   ],
   "source": [
    "# Use refseq MANE annotations to find gene cDNA sequences\n",
    "refseq_MANE_summary_filepath = 'MANE.GRCh38.v1.3.summary.txt'\n",
    "refseq_MANE_filepath = \"MANE.GRCh38.v1.3.refseq_rna.gbff\"\n",
    "refseq_MANE_summary = pd.read_csv(refseq_MANE_summary_filepath, sep='\\t')\n",
    "refseq_MANE_stream = SeqIO.parse(refseq_MANE_filepath, \"genbank\")\n",
    "\n",
    "# genes for assay\n",
    "genes_of_interest = \\\n",
    "['VWF']\n",
    "\n",
    "# find correct transcripts\n",
    "genes_to_MANE_transcripts = refseq_MANE_summary.\\\n",
    "    query('MANE_status == \"MANE Select\"').\\\n",
    "    query('symbol in @genes_of_interest')\n",
    "\n",
    "# find refseq\n",
    "gene_transcript_dict = {}\n",
    "for rec in refseq_MANE_stream:\n",
    "    if rec.id in genes_to_MANE_transcripts['RefSeq_nuc'].values:\n",
    "        print('Processing ' + rec.description)\n",
    "        for feature in rec.features:\n",
    "            if feature.type == \"CDS\":\n",
    "                gene_transcript_dict[rec.id] = [rec.description,\n",
    "                                                str(feature.location.extract(rec).seq)]\n",
    "genes_to_MANE_transcripts = genes_to_MANE_transcripts.\\\n",
    "    merge(pd.DataFrame.from_dict(gene_transcript_dict, orient='index', columns=['Description', 'CDS_Seq']),\n",
    "             left_on='RefSeq_nuc', right_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "36035e65-6086-4af8-b994-9e9b64baca55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VWF\n",
      "CDS: ATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTGACTTCGTCAACACCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCATTTGTTTGTCAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTTCAAGTCCTGCTGTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTGGTGTGAACGGGCATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAGACTTTGTGTGAGTGTGCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTGCGCCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCGGCACCTCCCTCTCTCGAGACTGCAACACCTGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGCCAGTACCTGCTGGCCCGGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTGCCATGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTATGCCGGGAAGACCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTGCAGAAGCAGCACAGCGATCCCTGCGCCCTCAACCCGCGCATGACCAGGTTCTCCGAGGAGGCGTGCGCGGTCCTGACGTCCCCCACATTCGAGGCCTGCCATCGTGCCGTCAGCCCGCTGCCCTACCTGCGGAACTGCCGCTACGACGTGTGCTCCTGCTCGGACGGCCGCGAGTGCCTGTGCGGCGCCCTGGCCAGCTATGCCGCGGCCTGCGCGGGGAGAGGCGTGCGCGTCGCGTGGCGCGAGCCAGGCCGCTGTGAGCTGAACTGCCCGAAAGGCCAGGTGTACCTGCAGTGCGGGACCCCCTGCAACCTGACCTGCCGCTCTCTCTCTTACCCGGATGAGGAATGCAATGAGGCCTGCCTGGAGGGCTGCTTCTGCCCCCCAGGGCTCTACATGGATGAGAGGGGGGACTGCGTGCCCAAGGCCCAGTGCCCCTGTTACTATGACGGTGAGATCTTCCAGCCAGAAGACATCTTCTCAGACCATCACACCATGTGCTACTGTGAGGATGGCTTCATGCACTGTACCATGAGTGGAGTCCCCGGAAGCTTGCTGCCTGACGCTGTCCTCAGCAGTCCCCTGTCTCATCGCAGCAAAAGGAGCCTATCCTGTCGGCCCCCCATGGTCAAGCTGGTGTGTCCCGCTGACAACCTGCGGGCTGAAGGGCTCGAGTGTACCAAAACGTGCCAGAACTATGACCTGGAGTGCATGAGCATGGGCTGTGTCTCTGGCTGCCTCTGCCCCCCGGGCATGGTCCGGCATGAGAACAGATGTGTGGCCCTGGAAAGGTGTCCCTGCTTCCATCAGGGCAAGGAGTATGCCCCTGGAGAAACAGTGAAGATTGGCTGCAACACTTGTGTCTGTCAGGACCGGAAGTGGAACTGCACAGACCATGTGTGTGATGCCACGTGCTCCACGATCGGCATGGCCCACTACCTCACCTTCGACGGGCTCAAATACCTGTTCCCCGGGGAGTGCCAGTACGTTCTGGTGCAGGATTACTGCGGCAGTAACCCTGGGACCTTTCGGATCCTAGTGGGGAATAAGGGATGCAGCCACCCCTCAGTGAAATGCAAGAAACGGGTCACCATCCTGGTGGAGGGAGGAGAGATTGAGCTGTTTGACGGGGAGGTGAATGTGAAGAGGCCCATGAAGGATGAGACTCACTTTGAGGTGGTGGAGTCTGGCCGGTACATCATTCTGCTGCTGGGCAAAGCCCTCTCCGTGGTCTGGGACCGCCACCTGAGCATCTCCGTGGTCCTGAAGCAGACATACCAGGAGAAAGTGTGTGGCCTGTGTGGGAATTTTGATGGCATCCAGAACAATGACCTCACCAGCAGCAACCTCCAAGTGGAGGAAGACCCTGTGGACTTTGGGAACTCCTGGAAAGTGAGCTCGCAGTGTGCTGACACCAGAAAAGTGCCTCTGGACTCATCCCCTGCCACCTGCCATAACAACATCATGAAGCAGACGATGGTGGATTCCTCCTGTAGAATCCTTACCAGTGACGTCTTCCAGGACTGCAACAAGCTGGTGGACCCCGAGCCATATCTGGATGTCTGCATTTACGACACCTGCTCCTGTGAGTCCATTGGGGACTGCGCCTGCTTCTGCGACACCATTGCTGCCTATGCCCACGTGTGTGCCCAGCATGGCAAGGTGGTGACCTGGAGGACGGCCACATTGTGCCCCCAGAGCTGCGAGGAGAGGAATCTCCGGGAGAACGGGTATGAGTGTGAGTGGCGCTATAACAGCTGTGCACCTGCCTGTCAAGTCACGTGTCAGCACCCTGAGCCACTGGCCTGCCCTGTGCAGTGTGTGGAGGGCTGCCATGCCCACTGCCCTCCAGGGAAAATCCTGGATGAGCTTTTGCAGACCTGCGTTGACCCTGAAGACTGTCCAGTGTGTGAGGTGGCTGGCCGGCGTTTTGCCTCAGGAAAGAAAGTCACCTTGAATCCCAGTGACCCTGAGCACTGCCAGATTTGCCACTGTGATGTTGTCAACCTCACCTGTGAAGCCTGCCAGGAGCCGGGAGGCCTGGTGGTGCCTCCCACAGATGCCCCGGTGAGCCCCACCACTCTGTATGTGGAGGACATCTCGGAACCGCCGTTGCACGATTTCTACTGCAGCAGGCTACTGGACCTGGTCTTCCTGCTGGATGGCTCCTCCAGGCTGTCCGAGGCTGAGTTTGAAGTGCTGAAGGCCTTTGTGGTGGACATGATGGAGCGGCTGCGCATCTCCCAGAAGTGGGTCCGCGTGGCCGTGGTGGAGTACCACGACGGCTCCCACGCCTACATCGGGCTCAAGGACCGGAAGCGACCGTCAGAGCTGCGGCGCATTGCCAGCCAGGTGAAGTATGCGGGCAGCCAGGTGGCCTCCACCAGCGAGGTCTTGAAATACACACTGTTCCAAATCTTCAGCAAGATCGACCGCCCTGAAGCCTCCCGCATCACCCTGCTCCTGATGGCCAGCCAGGAGCCCCAACGGATGTCCCGGAACTTTGTCCGCTACGTCCAGGGCCTGAAGAAGAAGAAGGTCATTGTGATCCCGGTGGGCATTGGGCCCCATGCCAACCTCAAGCAGATCCGCCTCATCGAGAAGCAGGCCCCTGAGAACAAGGCCTTCGTGCTGAGCAGTGTGGATGAGCTGGAGCAGCAAAGGGACGAGATCGTTAGCTACCTCTGTGACCTTGCCCCTGAAGCCCCTCCTCCTACTCTGCCCCCCGACATGGCACAAGTCACTGTGGGCCCGGGGCTCTTGGGGGTTTCGACCCTGGGGCCCAAGAGGAACTCCATGGTTCTGGATGTGGCGTTCGTCCTGGAAGGATCGGACAAAATTGGTGAAGCCGACTTCAACAGGAGCAAGGAGTTCATGGAGGAGGTGATTCAGCGGATGGATGTGGGCCAGGACAGCATCCACGTCACGGTGCTGCAGTACTCCTACATGGTGACTGTGGAGTACCCCTTCAGCGAGGCACAGTCCAAAGGGGACATCCTGCAGCGGGTGCGAGAGATCCGCTACCAGGGCGGCAACAGGACCAACACTGGGCTGGCCCTGCGGTACCTCTCTGACCACAGCTTCTTGGTCAGCCAGGGTGACCGGGAGCAGGCGCCCAACCTGGTCTACATGGTCACCGGAAATCCTGCCTCTGATGAGATCAAGAGGCTGCCTGGAGACATCCAGGTGGTGCCCATTGGAGTGGGCCCTAATGCCAACGTGCAGGAGCTGGAGAGGATTGGCTGGCCCAATGCCCCTATCCTCATCCAGGACTTTGAGACGCTCCCCCGAGAGGCTCCTGACCTGGTGCTGCAGAGGTGCTGCTCCGGAGAGGGGCTGCAGATCCCCACCCTCTCCCCTGCACCTGACTGCAGCCAGCCCCTGGACGTGATCCTTCTCCTGGATGGCTCCTCCAGTTTCCCAGCTTCTTATTTTGATGAAATGAAGAGTTTCGCCAAGGCTTTCATTTCAAAAGCCAATATAGGGCCTCGTCTCACTCAGGTGTCAGTGCTGCAGTATGGAAGCATCACCACCATTGACGTGCCATGGAACGTGGTCCCGGAGAAAGCCCATTTGCTGAGCCTTGTGGACGTCATGCAGCGGGAGGGAGGCCCCAGCCAAATCGGGGATGCCTTGGGCTTTGCTGTGCGATACTTGACTTCAGAAATGCATGGTGCCAGGCCGGGAGCCTCAAAGGCGGTGGTCATCCTGGTCACGGACGTCTCTGTGGATTCAGTGGATGCAGCAGCTGATGCCGCCAGGTCCAACAGAGTGACAGTGTTCCCTATTGGAATTGGAGATCGCTACGATGCAGCCCAGCTACGGATCTTGGCAGGCCCAGCAGGCGACTCCAACGTGGTGAAGCTCCAGCGAATCGAAGACCTCCCTACCATGGTCACCTTGGGCAATTCCTTCCTCCACAAACTGTGCTCTGGATTTGTTAGGATTTGCATGGATGAGGATGGGAATGAGAAGAGGCCCGGGGACGTCTGGACCTTGCCAGACCAGTGCCACACCGTGACTTGCCAGCCAGATGGCCAGACCTTGCTGAAGAGTCATCGGGTCAACTGTGACCGGGGGCTGAGGCCTTCGTGCCCTAACAGCCAGTCCCCTGTTAAAGTGGAAGAGACCTGTGGCTGCCGCTGGACCTGCCCCTGCGTGTGCACAGGCAGCTCCACTCGGCACATCGTGACCTTTGATGGGCAGAATTTCAAGCTGACTGGCAGCTGTTCTTATGTCCTATTTCAAAACAAGGAGCAGGACCTGGAGGTGATTCTCCATAATGGTGCCTGCAGCCCTGGAGCAAGGCAGGGCTGCATGAAATCCATCGAGGTGAAGCACAGTGCCCTCTCCGTCGAGCTGCACAGTGACATGGAGGTGACGGTGAATGGGAGACTGGTCTCTGTTCCTTACGTGGGTGGGAACATGGAAGTCAACGTTTATGGTGCCATCATGCATGAGGTCAGATTCAATCACCTTGGTCACATCTTCACATTCACTCCACAAAACAATGAGTTCCAACTGCAGCTCAGCCCCAAGACTTTTGCTTCAAAGACGTATGGTCTGTGTGGGATCTGTGATGAGAACGGAGCCAATGACTTCATGCTGAGGGATGGCACAGTCACCACAGACTGGAAAACACTTGTTCAGGAATGGACTGTGCAGCGGCCAGGGCAGACGTGCCAGCCCATCCTGGAGGAGCAGTGTCTTGTCCCCGACAGCTCCCACTGCCAGGTCCTCCTCTTACCACTGTTTGCTGAATGCCACAAGGTCCTGGCTCCAGCCACATTCTATGCCATCTGCCAGCAGGACAGTTGCCACCAGGAGCAAGTGTGTGAGGTGATCGCCTCTTATGCCCACCTCTGTCGGACCAACGGGGTCTGCGTTGACTGGAGGACACCTGATTTCTGTGCTATGTCATGCCCACCATCTCTGGTCTACAACCACTGTGAGCATGGCTGTCCCCGGCACTGTGATGGCAACGTGAGCTCCTGTGGGGACCATCCCTCCGAAGGCTGTTTCTGCCCTCCAGATAAAGTCATGTTGGAAGGCAGCTGTGTCCCTGAAGAGGCCTGCACTCAGTGCATTGGTGAGGATGGAGTCCAGCACCAGTTCCTGGAAGCCTGGGTCCCGGACCACCAGCCCTGTCAGATCTGCACATGCCTCAGCGGGCGGAAGGTCAACTGCACAACGCAGCCCTGCCCCACGGCCAAAGCTCCCACGTGTGGCCTGTGTGAAGTAGCCCGCCTCCGCCAGAATGCAGACCAGTGCTGCCCCGAGTATGAGTGTGTGTGTGACCCAGTGAGCTGTGACCTGCCCCCAGTGCCTCACTGTGAACGTGGCCTCCAGCCCACACTGACCAACCCTGGCGAGTGCAGACCCAACTTCACCTGCGCCTGCAGGAAGGAGGAGTGCAAAAGAGTGTCCCCACCCTCCTGCCCCCCGCACCGTTTGCCCACCCTTCGGAAGACCCAGTGCTGTGATGAGTATGAGTGTGCCTGCAACTGTGTCAACTCCACAGTGAGCTGTCCCCTTGGGTACTTGGCCTCAACTGCCACCAATGACTGTGGCTGTACCACAACCACCTGCCTTCCCGACAAGGTGTGTGTCCACCGAAGCACCATCTACCCTGTGGGCCAGTTCTGGGAGGAGGGCTGCGATGTGTGCACCTGCACCGACATGGAGGATGCCGTGATGGGCCTCCGCGTGGCCCAGTGCTCCCAGAAGCCCTGTGAGGACAGCTGTCGGTCGGGCTTCACTTACGTTCTGCATGAAGGCGAGTGCTGTGGAAGGTGCCTGCCATCTGCCTGTGAGGTGGTGACTGGCTCACCGCGGGGGGACTCCCAGTCTTCCTGGAAGAGTGTCGGCTCCCAGTGGGCCTCCCCGGAGAACCCCTGCCTCATCAATGAGTGTGTCCGAGTGAAGGAGGAGGTCTTTATACAACAAAGGAACGTCTCCTGCCCCCAGCTGGAGGTCCCTGTCTGCCCCTCGGGCTTTCAGCTGAGCTGTAAGACCTCAGCGTGCTGCCCAAGCTGTCGCTGTGAGCGCATGGAGGCCTGCATGCTCAATGGCACTGTCATTGGGCCCGGGAAGACTGTGATGATCGATGTGTGCACGACCTGCCGCTGCATGGTGCAGGTGGGGGTCATCTCTGGATTCAAGCTGGAGTGCAGGAAGACCACCTGCAACCCCTGCCCCCTGGGTTACAAGGAAGAAAATAACACAGGTGAATGTTGTGGGAGATGTTTGCCTACGGCTTGCACCATTCAGCTAAGAGGAGGACAGATCATGACACTGAAGCGTGATGAGACGCTCCAGGATGGCTGTGATACTCACTTCTGCAAGGTCAATGAGAGAGGAGAGTACTTCTGGGAGAAGAGGGTCACAGGCTGCCCACCCTTTGATGAACACAAGTGTCTGGCTGAGGGAGGTAAAATTATGAAAATTCCAGGCACCTGCTGTGACACATGTGAGGAGCCTGAGTGCAACGACATCACTGCCAGGCTGCAGTATGTCAAGGTGGGAAGCTGTAAGTCTGAAGTAGAGGTGGATATCCACTACTGCCAGGGCAAATGTGCCAGCAAAGCCATGTACTCCATTGACATCAACGATGTGCAGGACCAGTGCTCCTGCTGCTCTCCGACACGGACGGAGCCCATGCAGGTGGCCCTGCACTGCACCAATGGCTCTGTTGTGTACCATGAGGTTCTCAATGCCATGGAGTGCAAATGCTCCCCCAGGAAGTGCAGCAAGTGA\n",
      "AA: MIPARFAGVLLALALILPGTLCAEGTRGRSSTARCSLFGSDFVNTFDGSMYSFAGYCSYLLAGGCQKRSFSIIGDFQNGKRVSLSVYLGEFFDIHLFVNGTVTQGDQRVSMPYASKGLYLETEAGYYKLSGEAYGFVARIDGSGNFQVLLSDRYFNKTCGLCGNFNIFAEDDFMTQEGTLTSDPYDFANSWALSSGEQWCERASPPSSSCNISSGEMQKGLWEQCQLLKSTSVFARCHPLVDPEPFVALCEKTLCECAGGLECACPALLEYARTCAQEGMVLYGWTDHSACSPVCPAGMEYRQCVSPCARTCQSLHINEMCQERCVDGCSCPEGQLLDEGLCVESTECPCVHSGKRYPPGTSLSRDCNTCICRNSQWICSNEECPGECLVTGQSHFKSFDNRYFTFSGICQYLLARDCQDHSFSIVIETVQCADDRDAVCTRSVTVRLPGLHNSLVKLKHGAGVAMDGQDVQLPLLKGDLRIQHTVTASVRLSYGEDLQMDWDGRGRLLVKLSPVYAGKTCGLCGNYNGNQGDDFLTPSGLAEPRVEDFGNAWKLHGDCQDLQKQHSDPCALNPRMTRFSEEACAVLTSPTFEACHRAVSPLPYLRNCRYDVCSCSDGRECLCGALASYAAACAGRGVRVAWREPGRCELNCPKGQVYLQCGTPCNLTCRSLSYPDEECNEACLEGCFCPPGLYMDERGDCVPKAQCPCYYDGEIFQPEDIFSDHHTMCYCEDGFMHCTMSGVPGSLLPDAVLSSPLSHRSKRSLSCRPPMVKLVCPADNLRAEGLECTKTCQNYDLECMSMGCVSGCLCPPGMVRHENRCVALERCPCFHQGKEYAPGETVKIGCNTCVCQDRKWNCTDHVCDATCSTIGMAHYLTFDGLKYLFPGECQYVLVQDYCGSNPGTFRILVGNKGCSHPSVKCKKRVTILVEGGEIELFDGEVNVKRPMKDETHFEVVESGRYIILLLGKALSVVWDRHLSISVVLKQTYQEKVCGLCGNFDGIQNNDLTSSNLQVEEDPVDFGNSWKVSSQCADTRKVPLDSSPATCHNNIMKQTMVDSSCRILTSDVFQDCNKLVDPEPYLDVCIYDTCSCESIGDCACFCDTIAAYAHVCAQHGKVVTWRTATLCPQSCEERNLRENGYECEWRYNSCAPACQVTCQHPEPLACPVQCVEGCHAHCPPGKILDELLQTCVDPEDCPVCEVAGRRFASGKKVTLNPSDPEHCQICHCDVVNLTCEACQEPGGLVVPPTDAPVSPTTLYVEDISEPPLHDFYCSRLLDLVFLLDGSSRLSEAEFEVLKAFVVDMMERLRISQKWVRVAVVEYHDGSHAYIGLKDRKRPSELRRIASQVKYAGSQVASTSEVLKYTLFQIFSKIDRPEASRITLLLMASQEPQRMSRNFVRYVQGLKKKKVIVIPVGIGPHANLKQIRLIEKQAPENKAFVLSSVDELEQQRDEIVSYLCDLAPEAPPPTLPPDMAQVTVGPGLLGVSTLGPKRNSMVLDVAFVLEGSDKIGEADFNRSKEFMEEVIQRMDVGQDSIHVTVLQYSYMVTVEYPFSEAQSKGDILQRVREIRYQGGNRTNTGLALRYLSDHSFLVSQGDREQAPNLVYMVTGNPASDEIKRLPGDIQVVPIGVGPNANVQELERIGWPNAPILIQDFETLPREAPDLVLQRCCSGEGLQIPTLSPAPDCSQPLDVILLLDGSSSFPASYFDEMKSFAKAFISKANIGPRLTQVSVLQYGSITTIDVPWNVVPEKAHLLSLVDVMQREGGPSQIGDALGFAVRYLTSEMHGARPGASKAVVILVTDVSVDSVDAAADAARSNRVTVFPIGIGDRYDAAQLRILAGPAGDSNVVKLQRIEDLPTMVTLGNSFLHKLCSGFVRICMDEDGNEKRPGDVWTLPDQCHTVTCQPDGQTLLKSHRVNCDRGLRPSCPNSQSPVKVEETCGCRWTCPCVCTGSSTRHIVTFDGQNFKLTGSCSYVLFQNKEQDLEVILHNGACSPGARQGCMKSIEVKHSALSVELHSDMEVTVNGRLVSVPYVGGNMEVNVYGAIMHEVRFNHLGHIFTFTPQNNEFQLQLSPKTFASKTYGLCGICDENGANDFMLRDGTVTTDWKTLVQEWTVQRPGQTCQPILEEQCLVPDSSHCQVLLLPLFAECHKVLAPATFYAICQQDSCHQEQVCEVIASYAHLCRTNGVCVDWRTPDFCAMSCPPSLVYNHCEHGCPRHCDGNVSSCGDHPSEGCFCPPDKVMLEGSCVPEEACTQCIGEDGVQHQFLEAWVPDHQPCQICTCLSGRKVNCTTQPCPTAKAPTCGLCEVARLRQNADQCCPEYECVCDPVSCDLPPVPHCERGLQPTLTNPGECRPNFTCACRKEECKRVSPPSCPPHRLPTLRKTQCCDEYECACNCVNSTVSCPLGYLASTATNDCGCTTTTCLPDKVCVHRSTIYPVGQFWEEGCDVCTCTDMEDAVMGLRVAQCSQKPCEDSCRSGFTYVLHEGECCGRCLPSACEVVTGSPRGDSQSSWKSVGSQWASPENPCLINECVRVKEEVFIQQRNVSCPQLEVPVCPSGFQLSCKTSACCPSCRCERMEACMLNGTVIGPGKTVMIDVCTTCRCMVQVGVISGFKLECRKTTCNPCPLGYKEENNTGECCGRCLPTACTIQLRGGQIMTLKRDETLQDGCDTHFCKVNERGEYFWEKRVTGCPPFDEHKCLAEGGKIMKIPGTCCDTCEEPECNDITARLQYVKVGSCKSEVEVDIHYCQGKCASKAMYSIDINDVQDQCSCCSPTRTEPMQVALHCTNGSVVYHEVLNAMECKCSPRKCSK*\n"
     ]
    }
   ],
   "source": [
    "for gene in genes_of_interest:\n",
    "    print(gene)\n",
    "    gene_cds_seq = Seq(genes_to_MANE_transcripts[genes_to_MANE_transcripts['symbol']==gene]['CDS_Seq'].values[0].upper())\n",
    "    print('CDS: ' + str(gene_cds_seq))\n",
    "    print('AA: ' + str(gene_cds_seq.translate()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0a8d287c-ef1f-4fe3-abec-77935a3c4094",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define genes\n",
    "# clotting proteins:\n",
    "\n",
    "# truncated F8 (block1_notag) (2_1)\n",
    "FRED_SB1 = Seq('atgcaaatagagctctctacctgcttctttctgtgccttttgcgattctgctttagtgccaccagaagatactacctgggtgcagtggaactgtcatgggactatatgcaaagtgatctcggtgagctgcctgtggacgcaagatttcctcctagagtgccaaaatcttttccattcaacacctcagtcgtgtacaaaaagactctgtttgtagaattcacggatcaccttttcaacatcgctaagccaaggccaccctggatgggtctgctaggtcctaccatccaggctgaggtttatgatacagtggtcattacacttaagaacatggcttcccatcctgtcagtcttcatgctgttggtgtatcctactggaaagcttctgagggagctgaatatgatgatcagaccagtcaaagggagaaagaagatgataaagtcttccctggtggaagccatacatatgtctggcaggtcctgaaagagaatggtccaatggcctctgacccactgtgccttacctactcatatctttctcatgtggacctggtaaaagacttgaattcaggcctcattggagccctactagtatgtagagaagggagtctggccaaggaaaagacacagaccttgcacaaatttatactactttttgctgtatttgatgaagggaaaagttggcactcagaaacaaagaactccttgatgcaggatagggatgctgcatctgctcgggcctggcctaaaatgcacacagtcaatggttatgtaaacagatctctgccaggtctgattggatgccacaggaaatcagtctattggcatgtgattggaatgggcaccactcctgaagtgcactcaatattcctcgaaggtcacacatttcttgtgaggaaccatcgccaggcgtccttggaaatctcgccaataactttccttactgctcaaacactcttgatggaccttggacagtttctactgttttgtcatatctcttcccaccaacatgatggcatggaagcttatgtcaaagtagacagctgtccagaggaaccccaactacgaatgaaaaataatgaagaagcggaagactatgatgatgatcttactgattctgaaatggatgtggtcaggtttgatgatgacaactctccttcctttatccaaattcgctcagttgccaagaagcatcctaaaacttgggtacattacattgctgctgaagaggaggactgggactatgctcccttagtcctcgcccccgatgacagaagttataaaagtcaatatttgaacaatggccctcagcggattggtaggaagtacaaaaaagtccgatttatggcatacacagatgaaacctttaagactcgtgaagctattcagcatgaatcaggaatcttgggacctttactttatggggaagttggagacacactgttgattatatttaagaatcaagcaagcagaccatataacatctaccctcacggaatcactgatgtccgtcctttgtattcaaggagattaccaaaaggtgtaaaacatttgaaggattttccaattctgccaggagaaatattcaaatataaatggacagtgactgtagaagatgggccaactaaatcagatcctcggtgcctgacccgctattactct'.upper())\n",
    "\n",
    "# truncated F8 (block2_notag) (1_2)\n",
    "FRED_SB2 = Seq('ttcaaatataaatggacagtgactgtagaagatgggccaactaaatcagatcctcggtgcctgacccgctattactctagtttcgttaatatggagagagatctagcttcaggactcattggccctctcctcatctgctacaaagaatctgtagatcaaagaggaaaccagataatgtcagacaagaggaatgtcatcctgttttctgtatttgatgagaaccgaagctggtacctcacagagaatatacaacgctttctccccaatccagctggagtgcagcttgaggatccagagttccaagcctccaacatcatgcacagcatcaatggctatgtttttgatagtttgcagttgtcagtttgtttgcatgaggtggcatactggtacattctaagcattggagcacagactgacttcctttctgtcttcttctctggatataccttcaaacacaaaatggtctatgaagacacactcaccctattcccattctcaggagaaactgtcttcatgtcgatggaaaacccaggtctatggattctggggtgccacaactcagactttcggaacagaggcatgaccgccttactgaaggtttctagttgtgacaagaacactggtgattattacgaggacagttatgaagatatttcagcatacttgctgagtaaaaacaatgccattgaaccaagaagcttctcccagaattcaagacaccctagcactaggcaaaagcaatttaatgccaccacaattccagaaaatgacatagagaagactgacccttggtttgcacacagaacacctatgcctaaaatacaaaatgtctcctctagtgatttgttgatgctcttgcgacagagtcctactccacatgggctatccttatctgatctccaagaagccaaatatgagactttttctgatgatccatcacctggagcaatagacagtaataacagcctgtctgaaatgacacacttcaggccacagctccatcacagtggggacatggtatttacccctgagtcaggcctccaattaagattaaatgagaaactggggacaactgcagcaacagagttgaagaaacttgatttcaaagtttctagtacatcaaataatctgatttcaacaattccatcagacaatttggcagcaggtactgataatacaagttccttaggacccccaagtatgccagttcattatgatagtcaattagataccactctatttggcaaaaagtcatctccccttactgagtctggtggacctctgagcttgagtgaagaaaataatgattcaaagttgttagaatcaggtttaatgaatagccaagaaagttcatggggaaaaaatgtatcgtcacgggaaataactcgtactactcttcagtcagatcaagaggaaattgactatgatgataccatatcagttgaaatgaagaaggaagattttgacatttatgatgaggatgaaaatcagagcccccgcagctttcaaaagaaaacacgacactattttattgctgcagtggagaggctctgggattatgggatgagtagctccccacatgttctaagaaacagggctcagagtggcagtgtccctcagttcaagaaagttgttttccaggaatttactgatggctcctttactcagcccttataccgtggagaactaaatgaacatttgggactcctggggccatatataaga'.upper())\n",
    "\n",
    "# truncated F8 (block3_ctag) (3_2)\n",
    "FRED_SB3 = Seq('ccatatataagagcagaagttgaagataatatcatggtaactttcagaaatcaggcctctcgtccctattccttctattctagccttatttcttatgaggaagatcagaggcaaggagcagaacctagaaaaaactttgtcaagcctaatgaaaccaaaacttacttttggaaagtgcaacatcatatggcacccactaaagatgagtttgactgcaaagcctgggcttatttctctgatgttgacctggaaaaagatgtgcactcaggcctgattggaccccttctggtctgccacactaacacactgaaccctgctcatgggagacaagtgacagtacaggaatttgctctgtttttcaccatctttgatgaaaccaaaagctggtacttcactgaaaatatggaaagaaactgcagggctccctgcaatatccagatggaagatcccacttttaaagagaattatcgcttccatgcaatcaatggctacataatggatacactacctggcttagtaatggctcaggatcaaaggattcgatggtatctgctcagcatgggcagcaatgaaaacatccattctattcatttcagtggacatgtgttcactgtacgaaaaaaagaggagtataaaatggcactgtacaatctctatccaggtgtttttgagacagtggaaatgttaccatccaaagctggaatttggcgggtggaatgccttattggcgagcatctacatgctgggatgagcacactttttctggtgtacagcaataagtgtcagactcccctgggaatggcttctggacacattagagattttcagattacagcttcaggacaatatggacagtgggccccaaagctggccagacttcattattccggatcaatcaatgcctggagcaccaaggagcccttttcttggatcaaggtggatctgttggcaccaatgattattcacggcatcaagacccagggtgcccgtcagaagttctccagcctctacatctctcagtttatcatcatgtatagtcttgatgggaagaagtggcagacttatcgaggaaattccactggaaccttaatggtcttctttggcaatgtggattcatctgggataaaacacaatatttttaaccctccaattattgctcgatacatccgtttgcacccaactcattatagcattcgcagcactcttcgcatggagttgatgggctgtgatttaaatagttgcagcatgccattgggaatggagagtaaagcaatatcagatgcacagattactgcttcatcctactttaccaatatgtttgccacctggtcaccttcaaaagctcgacttcacctccaagggaggagtaatgcctggaggcctcaggtgaataatccaaaagagtggctgcaagtggacttccagaagacaatgaaagtcacaggagtaactactcagggagtaaaatctctgcttaccagcatgtatgtgaaggagttcctcatctccagcagtcaagatggccatcagtggactctcttttttcagaatggcaaagtaaaggtttttcagggaaatcaagactccttcacacctgtggtgaactctctagacccaccgttactgactcgctaccttcgaattcacccccagagttgggtgcaccagattgccctgaggatggaggttctgggctgcgaggcacaggacctctac'.upper())\n",
    "\n",
    "# VWF (block1_notag) (3_1)\n",
    "VWF_SB1 = Seq('ATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTGACTTCGTCAACACCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCATTTGTTTGTCAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTTCAAGTCCTGCTGTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTGGTGTGAACGGGCATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAGACTTTGTGTGAGTGTGCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTGCGCCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCGGCACCTCCCTCTCTCGAGACTGCAACACaTGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGCCAGTACCTGCTGGCCCGGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTGCCATGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTATGCCGGGAAGACCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTG'.upper())\n",
    "\n",
    "# VWF (block2_notag) (2_2)\n",
    "VWF_SB2 = Seq('CAGGACCTGCAGAAGCAGCACAGCGATCCCTGCGCCCTCAACCCGCGCATGACCAGGTTCTCCGAGGAGGCGTGCGCGGTCCTGACGTCCCCCACATTCGAGGCCTGCCATCGTGCCGTCAGCCCGCTGCCCTACCTGCGGAACTGCCGCTACGACGTGTGCTCCTGCTCGGACGGCCGCGAGTGCCTGTGCGGCGCCCTGGCCAGCTATGCCGCGGCCTGCGCGGGGAGAGGCGTGCGCGTCGCGTGGCGCGAGCCAGGCCGCTGTGAGCTGAACTGCCCGAAAGGCCAGGTGTACCTGCAGTGCGGGACCCCCTGCAACCTGACCTGCCGCTCTCTCTCTTACCCGGATGAGGAATGCAATGAGGCCTGCCTGGAGGGCTGCTTCTGCCCCCCAGGGCTCTACATGGATGAGAGGGGGGACTGCGTGCCCAAGGCCCAGTGCCCCTGTTACTATGACGGTGAGATCTTCCAGCCAGAAGACATCTTCTCAGACCATCACACCATGTGCTACTGTGAGGATGGCTTCATGCACTGTACCATGAGTGGAGTCCCCGGAAGCTTGCTGCCTGACGCTGTCCTCAGCAGTCCCCTGTCTCATCGCAGCAAAAGGAGCCTATCCTGTCGGCCCCCCATGGTCAAGCTGGTGTGTCCCGCTGACAACCTGCGGGCTGAAGGGCTCGAGTGTACCAAAACGTGCCAGAACTATGACCTGGAGTGCATGAGCATGGGCTGTGTCTCTGGCTGCCTCTGCCCCCCGGGCATGGTCCGGCATGAGAACAGATGTGTGGCCCTGGAAAGGTGTCCCTGCTTCCATCAGGGCAAGGAGTATGCCCCTGGAGAAACAGTGAAGATTGGCTGCAACACTTGTGTCTGTCAGGACCGGAAGTGGAACTGCACAGACCATGTGTGTGATGCCACGTGCTCCACGATCGGCATGGCCCACTACCTCACCTTCGACGGGCTCAAATACCTGTTCCCCGGGGAGTGCCAGTACGTTCTGGTGCAGGATTACTGCGGCAGTAACCCTGGGACCTTTCGGATCCTAGTGGGGAATAAGGGATGCAGCCACCCCTCAGTGAAATGCAAGAAACGGGTCACCATCCTGGTGGAGGGAGGAGAGATTGAGCTGTTTGACGGGGAGGTGAATGTGAAGAGGCCCATGAAGGATGAGACTCACTTTGAGGTGGTGGAGTCTGGCCGGTACATCATTCTGCTGCTGGGCAAAGCCCTCTCCGTGGTCTGGGACCGCCACCTGAGCATCTCCGTGGTCCTGAAGCAGACATACCAGGAGAAAGTGTGTGGCCTGTGTGGGAATTTTGATGGCATCCAGAACAATGACCTCACCAGCAGCAACCTCCAAGTGGAGGAAGACCCTGTGGACTTTGGGAACTCCTGGAAAGTGAGCTCGCAGTGTGCTGACACCAGAAAAGTGCCTCTGGACTCATCCCCTGCCACaTGCCATAACAACATCATGAAGCAGACGATGGTGGATTCCTCCTGTAGAATCCTTACCAGTGACGTCTTCCAGGACTGCAACAAGCTGGTGGACCCCGAGCCATATCTGGATGTCTGCATTTACGACACaTGCTCCTGTGAGTCCATTGGGGACTGCGCCTGCTTCTGCGACACCATTGCTGCCTATGCCCACGTGTGTGCCCAGCATGGCAAGGTGGTGACCTGGAGGACGGCCACATTGTGCCCCCAGAGCTGCGAGGAGAGGAATCTCCGG'.upper())\n",
    "\n",
    "# VWF (block3_notag) (5_3)\n",
    "VWF_SB3 = Seq('CGGGAGAACGGGTATGAGTGTGAGTGGCGCTATAACAGCTGTGCACCcGCCTGTCAAGTCACGTGTCAGCACCCTGAGCCACTGGCCTGCCCTGTGCAGTGTGTGGAGGGCTGCCATGCCCACTGCCCTCCAGGGAAAATCCTGGATGAGCTTTTGCAGACCTGCGTTGACCCTGAAGACTGTCCAGTGTGTGAGGTGGCTGGCCGGCGTTTTGCCTCAGGAAAGAAAGTCACCTTGAATCCCAGTGACCCTGAGCACTGCCAGATTTGCCACTGTGATGTTGTCAACCTCACCTGTGAAGCCTGCCAGGAGCCGGGAGGCCTGGTGGTGCCTCCCACAGATGCCCCGGTGAGCCCCACCACTCTGTATGTGGAGGACATCTCGGAACCGCCGTTGCACGATTTCTACTGCAGCAGGCTACTGGACCTGGTCTTCCTGCTGGATGGCTCCTCCAGGCTGTCCGAGGCTGAGTTTGAAGTGCTGAAGGCCTTTGTGGTGGACATGATGGAGCGGCTGCGCATCTCCCAGAAGTGGGTCCGCGTGGCCGTGGTGGAGTACCACGACGGCTCCCACGCCTACATCGGGCTCAAGGACCGGAAGCGACCGTCAGAGCTGCGGCGCATTGCCAGCCAGGTGAAGTATGCGGGCAGCCAGGTGGCCTCCACCAGCGAGGTCTTGAAATACACACTGTTCCAAATCTTCAGCAAGATCGACCGCCCTGAAGCCTCCCGCATCACCCTGCTCCTGATGGCCAGCCAGGAGCCCCAACGGATGTCCCGGAACTTTGTCCGCTACGTCCAGGGCCTGAAGAAGAAGAAGGTCATTGTGATCCCGGTGGGCATTGGGCCCCATGCCAACCTCAAGCAGATCCGCCTCATCGAGAAGCAGGCCCCTGAGAACAAGGCCTTCGTGCTGAGCAGTGTGGATGAGCTGGAGCAGCAAAGGGACGAGATCGTTAGCTACCTCTGTGACCTTGCCCCTGAAGCCCCTCCTCCTACTCTGCCCCCCGACATGGCACAAGTCACTGTGGGCCCGGGGCTCTTGGGGGTTTCGACCCTGGGGCCCAAGAGGAACTCCATGGTTCTGGATGTGGCGTTCGTCCTGGAAGGATCGGACAAAATTGGTGAAGCCGACTTCAACAGGAGCAAGGAGTTCATGGAGGAGGTGATTCAGCGGATGGATGTGGGCCAGGACAGCATCCACGTCACGGTGCTGCAGTACTCCTACATGGTGACTGTGGAGTACCCCTTCAGCGAGGCACAGTCCAAAGGGGACATCCTGCAGCGGGTGCGAGAGATCCGCTACCAGGGCGGCAACAGGACCAACACTGGGCTGGCCCTGCGGTACCTCTCTGACCACAGCTTCTTGGTCAGCCAGGGTGACCGGGAGCAGGCGCCCAACCTGGTCTACATGGTCACCGGAAATCCTGCCTCTGATGAGATCAAGAGGCTGCCTGGAGACATCCAGGTGGTGCCCATTGGAGTGGGCCCTAATGCCAACGTGCAGGAGCTGGAGAGGATTGGCTGGCCCAATGCCCCTATCCTCATCCAGGACTTTGAAACGCTCCCCCGAGAGGCTCCTGACCTGGTGCTGCAGAGGTGCTGCTCCGGAGAGGGGCTGCAGATCCCCACCCTCTCCCCTGCACCTGACTGCAGCCAGCCCCTGGACGTGATCCTTCTCCTGGATGGCTCCTCC'.upper())\n",
    "\n",
    "# VWF (block4_notag) (3_4)\n",
    "VWF_SB4 = Seq('GTGATCCTTCTCCTGGATGGCTCCTCCAGTTTCCCAGCTTCTTATTTTGATGAAATGAAGAGTTTCGCCAAGGCTTTCATTTCAAAAGCCAATATAGGGCCTAGACTCACTCAGGTGTCAGTGCTGCAGTATGGAAGCATCACCACCATTGACGTGCCATGGAACGTGGTCCCGGAGAAAGCCCATTTGCTGAGCCTTGTGGACGTCATGCAGCGGGAGGGAGGCCCCAGCCAAATCGGGGATGCCTTGGGCTTTGCTGTGCGATACTTGACTTCAGAAATGCATGGTGCCAGGCCGGGAGCCTCAAAGGCGGTGGTCATCCTGGTCACGGACGTGTCTGTGGATTCAGTGGATGCAGCAGCTGATGCCGCCAGGTCCAACAGAGTGACAGTGTTCCCTATTGGAATTGGAGATCGCTACGATGCAGCCCAGCTACGGATCTTGGCAGGCCCAGCAGGCGACTCCAACGTGGTGAAGCTCCAGCGAATCGAAGACCTCCCTACCATGGTCACCTTGGGCAATTCCTTCCTCCACAAACTGTGCTCTGGATTTGTTAGGATTTGCATGGATGAGGATGGGAATGAGAAGAGGCCCGGGGACGTCTGGACCTTGCCAGACCAGTGCCACACCGTGACTTGCCAGCCAGATGGCCAGACCTTGCTGAAGAGTCATCGGGTCAACTGTGACCGGGGGCTGAGGCCTTCGTGCCCTAACAGCCAGTCCCCTGTTAAAGTGGAAGAaACCTGTGGCTGCCGCTGGACCTGCCCCTGCGTGTGCACAGGCAGCTCCACTCGGCACATCGTGACCTTTGATGGGCAGAATTTCAAGCTGACTGGCAGCTGTTCTTATGTCCTATTTCAAAACAAGGAGCAGGACCTGGAGGTGATTCTCCATAATGGTGCCTGCAGCCCTGGAGCAAGGCAGGGCTGCATGAAATCCATCGAGGTGAAGCACAGTGCCCTCTCCGTCGAGCTGCACAGTGACATGGAGGTGACGGTGAATGGGAGACTGGTgTCTGTTCCTTACGTGGGTGGGAACATGGAAGTCAACGTTTATGGTGCCATCATGCATGAGGTCAGATTCAATCACCTTGGTCACATCTTCACATTCACTCCACAAAACAATGAGTTCCAACTGCAGCTCAGCCCCAAGACTTTTGCTTCAAAGACGTATGGTCTGTGTGGGATCTGTGATGAGAACGGAGCCAATGACTTCATGCTGAGGGATGGCACAGTCACCACAGACTGGAAAACACTTGTTCAGGAATGGACTGTGCAGCGGCCAGGGCAGACGTGCCAGCCCATCCTGGAGGAGCAGTGTCTTGTCCCCGACAGCTCCCACTGCCAGGTCCTCCTCTTACCACTGTTTGCTGAATGCCACAAGGTCCTGGCTCCAGCCACATTCTATGCCATCTGCCAGCAGGACAGTTGCCACCAGGAGCAAGTGTGTGAGGTGATCGCCTCTTATGCCCACCTCTGTCGGACCAACGGGGTCTGCGTTGACTGGAGGACACCTGATTTCTGTGCTATGTCATGCCCACCATCTCTGGTCTACAACCACTGTGAGCATGGCTGTCCCCGGCACTGTGATGGCAACGTGAGCTCCTGTGGGGACCATCCCTCCGAAGGCTGTTTCTGCCCTCCAGATAAAGTCATGTTGGAAGGCAGCTGTGTCCCTGAAGAGGCCTGC'.upper())\n",
    "\n",
    "# VWF (block5_ctag) (2_5)\n",
    "VWF_SB5 = Seq('GAGGCCTGCACTCAGTGCATTGGTGAGGATGGAGTCCAGCACCAGTTCCTGGAAGCCTGGGTCCCGGACCACCAGCCCTGTCAGATCTGCACATGCCTCAGCGGGCGGAAGGTCAACTGCACAACGCAGCCCTGCCCCACGGCCAAAGCTCCCACGTGTGGCCTGTGTGAAGTAGCCCGCCTCCGCCAGAATGCAGACCAGTGCTGCCCCGAGTATGAGTGTGTGTGTGACCCAGTGAGCTGTGACCTGCCCCCAGTGCCTCACTGTGAACGTGGCCTCCAGCCCACACTGACCAACCCTGGCGAGTGCAGACCCAACTTCACaTGCGCCTGCAGGAAGGAGGAGTGCAAAAGAGTGTCCCCACCCTCCTGCCCCCCGCACCGTTTGCCCACCCTTCGGAAGACCCAGTGCTGTGATGAGTATGAGTGTGCCTGCAACTGTGTCAACTCCACAGTGAGCTGTCCCCTTGGGTACTTGGCCTCAACTGCCACCAATGACTGTGGCTGTACCACAACCACaTGCCTTCCCGACAAGGTGTGTGTCCACCGAAGCACCATCTACCCTGTGGGCCAGTTCTGGGAGGAGGGCTGCGATGTGTGCACaTGCACCGACATGGAGGATGCCGTGATGGGCCTCCGCGTGGCCCAGTGCTCCCAGAAGCCCTGTGAGGACAGCTGTCGGTCGGGCTTCACTTACGTTCTGCATGAAGGCGAGTGCTGTGGAAGGTGCCTGCCATCTGCCTGTGAGGTGGTGACTGGCTCACCGCGGGGGGACTCCCAGTCTTCCTGGAAGAGTGTCGGCTCCCAGTGGGCCTCCCCGGAGAACCCCTGCCTCATCAATGAGTGTGTCCGAGTGAAGGAGGAGGTCTTTATACAACAAAGGAACGTGTCCTGCCCCCAGCTGGAGGTCCCTGTCTGCCCCTCGGGCTTTCAGCTGAGCTGTAAGACCTCAGCGTGCTGCCCAAGCTGTCGCTGTGAGCGCATGGAGGCCTGCATGCTCAATGGCACTGTCATTGGGCCCGGGAAGACTGTGATGATCGATGTGTGCACGACCTGCCGCTGCATGGTGCAGGTcGGGGTCATCTCTGGATTCAAGCTGGAGTGCAGGAAGACCACaTGCAACCCCTGCCCCCTGGGTTACAAGGAAGAAAATAACACAGGTGAATGTTGTGGGAGATGTTTGCCTACGGCTTGCACCATTCAGCTAAGAGGAGGACAGATCATGACACTGAAGCGTGATGAAACGCTCCAGGATGGCTGTGATACTCACTTCTGCAAGGTCAATGAGAGAGGAGAGTACTTCTGGGAGAAGAGGGTCACAGGCTGCCCACCCTTTGATGAACACAAGTGTCTGGCTGAGGGAGGTAAAATTATGAAAATTCCAGGCACaTGCTGTGACACATGTGAGGAGCCTGAGTGCAACGACATCACTGCCAGGCTGCAGTATGTCAAGGTGGGAAGCTGTAAGTCTGAAGTAGAGGTGGATATCCACTACTGCCAGGGCAAATGTGCCAGCAAAGCCATGTACTCCATTGACATCAACGATGTGCAGGACCAGTGCTCCTGCTGCTCTCCGACACGGACGGAGCCCATGCAGGTcGCCCTGCACTGCACCAATGGCTCTGTTGTGTACCATGAGGTTCTCAATGCCATGGAGTGCAAATGCTCCCCCAGGAAGTGCAGCAAG'.upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "13c09211-8c41-48b8-8a94-71559c507266",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Processing forward primers against FRED1\n",
      "Found non-specific match at 1280bp:\n",
      " match:GCTTTCCAGTAGGATACACC\n",
      "primer:AAGGCCCAGAAGGATACAAC Tm:20.43270409953766\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/lib/python3.13/site-packages/primer3/bindings.py:305: UserWarning: Function deprecated please use \"calc_heterodimer\" instead\n",
      "  return THERMO_ANALYSIS.calcHeterodimer(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Processing forward primers against FRED2\n",
      "Found non-specific match using Primer3 at 1586bp:\n",
      " match:TAGCAGATGAGGAGAGGGCC\n",
      "primer:GACCATGCAAGGAGAGGTAC Tm:19.7\n",
      "\n",
      " Processing forward primers against FRED3\n",
      "Found non-specific match at 548bp:\n",
      " match:ATCTGCTCAGCATGGGCAGC\n",
      "primer:AGTATCTCAGCAAGGGCAAC Tm:25.3\n",
      "\n",
      " Processing forward primers against VWF1\n",
      "Found non-specific match at 1006bp:\n",
      " match:CTCCTGGATGAAGGCCTCTG\n",
      "primer:AAAGCACTCTTAGGCCTCTG Tm:23.5\n",
      "Found non-specific match at 1478bp:\n",
      " match:AAGGAGCGTTTCTGGCAGCC\n",
      "primer:AATCAGTTTCTTTGGCAGCC Tm:22.038901619925298\n",
      "\n",
      " Processing forward primers against VWF2\n",
      "Found non-specific match at 29bp:\n",
      " match:CCTGCGCCCTCAACCCGCGC\n",
      "primer:AGTTGTAATATCACCCGCGC Tm:32.6\n",
      "Found non-specific match at 1492bp:\n",
      " match:GCGCACGCCTCTCCCCGCGC\n",
      "primer:AGTTGTAATATCACCCGCGC Tm:26.79564064854054\n",
      "Found non-specific match at 330bp:\n",
      " match:CCGCTCTCTCTCTTACCCGG\n",
      "primer:AATAGGAACCTCTTACGCGG Tm:23.0\n",
      "Found non-specific match at 184bp:\n",
      " match:TGCCTGTGCGGCGCCCTGGC\n",
      "primer:ACATTAAATTTCGCCGTGGC Tm:21.3\n",
      "\n",
      " Processing forward primers against VWF3\n",
      "Found non-specific match at 1375bp:\n",
      " match:CAGGGTGACCGGGAGCAGGC\n",
      "primer:TCCAATTATACGGAGCAGGC Tm:27.4\n",
      "Found non-specific match at 1356bp:\n",
      " match:CCACAGCTTCTTGGTCAGCC\n",
      "primer:AATCAGTTTCTTTGGCAGCC Tm:21.2\n",
      "Found non-specific match at 1140bp:\n",
      " match:ACTCCACCACGGCCACGCGG\n",
      "primer:ATACTGTAAGAACCACGCGG Tm:29.873359153167996\n",
      "\n",
      " Processing forward primers against VWF4\n",
      "Found non-specific match at 1311bp:\n",
      " match:GGAGCAGTGTCTTGTCCCCG\n",
      "primer:GATACATAGACTTGGCCCCG Tm:21.9\n",
      "\n",
      " Processing forward primers against VWF5\n",
      "Found non-specific match at 689bp:\n",
      " match:AGCATGCAGGCCTCCATGCG\n",
      "primer:ATAGATGGTGCCTACATGCG Tm:25.47404059538661\n",
      "Found non-specific match at 1032bp:\n",
      " match:GGGAGCACTGGGCCACGCGG\n",
      "primer:ATACTGTAAGAACCACGCGG Tm:29.873359153167996\n",
      "Found non-specific match at 258bp:\n",
      " match:GCCTCACTGTGAACGTGGCC\n",
      "primer:TACTTGAATACCACGTGGCC Tm:24.2\n",
      "\n",
      " Processing reverse primers against FRED1\n",
      "Found non-specific match at 481bp:\n",
      " match:AAAGAGAATGGTCCAATGGC\n",
      "primer:ACGTTCAGTTTTCCAATGGC Tm:20.9\n",
      "\n",
      " Processing reverse primers against FRED2\n",
      "Found non-specific match at 993bp:\n",
      " match:ACAGCTCCATCACAGTGGGG\n",
      "primer:AGCTCAATCTAACAGTGGGG Tm:24.1\n",
      "\n",
      " Processing reverse primers against FRED3\n",
      "Found non-specific match at 1181bp:\n",
      " match:CTTTGATCCTGAGCCATTAC\n",
      "primer:AAGACGATCCGAGCCATTAC Tm:20.756945576310443\n",
      "\n",
      " Processing reverse primers against VWF1\n",
      "Found non-specific match at 355bp:\n",
      " match:TATCTAGAAACTGAGGCTGG\n",
      "primer:AGACAACAATCTGAGGCTGG Tm:28.1\n",
      "Found non-specific match at 1135bp:\n",
      " match:AGGGTCCGAGGTCAAGGTCC\n",
      "primer:AGTATTAGGCGTCAAGGTCC Tm:25.5344824334357\n",
      "Found non-specific match at 1128bp:\n",
      " match:AGTCATAAGGGTCCGAGGTC\n",
      "primer:AGTTATAAGGGTCCGATGCC Tm:24.266763984606712\n",
      "\n",
      " Processing reverse primers against VWF2\n",
      "Found non-specific match at 506bp:\n",
      " match:TGTGCTACTGTGAGGATGGC\n",
      "primer:CCACGAGATAAGAGGATGGC Tm:21.2\n",
      "Found non-specific match at 1644bp:\n",
      " match:TGCCCACGTGTGTGCCCAGC\n",
      "primer:AGTAAAGCATAGTGCCCAGC Tm:28.3\n",
      "Found non-specific match at 495bp:\n",
      " match:CGGAGAGGGCTTTGCCCAGC\n",
      "primer:AGTAAAGCATAGTGCCCAGC Tm:22.038901619925184\n",
      "Found non-specific match using Primer3 at 1545bp:\n",
      " match:GGTGGACCCCGAGCCATATC\n",
      "primer:AAGACGATCCGAGCCATTAC Tm:24.4\n",
      "Found non-specific match at 729bp:\n",
      " match:CCAGAACGTACTGGCACTCC\n",
      "primer:ACAGAACGAACAGGCACTAC Tm:32.41159257503773\n",
      "\n",
      " Processing reverse primers against VWF3\n",
      "\n",
      " Processing reverse primers against VWF4\n",
      "Found non-specific match at 209bp:\n",
      " match:TGCAGCGGGAGGGAGGCCCC\n",
      "primer:GCTAAATAGAGGGAAGCCCC Tm:27.5\n",
      "\n",
      " Processing reverse primers against VWF5\n",
      "Found non-specific match at 906bp:\n",
      " match:GGTCCCTGTCTGCCCCTCGG\n",
      "primer:AAGAATTACTGACCCCTCGG Tm:25.1\n"
     ]
    }
   ],
   "source": [
    "# Define genes and check primers against them\n",
    "genes = {\n",
    "           'FRED1':FRED_SB1,\n",
    "           'FRED2':FRED_SB2,\n",
    "           'FRED3':FRED_SB3,\n",
    "           'VWF1':VWF_SB1,\n",
    "           'VWF2':VWF_SB2,\n",
    "           'VWF3':VWF_SB3,\n",
    "           'VWF4':VWF_SB4,\n",
    "           'VWF5':VWF_SB5\n",
    "        }\n",
    "\n",
    "forward_primer_array = np.zeros((len(orthogonal_F['PrimerEnd']),len(genes)))\n",
    "for j, genename in enumerate(genes.keys()):\n",
    "    print('\\n Processing forward primers against ' + genename)\n",
    "    gene = genes[genename]\n",
    "    for i, primer in enumerate(orthogonal_F['PrimerEnd']):\n",
    "        forward_primer_array[i,j] = check_nonspecific(primer, gene)\n",
    "        \n",
    "reverse_primer_array = np.zeros((len(orthogonal_R['PrimerEnd']),len(genes)))\n",
    "for j, genename in enumerate(genes.keys()):\n",
    "    print('\\n Processing reverse primers against ' + genename)\n",
    "    gene = genes[genename]\n",
    "    for i, primer in enumerate(orthogonal_R['PrimerEnd']):\n",
    "        reverse_primer_array[i,j] = check_nonspecific(primer, gene)\n",
    "        \n",
    "# Compute number of nonspecific binding sites for each primer\n",
    "orthogonal_F['Num_Nonspecific_Binding_Sites'] = np.sum(forward_primer_array, axis=1)\n",
    "orthogonal_R['Num_Nonspecific_Binding_Sites'] = np.sum(reverse_primer_array, axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a443b147-464c-4e4e-9652-4f821fe354a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Process primers\n",
    "orthogonal_F_touse = orthogonal_F[(orthogonal_F.BsaI_Site_Present == False) & \\\n",
    "                                  (orthogonal_F.Num_Nonspecific_Binding_Sites == 0) & \\\n",
    "                                  (orthogonal_F.Exclude == False)][['Well Position','PrimerEnd','Worse']]\n",
    "orthogonal_R_touse = orthogonal_R[(orthogonal_R.BsaI_Site_Present == False) & \\\n",
    "                                  (orthogonal_R.Num_Nonspecific_Binding_Sites == 0) & \\\n",
    "                                  (orthogonal_R.Exclude == False)][['Well Position','PrimerEnd','Worse']]\n",
    "\n",
    "# Remove primer sequences manually - they bind to one of the genes on the gene list and don't pass oligo qc\n",
    "primer_seqs_to_remove = ['AGTTGTAATATCACCCGCGC', 'ACAGAACGAACAGGCACTAC' ,'ATACTGTAAGAACCACGCGG']\n",
    "orthogonal_F_touse_filt = orthogonal_F_touse[~orthogonal_F_touse['PrimerEnd'].isin(primer_seqs_to_remove)]\n",
    "orthogonal_R_touse_filt = orthogonal_R_touse[~orthogonal_R_touse['PrimerEnd'].isin(primer_seqs_to_remove)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "feeb72e8-013e-46d4-a0dd-95c91ca6bf89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Block</th>\n",
       "      <th>Validated F</th>\n",
       "      <th>Validated R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARAFcterm</td>\n",
       "      <td>6</td>\n",
       "      <td>F1</td>\n",
       "      <td>A4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARAFnterm</td>\n",
       "      <td>5</td>\n",
       "      <td>A11</td>\n",
       "      <td>A11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARAFnterm</td>\n",
       "      <td>2</td>\n",
       "      <td>E10</td>\n",
       "      <td>F10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRAFcterm</td>\n",
       "      <td>6</td>\n",
       "      <td>A4</td>\n",
       "      <td>A4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRAFcterm</td>\n",
       "      <td>7</td>\n",
       "      <td>B4</td>\n",
       "      <td>B4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>ERBB2cterm</td>\n",
       "      <td>2</td>\n",
       "      <td>B2</td>\n",
       "      <td>A7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>ERBB2cterm</td>\n",
       "      <td>3</td>\n",
       "      <td>C2</td>\n",
       "      <td>B7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>ERBB2cterm</td>\n",
       "      <td>4</td>\n",
       "      <td>D2</td>\n",
       "      <td>C7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>ERBB2cterm</td>\n",
       "      <td>10</td>\n",
       "      <td>B3</td>\n",
       "      <td>A8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>SHP2</td>\n",
       "      <td>10</td>\n",
       "      <td>F4</td>\n",
       "      <td>G9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene  Block Validated F Validated R\n",
       "0    ARAFcterm      6          F1          A4\n",
       "1    ARAFnterm      5         A11         A11\n",
       "2    ARAFnterm      2         E10         F10\n",
       "3    BRAFcterm      6          A4          A4\n",
       "4    BRAFcterm      7          B4          B4\n",
       "..         ...    ...         ...         ...\n",
       "90  ERBB2cterm      2          B2          A7\n",
       "91  ERBB2cterm      3          C2          B7\n",
       "92  ERBB2cterm      4          D2          C7\n",
       "93  ERBB2cterm     10          B3          A8\n",
       "94        SHP2     10          F4          G9\n",
       "\n",
       "[95 rows x 4 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import prevalidated primer combos\n",
    "validated_primer_combos = pd.read_csv('./Validated Gene Primer Pair Combinations.csv')\n",
    "validated_primer_combos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e06b91b7-b5df-4f55-9246-22fdfed65a4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward wells validated but not usable:\n",
      "Reverse wells validated but not usable\n",
      "B10\n",
      "H7\n"
     ]
    }
   ],
   "source": [
    "# Remove validated combos that are not usable since they bind to other genes\n",
    "print('Forward wells validated but not usable:')\n",
    "validated_not_usable_F = []\n",
    "validated_not_usable_R = []\n",
    "for f_well in np.unique(validated_primer_combos['Validated F'].values):\n",
    "    if f_well not in orthogonal_F_touse_filt['Well Position'].values:\n",
    "        print(f_well)\n",
    "        validated_not_usable_F.append(f_well)\n",
    "        \n",
    "print('Reverse wells validated but not usable')\n",
    "for r_well in np.unique(validated_primer_combos['Validated R'].values):\n",
    "    if r_well not in orthogonal_R_touse_filt['Well Position'].values:\n",
    "        print(r_well)\n",
    "        validated_not_usable_R.append(r_well)\n",
    "\n",
    "# filter out unusable primers\n",
    "validated_primer_combos_filtered = validated_primer_combos[~validated_primer_combos['Validated F'].isin(validated_not_usable_F) & \\\n",
    "                                                           ~validated_primer_combos['Validated R'].isin(validated_not_usable_R)] \\\n",
    "                        .rename(columns={'Validated F':'Forward Name',\n",
    "                                                 'Validated R':'Reverse Name'})\n",
    "\n",
    "# add primer sequences\n",
    "validated_primer_combos_filtered = pd.merge(validated_primer_combos_filtered,\n",
    "         orthogonal_F[['Well Position','PrimerEnd']].rename(columns={'Well Position':'Forward Name',\n",
    "                                                                     'PrimerEnd': 'Forward Primer'}), on='Forward Name')\n",
    "validated_primer_combos_filtered = pd.merge(validated_primer_combos_filtered,\n",
    "         orthogonal_R[['Well Position','PrimerEnd']].rename(columns={'Well Position':'Reverse Name',\n",
    "                                                                     'PrimerEnd': 'Reverse Primer'}), on='Reverse Name')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "4813e39e-8472-4c92-8a47-d59c9e9de333",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate more random combos of orthogonal F and reverse primers that are not \"red\" and that are not validated\n",
    "# concatenate F and R primers into pandas dataframe\n",
    "niters = 2\n",
    "minlens_FR = min(len(orthogonal_F_touse_filt), len(orthogonal_R_touse_filt))\n",
    "orthogonal_primers_iter = []\n",
    "for i in range(niters):\n",
    "    orthogonal_primers_iter.append(pd.concat([orthogonal_F_touse_filt.sample(n=minlens_FR).reset_index().\\\n",
    "                                                  drop(['index'], axis=1).rename(columns={'Well Position':'Forward Name',\n",
    "                                                                                         'PrimerEnd':'Forward Primer'}),\n",
    "                                              orthogonal_R_touse_filt.sample(n=minlens_FR).reset_index().\\\n",
    "                                                  drop(['index'], axis=1).rename(columns={'Well Position':'Reverse Name',\n",
    "                                                                                         'PrimerEnd':'Reverse Primer'})],\n",
    "                                                axis=1)\\\n",
    "                                    .drop(columns='Worse'))\n",
    "orthogonal_primers_touse = pd.concat(orthogonal_primers_iter).drop_duplicates()\n",
    "    \n",
    "# remove any prevalidated combos\n",
    "orthogonal_primers_touse = orthogonal_primers_touse[~orthogonal_primers_touse[['Forward Name','Reverse Name']]\\\n",
    "                                .apply(lambda a: (validated_primer_combos_filtered[['Forward Name','Reverse Name']] == a).all(1).any(),axis=1)]\\\n",
    "                                .reset_index(drop=True)\n",
    "orthogonal_primers_touse.to_csv('orthogonal_primer_random_combos_used_DH_091125.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "f88ae21b-1cae-43bf-8094-91b1f2e74fed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAAA</th>\n",
       "      <th>AAAC</th>\n",
       "      <th>AAAG</th>\n",
       "      <th>AAAT</th>\n",
       "      <th>AACA</th>\n",
       "      <th>AACC</th>\n",
       "      <th>AACG</th>\n",
       "      <th>AACT</th>\n",
       "      <th>AAGA</th>\n",
       "      <th>AAGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TTCG</th>\n",
       "      <th>TTCT</th>\n",
       "      <th>TTGA</th>\n",
       "      <th>TTGC</th>\n",
       "      <th>TTGG</th>\n",
       "      <th>TTGT</th>\n",
       "      <th>TTTA</th>\n",
       "      <th>TTTC</th>\n",
       "      <th>TTTG</th>\n",
       "      <th>TTTT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Overhang</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TTTT</th>\n",
       "      <td>636</td>\n",
       "      <td>9</td>\n",
       "      <td>41</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTT</th>\n",
       "      <td>4</td>\n",
       "      <td>477</td>\n",
       "      <td>5</td>\n",
       "      <td>46</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTT</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>597</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATTT</th>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>643</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTT</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>494</td>\n",
       "      <td>17</td>\n",
       "      <td>65</td>\n",
       "      <td>57</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>480</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAAA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>362</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAAA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>716</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAAA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>486</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>636</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>256 rows Ã— 256 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          AAAA  AAAC  AAAG  AAAT  AACA  AACC  AACG  AACT  AAGA  AAGC  ...  \\\n",
       "Overhang                                                              ...   \n",
       "TTTT       636     9    41    17     3     1     1     1     8     1  ...   \n",
       "GTTT         4   477     5    46     1    21     1     2     1    16  ...   \n",
       "CTTT         2     2   597     3     1     1    19     1     1     1  ...   \n",
       "ATTT         9     5     2   643     1     1     1     7     1     2  ...   \n",
       "TGTT         1     1     1     1   494    17    65    57     3     1  ...   \n",
       "...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   \n",
       "ACAA         1     1     1     1     1     1     1     1     1     1  ...   \n",
       "TAAA         1     1     1     1     1     1     1     1     1     1  ...   \n",
       "GAAA         1     1     1     1     1     1     1     1     1     1  ...   \n",
       "CAAA         1     1     1     1     1     1     1     1     1     1  ...   \n",
       "AAAA         1     1     1     1     1     1     1     1     1     1  ...   \n",
       "\n",
       "          TTCG  TTCT  TTGA  TTGC  TTGG  TTGT  TTTA  TTTC  TTTG  TTTT  \n",
       "Overhang                                                              \n",
       "TTTT         1     1     1     1     1     1     1     1     1     1  \n",
       "GTTT         1     1     1     1     1     1     1     1     1     1  \n",
       "CTTT         1     1     1     1     1     1     1     1     1     1  \n",
       "ATTT         1     1     1     1     1     1     1     1     1     1  \n",
       "TGTT         1     1     1     1     1     1     1     1     1     1  \n",
       "...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  \n",
       "ACAA         1     1    11     3     8   480     1     1     1     1  \n",
       "TAAA         1     1     1     1     1     1   362     1    11     4  \n",
       "GAAA         1     1     1     3     1     1     6   716     2    20  \n",
       "CAAA         5     1     1     1     1     1     4     1   486     1  \n",
       "AAAA         1     1     1     1     1     1     5     4     2   636  \n",
       "\n",
       "[256 rows x 256 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Import BsaI data\n",
    "bsaI_empirical = pd.read_csv('bsaI_empirical.csv')\n",
    "bsaI_empirical.index = bsaI_empirical['Overhang']\n",
    "bsaI_empirical = bsaI_empirical.drop(columns=['Overhang'])\n",
    "bsaI_empirical = bsaI_empirical + 1\n",
    "bsaI_empirical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c10bbf26-c403-47c0-900c-12b6a92cd2d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import information about codon usage for mutagenesis\n",
    "codons_ranked_by_usage = {\n",
    "    \"A\": [\"GCC\", \"GCT\", \"GCA\", \"GCG\"],\n",
    "    \"C\": [\"TGC\", \"TGT\"],\n",
    "    \"D\": [\"GAC\", \"GAT\"],\n",
    "    \"E\": [\"GAG\", \"GAA\"],\n",
    "    \"F\": [\"TTC\", \"TTT\"],\n",
    "    \"G\": [\"GGC\", \"GGA\", \"GGG\", \"GGT\"],\n",
    "    \"H\": [\"CAC\", \"CAT\"],\n",
    "    \"I\": [\"ATC\", \"ATT\", \"ATA\"],\n",
    "    \"K\": [\"AAG\", \"AAA\"],\n",
    "    \"L\": [\"CTG\", \"CTC\", \"CTT\", \"TTG\", \"TTA\", \"CTA\"],\n",
    "    \"M\": [\"ATG\"],\n",
    "    \"N\": [\"AAC\", \"AAT\"],\n",
    "    \"P\": [\"CCC\", \"CCT\", \"CCA\", \"CCG\"],\n",
    "    \"Q\": [\"CAG\", \"CAA\"],\n",
    "    \"R\": [\"CGG\", \"AGA\", \"AGG\", \"CGC\", \"CGA\", \"CGT\"],\n",
    "    \"S\": [\"AGC\", \"TCC\", \"TCT\", \"AGT\", \"TCA\", \"TCG\"],\n",
    "    \"T\": [\"ACC\", \"ACA\", \"ACT\", \"ACG\"],\n",
    "    \"V\": [\"GTG\", \"GTC\", \"GTT\", \"GTA\"],\n",
    "    \"W\": [\"TGG\"],\n",
    "    \"Y\": [\"TAC\", \"TAT\"],\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "0ee3a220-2309-48e8-a8f6-f98cb0e92266",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final overhang blacklist: ['AATT', 'ACGC', 'ACGT', 'AGCC', 'AGCG', 'AGCT', 'ATAT', 'CAAC', 'CACC', 'CACG', 'CAGC', 'CATG', 'CCAC', 'CCCC', 'CCGC', 'CCGG', 'CGAC', 'CGAG', 'CGCA', 'CGCC', 'CGCG', 'CGCT', 'CGGC', 'CGTC', 'CTAG', 'GACG', 'GATC', 'GATG', 'GCAC', 'GCCC', 'GCCG', 'GCGC', 'GCGG', 'GCGT', 'GCTG', 'GGCC', 'GGCG', 'GGCT', 'GGGC', 'GGGG', 'GGTC', 'GGTG', 'GTAC', 'GTCG', 'GTGC', 'GTGG', 'GTTG', 'TATA', 'TCGA', 'TGCA', 'TGCG', 'TTAA']\n"
     ]
    }
   ],
   "source": [
    "# Make the blacklist of overhangs to exclude\n",
    "# Tiny helper that works with either str or Seq\n",
    "def rc(s: str) -> str:\n",
    "    return str(Seq(str(s)).reverse_complement())\n",
    "\n",
    "# Build blacklist \n",
    "manual_blacklist = [\n",
    "    \"CGCC\",\"GCGG\",\"GGCG\",\"CGCG\",\"CCGC\",\"GGCC\",\"CGGC\",\"GGTG\",\"CACC\",\"CGCT\",\"GCGC\",\"GTCG\",\"GCCG\",\"AGCG\",\n",
    "    \"CACG\",\"GTGG\",\"GGCT\",\"GCTG\",\"CGAC\",\"GTTG\",\"GTGC\",\"GACG\",\"AGCC\",\"GGGC\",\"ACGC\",\"GCAC\",\"CGTC\",\n",
    "    \"CAGC\",\"GGGG\",\"GCGT\",\"CGCA\",\"GGTC\",\"TGCG\",\"GATG\",\"GCCC\",\"CGAG\"\n",
    "]\n",
    "\n",
    "overhang_blacklist = manual_blacklist.copy()\n",
    "\n",
    "# Ensure consistent casing\n",
    "bsaI_empirical.index = bsaI_empirical.index.map(lambda x: str(x).upper())\n",
    "bsaI_empirical.columns = bsaI_empirical.columns.map(lambda x: str(x).upper())\n",
    "\n",
    "for codon in bsaI_empirical.index:\n",
    "    codon_str = str(codon).upper()\n",
    "    rc_str = rc(codon_str)\n",
    "\n",
    "    # drop palindromes\n",
    "    if codon_str == rc_str:\n",
    "        overhang_blacklist.append(codon_str)\n",
    "        continue\n",
    "\n",
    "    # Look up efficiency\n",
    "    if (codon_str in bsaI_empirical.index) and (rc_str in bsaI_empirical.columns):\n",
    "        eff = bsaI_empirical.at[codon_str, rc_str]\n",
    "        if eff < 300:\n",
    "            overhang_blacklist.append(codon_str)\n",
    "    else:\n",
    "        print(f\"Warning: no data for {codon_str} â†’ {rc_str}\")\n",
    "\n",
    "# Deduplicate + sort\n",
    "overhang_blacklist = sorted(set(overhang_blacklist))\n",
    "print(\"Final overhang blacklist:\", overhang_blacklist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "e6cb5ca3-87e2-4a13-a102-428ddc2bf13a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set defaults\n",
    "block_size_range = [153, 198]\n",
    "max_oligo_size = 250\n",
    "first_last_block_reduction = 8\n",
    "slack = 5\n",
    "randomsequencepad = \"ACGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGTAATACCATGGTCCCTGAAATTCGGCTCAGTTACTTCGAGCGTAATGTCTCAAATGGCGTAGAACGGCAATGACTGTTTGACACTAGGTGGTGTTCAGTTCGGTAACGGAGAGTCTGTGCGGCATTCTTATTAATACATTTGAAACGCGCCCAACTGACGCTAGGCAAGTCAGTGCAGGCTCCCGTGTTAGGATAAGGGTAAACATACAAGTCGATAGAAGATGGGTAGGGGCCTTCAATTCATCCAGCACTCTACG\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "137da5af-559a-45ba-95a9-22c301610748",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5p libraries\n",
    "def post_qc(amp_primer_set, wt_oligos, primer_set, melt_temp_threshold = 35, check_all_primers=True):\n",
    "    print(\"Running QC for primer specificity on WT oligos\")\n",
    "    f_primer_map = {}\n",
    "    r_primer_map = {}\n",
    "    # invert the primer to subpool map\n",
    "    for k, v in amp_primer_set.items():\n",
    "        f_primer_map[v[1]] = f_primer_map.get(v[1], []) + [k]\n",
    "        r_primer_map[v[3]] = r_primer_map.get(v[3], []) + [k]\n",
    "    \n",
    "    # initialize list of nonspecific problems\n",
    "    nonspecific = {}\n",
    "    \n",
    "    # add unused primers if check_all_primers\n",
    "    if check_all_primers:\n",
    "        all_f_primers = np.unique(primer_set['Forward Primer'])\n",
    "        all_r_primers = np.unique(primer_set['Reverse Primer'])\n",
    "        for f_primer in all_f_primers:\n",
    "            if f_primer not in f_primer_map.keys():\n",
    "                f_primer_map[f_primer] = []\n",
    "        for r_primer in all_r_primers:\n",
    "            if r_primer not in r_primer_map.keys():\n",
    "                r_primer_map[r_primer] = []\n",
    "        \n",
    "    for f_primer, subpools_used in f_primer_map.items():\n",
    "    # iterate over every barcode primer pair and match to each oligo to check for nonspecific amplification\n",
    "        anneal_locs = []\n",
    "        for subpoolcheck, fragmentcheck in wt_oligos.items():  # iterate over every WT oligo\n",
    "            if (subpoolcheck not in subpools_used):  # ignore designed annealing (same name)\n",
    "                if check_nonspecific(f_primer, fragmentcheck, Tm_rem = melt_temp_threshold, verbose=False) > 0: #use high Tm_rem\n",
    "                    anneal_locs.append(subpoolcheck)\n",
    "        if anneal_locs:\n",
    "            nonspecific.update({f_primer:[a[0] + '_block' + str(a[1]+1) for a in anneal_locs]})\n",
    "    for r_primer, subpools_used in r_primer_map.items():\n",
    "    # iterate over every barcode primer pair and match to each oligo to check for nonspecific amplification\n",
    "        anneal_locs = []\n",
    "        for subpoolcheck, fragmentcheck in wt_oligos.items():  # iterate over every WT oligo\n",
    "            if (subpoolcheck not in subpools_used):  # ignore designed annealing (same name)\n",
    "                if check_nonspecific(r_primer, fragmentcheck, Tm_rem = melt_temp_threshold, verbose=False) > 0: #use high Tm_rem\n",
    "                    anneal_locs.append(subpoolcheck)\n",
    "        if anneal_locs:\n",
    "            nonspecific.update({r_primer:[a[0] + '_block' + str(a[1]+1) for a in anneal_locs]})\n",
    "    if nonspecific:\n",
    "        print(\"Nonspecific Primers: (Manually removing primer sequence recommended)\")\n",
    "        print(nonspecific)\n",
    "    else:\n",
    "        print(\"No non-specific primers detected\")\n",
    "        \n",
    "    return nonspecific\n",
    "\n",
    "def build_kmers(sequence, \n",
    "                ksize):\n",
    "    kmers = []\n",
    "    n_kmers = len(sequence) - ksize + 1\n",
    "\n",
    "    for i in range(n_kmers):\n",
    "        kmer = sequence[i:i + ksize]\n",
    "        kmers.append(kmer)\n",
    "\n",
    "    return kmers\n",
    "\n",
    "def compute_overlaps(breakpoints, \n",
    "                     inclusion_array, \n",
    "                     gene):\n",
    "    \n",
    "    overlaps = [[gene[val:val+4].reverse_complement(), gene[val:val+4]] for val in breakpoints]\n",
    "    counter = 0\n",
    "    for val in inclusion_array:\n",
    "        if val == -1:\n",
    "            (overlaps[counter][1],overlaps[counter+1][0]) = (overlaps[counter+1][0],overlaps[counter][1])\n",
    "            counter += 1\n",
    "        elif val == 0:\n",
    "            overlaps[counter][1] = overlaps[counter+1][1]\n",
    "            del overlaps[counter+1]\n",
    "        \n",
    "    return overlaps\n",
    "\n",
    "def score_breakpoints(gene, \n",
    "                      breakpoint_pair, \n",
    "                      empirical, \n",
    "                      overhang_blacklist=overhang_blacklist):\n",
    "    \n",
    "    #subset empirical matrix by the set of all overlaps\n",
    "    all_overlaps = []\n",
    "    for breakpoint in breakpoint_pair:\n",
    "        all_overlaps.append(gene[breakpoint:(breakpoint+4)])\n",
    "        all_overlaps.append(gene[breakpoint:(breakpoint+4)].reverse_complement())\n",
    "    all_overlaps = [str(o) for o in all_overlaps]\n",
    "    if (len(np.unique(all_overlaps)) == len(all_overlaps)) & (len(set(all_overlaps).intersection(set(overhang_blacklist))) == 0):\n",
    "        empirical_subset = empirical.loc[all_overlaps,all_overlaps]\n",
    "\n",
    "        #compute fidelity score\n",
    "        empirical_subset = empirical_subset/empirical_subset.sum(axis=1)\n",
    "        fidelity_score = 1\n",
    "        for breakpoint in breakpoint_pair:\n",
    "            fidelity_score = fidelity_score * \\\n",
    "                empirical_subset.loc[str(gene[breakpoint:(breakpoint+4)]),\n",
    "                                     str(gene[breakpoint:(breakpoint+4)].reverse_complement())]\n",
    "    \n",
    "    else:\n",
    "        fidelity_score = 0\n",
    "    \n",
    "    return fidelity_score\n",
    "    \n",
    "def optimize_breakpoints(gene, \n",
    "                         breakpoint_pair, \n",
    "                         indices_to_shift, \n",
    "                         indices_of_array,\n",
    "                         slack, \n",
    "                         empirical=bsaI_empirical, \n",
    "                         overhang_blacklist=overhang_blacklist):\n",
    "    \n",
    "    #compute all enrichments\n",
    "    shifts = list(range(-slack,slack+1))\n",
    "    if (len(indices_to_shift) > 2) | (len(indices_to_shift) < 1):\n",
    "        print('Error -- too many or too few breakpoints!')\n",
    "        optimum_breakpoint = breakpoint_pair\n",
    "        optimum_score = 0\n",
    "    elif (len(indices_to_shift) == 1): #external pair\n",
    "        scores = [0]*len(shifts)\n",
    "        for i,shift in enumerate(shifts):\n",
    "            scores[i] = score_breakpoints(gene, breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shift] + breakpoint_pair[(indices_to_shift[0]+1):], \n",
    "                                          empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "        \n",
    "        optimum_shift = np.argmax(scores)\n",
    "        optimum_breakpoint = breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shifts[optimum_shift]] + breakpoint_pair[(indices_to_shift[0]+1):]\n",
    "        optimum_score = scores[optimum_shift]\n",
    "        optimum_length = optimum_breakpoint[indices_of_array[1]] - optimum_breakpoint[indices_of_array[0]]\n",
    "            \n",
    "            \n",
    "    else: #internal pair\n",
    "        indices_to_shift = sorted(indices_to_shift)\n",
    "        scores = np.zeros((len(shifts),len(shifts)))\n",
    "        for i,shift1 in enumerate(shifts):\n",
    "            for j,shift2 in enumerate(shifts):\n",
    "                scores[i,j] = score_breakpoints(gene, breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shift1] + \\\n",
    "                                                            breakpoint_pair[(indices_to_shift[0]+1):indices_to_shift[1]] + \\\n",
    "                                                            [breakpoint_pair[indices_to_shift[1]]+shift2] + breakpoint_pair[(indices_to_shift[1]+1):], \n",
    "                                              empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "                \n",
    "        optimum_shift = np.unravel_index(np.argmax(scores,axis=None), scores.shape)\n",
    "        optimum_breakpoint = breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shifts[optimum_shift[0]]] + \\\n",
    "                                            breakpoint_pair[(indices_to_shift[0]+1):indices_to_shift[1]] + \\\n",
    "                                            [breakpoint_pair[indices_to_shift[1]]+shifts[optimum_shift[1]]] + breakpoint_pair[(indices_to_shift[1]+1):]\n",
    "        optimum_score = scores[optimum_shift]\n",
    "        optimum_length = optimum_breakpoint[indices_of_array[1]] - optimum_breakpoint[indices_of_array[0]] + 4\n",
    "    \n",
    "    return optimum_breakpoint, optimum_score, optimum_length\n",
    "\n",
    "def optimize_gene(gene, \n",
    "                  max_tile_size,\n",
    "                  first_last_block_reduction,\n",
    "                  block_size_range=block_size_range, \n",
    "                  slack=slack, \n",
    "                  empirical=bsaI_empirical, \n",
    "                  overhang_blacklist=overhang_blacklist): \n",
    "    \n",
    "    #setup initial inputs to optimization\n",
    "    gene_size = len(gene)\n",
    "    protein_size = len(gene.translate())\n",
    "        \n",
    "    #exclude gene if it is too big\n",
    "    if protein_size > 1000:\n",
    "        print('Protein size too big!')\n",
    "        \n",
    "    #divide genes between 500 and 1000aa into two blocks\n",
    "    elif protein_size > 620:\n",
    "        print('Protein size too big! Will add two superblock (620aa+ proteins) soon.')\n",
    "        \n",
    "    else:\n",
    "        #gene is one superblock \n",
    "        #print('Protein is one superblock.')\n",
    "        block_size = block_size_range[0] + np.argmin(\n",
    "            [abs((gene_size+2*first_last_block_reduction)/(i+block_size_range[0]) - \\\n",
    "                     round((gene_size+2*first_last_block_reduction)/(i+block_size_range[0]))) \\\n",
    "                 for i in range(0, block_size_range[1]-block_size_range[0])])\n",
    "        \n",
    "        \n",
    "        # now, set initial breakpoints\n",
    "        fragment_number = int((gene_size+2*first_last_block_reduction)/block_size)\n",
    "        \n",
    "        # if any of the tiles are too big?\n",
    "        tile_lengths = [1000]\n",
    "        while max(tile_lengths) > (max_tile_size-2*slack):\n",
    "            fragment_number = fragment_number + 1\n",
    "            first_breakpoint = 0\n",
    "            last_breakpoint = gene_size-4\n",
    "            step = (last_breakpoint - first_breakpoint + 2*first_last_block_reduction)/fragment_number\n",
    "            evenly_spaced_floats = [first_breakpoint] + [step * i - first_last_block_reduction for i in range(1,fragment_number)] + [last_breakpoint]        \n",
    "            initial_breakpoints = [[first_breakpoint,int(evenly_spaced_floats[1])+slack+2,last_breakpoint]] + \\\n",
    "                                    [[0,int(evenly_spaced_floats[i])-2-slack,int(evenly_spaced_floats[i+1])+slack+2,last_breakpoint] for i in range(1,len(evenly_spaced_floats)-2)] + \\\n",
    "                                    [[0,int(evenly_spaced_floats[-2])-2-slack,last_breakpoint]]\n",
    "            tile_lengths = [int(evenly_spaced_floats[1])+slack+2-first_breakpoint+first_last_block_reduction+4] + \\\n",
    "                                    [int(evenly_spaced_floats[i+1])-int(evenly_spaced_floats[i])+2*(slack+2)+4 for i in range(1,len(evenly_spaced_floats)-2)] + \\\n",
    "                                    [last_breakpoint+4-int(evenly_spaced_floats[-2])+slack+2+first_last_block_reduction]\n",
    "            \n",
    "\n",
    "        #optimize each breakpoint\n",
    "        optimum_breakpoints = []\n",
    "        optimum_scores = []\n",
    "        optimum_lengths = []\n",
    "        oligo_array_indices = []\n",
    "        for k,breakpoint in enumerate(initial_breakpoints):\n",
    "            if len(breakpoint) == 3:\n",
    "                indices_of_array = [0, 1] if k==0 else [1, 2]\n",
    "                optimum_breakpoint, optimum_score, optimum_length = optimize_breakpoints(gene, breakpoint, [1], indices_of_array,\n",
    "                                                            slack, empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "                optimum_breakpoints.append(optimum_breakpoint)\n",
    "                optimum_scores.append(optimum_score)\n",
    "                optimum_lengths.append(optimum_length)\n",
    "                oligo_array_indices.append(indices_of_array)\n",
    "            else:\n",
    "                indices_of_array = [1, 2]\n",
    "                optimum_breakpoint, optimum_score, optimum_length = optimize_breakpoints(gene, breakpoint, [1, 2], indices_of_array,\n",
    "                                                            slack, empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "                optimum_breakpoints.append(optimum_breakpoint)\n",
    "                optimum_scores.append(optimum_score)\n",
    "                optimum_lengths.append(optimum_length)\n",
    "                oligo_array_indices.append(indices_of_array)\n",
    "    \n",
    "    optimum_overlaps = [[str(gene[t:(t+4)]) for t in s] for s in optimum_breakpoints]\n",
    "    if all([s >= 0.95 for s in optimum_scores]):\n",
    "        print('All regions are high fidelity!')\n",
    "    elif all([s >= 0.9 for s in optimum_scores]):\n",
    "        print('Some regions are medium fidelity.')\n",
    "    else:\n",
    "        print('Some regions are low fidelity. Look closer')\n",
    "        \n",
    "    return optimum_breakpoints, optimum_overlaps, optimum_scores, optimum_lengths, oligo_array_indices\n",
    "\n",
    "def generate_primer(DNA_seq,\n",
    "                     Fwd=True,\n",
    "                     extendtoCG=False,\n",
    "                     smallest_primer_size=16,\n",
    "                     largest_primer_size=30,\n",
    "                     Tm=55):\n",
    "    \n",
    "    #Setup melting temperature arrays\n",
    "    melt_temp_array = np.zeros(largest_primer_size-smallest_primer_size+1)\n",
    "    \n",
    "    if Fwd:\n",
    "        DNA_seq_touse = DNA_seq\n",
    "    else:\n",
    "        DNA_seq_touse = DNA_seq.reverse_complement()\n",
    "            \n",
    "    #Make melting temperature arrays\n",
    "    primer_length = 0\n",
    "    for i in range(smallest_primer_size,largest_primer_size+1):\n",
    "        melt_temp_array[i-smallest_primer_size] = mt.Tm_NN(DNA_seq_touse[0:i])\n",
    "        \n",
    "        #Pick F primer when Tm is first >F_Tm\n",
    "        if (melt_temp_array[i-smallest_primer_size] >= Tm) & (primer_length==0):\n",
    "            primer_length = i\n",
    "    \n",
    "    #If Tm isnt high enough after max bases, just set primer length to be max and hope it works\n",
    "    if (primer_length == 0):\n",
    "        primer_length = largest_primer_size\n",
    "        \n",
    "    if extendtoCG:\n",
    "        while ((DNA_seq_touse[primer_length-1] == 'A') | (DNA_seq_touse[primer_length-1] == 'T')) & \\\n",
    "                    (primer_length < largest_primer_size):\n",
    "            primer_length += 1\n",
    "    \n",
    "    return DNA_seq_touse[0:primer_length]\n",
    "\n",
    "def make_mutations(region_name,\n",
    "                       region,\n",
    "                       region_flanks=[Seq(''),Seq('')],\n",
    "                       nt_start=0, #zero-indexed!\n",
    "                       wt_only=False,\n",
    "                       synonymous=True,\n",
    "                       stops='TAA',\n",
    "                       all3ntdeletions=True,\n",
    "                       mutation_list=False,\n",
    "                       codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                       aa_start=0):\n",
    "                       \n",
    "    oligo_array = {}\n",
    "    #Check that region has size divisible by three\n",
    "    if (len(region)/3 != len(region)//3) | (nt_start/3 != nt_start//3):\n",
    "        print('Region is not translatable!')\n",
    "        \n",
    "    else:\n",
    "        #add wt seq to oligo array\n",
    "        oligo_name = region_name + '_WT'\n",
    "        wt_seq = \\\n",
    "            region_flanks[0] + region + region_flanks[1]\n",
    "        oligo_array[oligo_name] = wt_seq\n",
    "        \n",
    "        if not wt_only:\n",
    "            \n",
    "            # see if a mutation list was given\n",
    "            if mutation_list == False:\n",
    "                    \n",
    "                #loop over amino acids\n",
    "                for j in range(0,len(region),3):\n",
    "\n",
    "                    #add all missense variants\n",
    "                    aa = region[j:(j+3)].translate()\n",
    "                    for aa_to in codons_ranked_by_usage.keys():\n",
    "                        if aa_to != aa:\n",
    "                            oligo_name = region_name + '_' + str(aa) + str((nt_start+j)//3+1+aa_start) + str(aa_to)\n",
    "                            seq_to_append = \\\n",
    "                                region_flanks[0] + \\\n",
    "                                region[0:j] + Seq(codons_ranked_by_usage[aa_to][0]) + \\\n",
    "                                region[(j+3):] + \\\n",
    "                                region_flanks[1]\n",
    "                            oligo_array[oligo_name] = seq_to_append\n",
    "\n",
    "                    #add synonymous variant if True and if possible, \n",
    "                    # using the most common codon that is NOT the codon in the gene\n",
    "                    if synonymous:\n",
    "                        if len(codons_ranked_by_usage[aa]) > 1:\n",
    "                            oligo_name = region_name + '_' + str(aa) + str((nt_start+j)//3+1+aa_start) + str(aa)\n",
    "                            possible_codons = codons_ranked_by_usage[aa].copy()\n",
    "                            possible_codons.remove(region[j:(j+3)])\n",
    "                            seq_to_append = \\\n",
    "                                region_flanks[0] + \\\n",
    "                                region[0:j] + Seq(possible_codons[0]) + \\\n",
    "                                region[(j+3):] + \\\n",
    "                                region_flanks[1]\n",
    "                            oligo_array[oligo_name] = seq_to_append\n",
    "\n",
    "                    #add stops if true\n",
    "                    if stops:\n",
    "                        oligo_name = region_name + '_' + str(aa) + str((nt_start+j)//3+1+aa_start) + 'X'\n",
    "                        seq_to_append = \\\n",
    "                            region_flanks[0] + \\\n",
    "                            region[0:j] + Seq(stops) + \\\n",
    "                            region[(j+3):] + \\\n",
    "                            region_flanks[1]\n",
    "                        oligo_array[oligo_name] = seq_to_append\n",
    "\n",
    "                    #add all 3nt deletions if True\n",
    "                    if all3ntdeletions:\n",
    "                        for k in range(0,3):\n",
    "                            if j+k+3 <= len(region):\n",
    "                                oligo_name = region_name + '_' + 'del' + str(nt_start+j+k+1+3*aa_start)\n",
    "                                seq_to_append = \\\n",
    "                                    region_flanks[0] + \\\n",
    "                                    region[0:(j+k)] + \\\n",
    "                                    region[(j+k+3):] + \\\n",
    "                                    region_flanks[1]\n",
    "                                oligo_array[oligo_name] = seq_to_append\n",
    "                                \n",
    "            else:\n",
    "                \n",
    "                #loop over mutation list\n",
    "                for i in range(len(mutation_list)):\n",
    "                        \n",
    "                    #iterate over every single aa change\n",
    "                    oligo_name = region_name + '_' + 'variant' + str(i+1)\n",
    "                    seq_to_append = region\n",
    "                    for k,v in enumerate(mutation_list[i]):\n",
    "                        aa_from = v[0]\n",
    "                        aa_to = v[-1]\n",
    "                        pos = int(v[1:-1])\n",
    "                        j=3*(pos-(nt_start//3+1))\n",
    "                        aa = region[j:(j+3)].translate()\n",
    "                        if aa != aa_from:\n",
    "                            print('Check mutation list!')\n",
    "                        else:\n",
    "                            seq_to_append = \\\n",
    "                                seq_to_append[0:j] + \\\n",
    "                                Seq(codons_ranked_by_usage[aa_to][0]) + \\\n",
    "                                seq_to_append[(j+3):]\n",
    "\n",
    "                    # append oligo to array\n",
    "                    seq_to_append = region_flanks[0] + \\\n",
    "                                    seq_to_append + \\\n",
    "                                    region_flanks[1]\n",
    "                    oligo_array[oligo_name] = seq_to_append\n",
    "        \n",
    "    return oligo_array\n",
    "\n",
    "\n",
    "def write_oligo_library(genes,\n",
    "                        oligo_file='./oligo_test.csv',\n",
    "                        primer_file='./primer_test.tsv',\n",
    "                        gbl_file='./gbl_test.tsv',\n",
    "                        gbl_large_file='./gbl_test_large.tsv',\n",
    "                        amp_primer_key_file='./amp_primer_key.tsv',\n",
    "                        breakpoint_file='./breakpoints.tsv',\n",
    "                        primer_set=orthogonal_primers_touse,\n",
    "                        codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                        block_size_range=block_size_range, \n",
    "                        max_oligo_size=max_oligo_size,\n",
    "                        slack=slack, \n",
    "                        empirical=bsaI_empirical, \n",
    "                        overhang_blacklist=overhang_blacklist,\n",
    "                        validated_primer_set=False,\n",
    "                        aa_start=False,\n",
    "                        wt_only=False,\n",
    "                        synonymous=True,\n",
    "                        stops='TAA',\n",
    "                        all3ntdeletions=True,\n",
    "                        mutations_to_use=False,\n",
    "                        find_pcr_primers=True,\n",
    "                        smallest_primer_size=16,\n",
    "                        largest_primer_size=30,\n",
    "                        Tm=55,\n",
    "                        extendtoCG=True,\n",
    "                        bsaI_firstoverlap='ACGA',\n",
    "                        bsaI_lastoverlap='TTTC',\n",
    "                        all_blocks=True,\n",
    "                        blocks_to_include=False,\n",
    "                        tile_boundaries=False,\n",
    "                        paqcIcapF=True,\n",
    "                        paqcIcapR=True,\n",
    "                        check_all_primers=True,\n",
    "                        qc_melt_temp_threshold=32,\n",
    "                        gblock_min_size=300,\n",
    "                        gblock_large_threshold=1000,\n",
    "                        randomsequencepad=randomsequencepad):\n",
    "    \n",
    "    #Split up primer set into F and R primers, cannot do more than 82 sublibraries\n",
    "    oligo_primer_counter = 0\n",
    "    oligo_array = {}\n",
    "    amp_primers = {}\n",
    "    gblocks = {}\n",
    "    num_blocks = {}\n",
    "    amp_primer_dict = {}\n",
    "    breakpoint_dict = {}\n",
    "    \n",
    "    #Convert genes to Seq and genes to list\n",
    "    gene_names = list(genes.keys())\n",
    "    genes = [Seq(genes[gene_name].upper()) for gene_name in gene_names]\n",
    "    \n",
    "    #PaqCI and BsaI site sequences and overhangs for paqcI\n",
    "    paqcI_seq = Seq('CACCTGC')\n",
    "    paqcI_overhang_nterm_ctag = Seq('CCAC')\n",
    "    paqcI_overhang_cterm_ctag = Seq('CGGG')\n",
    "    paqcI_overhang_nterm_ntag = Seq('TGGC')\n",
    "    paqcI_overhang_cterm_ntag = Seq('TAGG')\n",
    "    paqcI_seqplusfour = Seq('CACCTGCCTAG')\n",
    "    bsaI_seq = Seq('GGTCTC')\n",
    "    bsaI_seqplusone = Seq('GGTCTCT')\n",
    "    pcr_capseq = Seq('GGCTAC') + bsaI_seqplusone\n",
    "    gbl_capseq_F = Seq('CCGCGTGATTACGAGTCG') + pcr_capseq\n",
    "    gbl_capseq_R = Seq('GGGTTAGCAAGTGGCAGCCT') + pcr_capseq\n",
    "    \n",
    "    # set max size of a tile\n",
    "    primer_len = len(primer_set['Forward Primer'][0])\n",
    "    max_tile_size = max_oligo_size - 2*primer_len - 2*len(bsaI_seqplusone)\n",
    "    first_last_block_reduction = len(paqcI_seqplusfour)\n",
    "    \n",
    "    # iterate over genes\n",
    "    for r,gene in enumerate(genes):\n",
    "        \n",
    "        gene_name = gene_names[r]\n",
    "        print('Processing gene ' + str(r+1) + ': ' + gene_name)\n",
    "        \n",
    "        # amino acid to start at\n",
    "        if aa_start != False:\n",
    "            if gene_name in aa_start.keys():\n",
    "                aa_start_gene=aa_start[gene_name]-1\n",
    "            else:\n",
    "                aa_start_gene=0\n",
    "        else:\n",
    "            aa_start_gene=0\n",
    "        \n",
    "        #exclude if gene size is not divisible by three\n",
    "        if len(gene)/3 != len(gene)//3:\n",
    "            print('Gene length is not divisible by 3!')\n",
    "    \n",
    "        #exclude if there is a paqcI site in the gene\n",
    "        elif any([True for kmer in build_kmers(gene, len(paqcI_seq)) if kmer==paqcI_seq]) | \\\n",
    "            any([True for kmer in build_kmers(gene.reverse_complement(), len(paqcI_seq)) if kmer==paqcI_seq]):\n",
    "            print('Gene has paqcI site!')\n",
    "        \n",
    "        #exclude if there is a BsaI site in the gene\n",
    "        elif any([True for kmer in build_kmers(gene, len(bsaI_seq)) if kmer==bsaI_seq]) | \\\n",
    "            any([True for kmer in build_kmers(gene.reverse_complement(), len(bsaI_seq)) if kmer==bsaI_seq]):\n",
    "            print('Gene has BsaI site!')\n",
    "            \n",
    "        else:\n",
    "            print('Gene has no paqcI or BsaI sites! Performing GoldenGate optimization...')\n",
    "            \n",
    "            #cap gene with BsaI breakpoints and possible paqcI sites \n",
    "            if paqcIcapF & paqcIcapR:\n",
    "                gene_capped = bsaI_firstoverlap + paqcI_seqplusfour +  paqcI_overhang_nterm_ntag + \\\n",
    "                        gene + paqcI_overhang_cterm_ntag + paqcI_seqplusfour.reverse_complement() + bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap + paqcI_seqplusfour + paqcI_overhang_nterm_ntag)\n",
    "                capping_length_R = len(paqcI_overhang_cterm_ntag + paqcI_seqplusfour.reverse_complement() + \\\n",
    "                                   bsaI_lastoverlap)\n",
    "            elif paqcIcapF:\n",
    "                gene_capped = bsaI_firstoverlap + paqcI_seqplusfour + paqcI_overhang_nterm_ntag + \\\n",
    "                        gene + bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap + paqcI_seqplusfour + paqcI_overhang_nterm_ntag)\n",
    "                capping_length_R = len(bsaI_lastoverlap)\n",
    "            elif paqcIcapR:\n",
    "                gene_capped = bsaI_firstoverlap + \\\n",
    "                        gene + paqcI_overhang_cterm_ntag + paqcI_seqplusfour.reverse_complement() + bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap)\n",
    "                capping_length_R = len(paqcI_overhang_cterm_ntag + paqcI_seqplusfour.reverse_complement() + \\\n",
    "                                   bsaI_lastoverlap)\n",
    "            else:\n",
    "                gene_capped = bsaI_firstoverlap + gene + bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap)\n",
    "                capping_length_R = len(bsaI_lastoverlap)\n",
    "            \n",
    "            #Optimize gene if tile boundaries are not given\n",
    "            if tile_boundaries is False:\n",
    "                optimum_breakpoints, optimum_overlaps, optimum_scores, optimum_lengths, oligo_array_indices = \\\n",
    "                optimize_gene(gene_capped, \n",
    "                          max_tile_size=max_tile_size,\n",
    "                          first_last_block_reduction=first_last_block_reduction,\n",
    "                          block_size_range=block_size_range, \n",
    "                          slack=slack, \n",
    "                          empirical=bsaI_empirical, \n",
    "                          overhang_blacklist=overhang_blacklist)\n",
    "                pprint.pprint({'Optimum Breakpoints': optimum_breakpoints, \n",
    "                       'Optimum Overlaps': optimum_overlaps, \n",
    "                       'Optimum Scores': optimum_scores})\n",
    "                num_blocks[gene_name] = len(optimum_breakpoints)\n",
    "            else:\n",
    "                optimum_breakpoints = tile_boundaries[gene_name]\n",
    "                if len(optimum_breakpoints[0])==4:\n",
    "                    if len(optimum_breakpoints)>1:\n",
    "                        #multiple tiles, including one at beginning of gene\n",
    "                        oligo_array_indices = [[0,1]] + [[1,2]]*(len(optimum_breakpoints)-1)\n",
    "                    else:\n",
    "                        # one tile\n",
    "                        if ((optimum_breakpoints[1]-optimum_breakpoints[0]) > (optimum_breakpoints[2]-optimum_breakpoints[1])):\n",
    "                            # at end of gene\n",
    "                            oligo_array_indices = [[1,2]]\n",
    "                        else:\n",
    "                            # at beginning of gene\n",
    "                            oligo_array_indices = [[0,1]]\n",
    "                else:\n",
    "                    # multiple tiles, starting in the middle\n",
    "                    oligo_array_indices = [[1,2]]*(len(optimum_breakpoints))\n",
    "                num_blocks[gene_name] = len(optimum_breakpoints)\n",
    "                \n",
    "            \n",
    "            #add primers for gene_F and gene_R that are repeated constantly throughout the PCRs\n",
    "            #note: should probably prevalidate these primers!\n",
    "            if find_pcr_primers:\n",
    "                F_primer = generate_primer(gene,\n",
    "                                           Fwd=True,\n",
    "                                           extendtoCG=extendtoCG,\n",
    "                                           smallest_primer_size=smallest_primer_size,\n",
    "                                           largest_primer_size=largest_primer_size,\n",
    "                                           Tm=Tm)\n",
    "                F_primer = pcr_capseq + bsaI_firstoverlap + paqcI_seqplusfour + F_primer\n",
    "                amp_primers[gene_name+'_gene'+'_ampF'] = F_primer\n",
    "                R_primer = generate_primer(gene,\n",
    "                                           Fwd=False,\n",
    "                                           extendtoCG=extendtoCG,\n",
    "                                           smallest_primer_size=smallest_primer_size,\n",
    "                                           largest_primer_size=largest_primer_size,\n",
    "                                           Tm=Tm)\n",
    "                R_primer = pcr_capseq + Seq(bsaI_lastoverlap).reverse_complement() + \\\n",
    "                            paqcI_seqplusfour + R_primer\n",
    "                amp_primers[gene_name+'_gene'+'_ampR'] = R_primer\n",
    "            \n",
    "            #make oligos, primers, gblocks for each block\n",
    "            for i,breakpoint in enumerate(optimum_breakpoints):\n",
    "                \n",
    "                #find indices of breakpoint that correspond to oligo vs need to be PCRed/gblock\n",
    "                pcr_indices = [[j,j+1] for j in range(len(breakpoint)-1)]\n",
    "                pcr_indices.remove(oligo_array_indices[i])\n",
    "                \n",
    "                #find mutagenic window of oligo\n",
    "                oligo_breaks = [breakpoint[j] for j in oligo_array_indices[i]]\n",
    "                oligo_mutagenic_window = [int(3*np.ceil(max(oligo_breaks[0]+4-capping_length_F,3)/3)), int(3*np.floor(min(oligo_breaks[1]-capping_length_F,len(gene)-1)/3))]\n",
    "                \n",
    "                #subset the right block if needed\n",
    "                if (all_blocks == True) | ((i+1) in blocks_to_include[r] if blocks_to_include != False else True): #subset on allowed blocks\n",
    "                \n",
    "                    #add pcr primers and gblocks, one segment at a time\n",
    "                    for k,pcr_index in enumerate(pcr_indices):\n",
    "                        piece_name = gene_name + '_block' + str(i+1) + '_s' + str(k+1)\n",
    "                        pcr_breaks = [breakpoint[j] for j in pcr_index]\n",
    "                                            \n",
    "                        #get pcr primers\n",
    "                        if find_pcr_primers:\n",
    "                            if pcr_breaks[0] == breakpoint[0]: #Fragment beginning at gene start \n",
    "                                R_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=False,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                R_primer = pcr_capseq + R_primer\n",
    "                                amp_primers[piece_name+'_ampR'] = R_primer\n",
    "                            elif pcr_breaks[1] == breakpoint[-1]: #Fragment ending at gene end\n",
    "                                F_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=True,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                F_primer = pcr_capseq + F_primer\n",
    "                                amp_primers[piece_name+'_ampF'] = F_primer\n",
    "                            else:\n",
    "                                F_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=True,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                F_primer = pcr_capseq + F_primer\n",
    "                                R_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=False,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                R_primer = pcr_capseq + R_primer\n",
    "                                amp_primers[piece_name+'_ampF'] = F_primer\n",
    "                                amp_primers[piece_name+'_ampR'] = R_primer\n",
    "\n",
    "                        #make gblocks\n",
    "                        gbl = gbl_capseq_F + gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)] + gbl_capseq_R.reverse_complement()\n",
    "                        gblocks[piece_name] = gbl\n",
    "                        \n",
    "                    #add oligos to oligo array, checking first for validated primers if they are given\n",
    "                    validated=False\n",
    "                    if validated_primer_set is not False:\n",
    "                        validated_combo = validated_primer_set.query('Gene == @gene_name & Block == (@i+1)')\n",
    "                        if not validated_combo.empty:\n",
    "                            validated=True\n",
    "                            name_primer_F, primer_F, name_primer_R, primer_R = \\\n",
    "                                validated_combo[['Forward Name',\n",
    "                                                  'Forward Primer',\n",
    "                                                  'Reverse Name',\n",
    "                                                  'Reverse Primer']].values[0]\n",
    "                        else:\n",
    "                            name_primer_F, primer_F, name_primer_R, primer_R = \\\n",
    "                                primer_set.iloc[oligo_primer_counter,][['Forward Name',\n",
    "                                                                      'Forward Primer',\n",
    "                                                                      'Reverse Name',\n",
    "                                                                      'Reverse Primer']]\n",
    "                            oligo_primer_counter += 1\n",
    "                    else:\n",
    "                        name_primer_F, primer_F, name_primer_R, primer_R = \\\n",
    "                                primer_set.iloc[oligo_primer_counter,][['Forward Name',\n",
    "                                                                      'Forward Primer',\n",
    "                                                                      'Reverse Name',\n",
    "                                                                      'Reverse Primer']]\n",
    "                        oligo_primer_counter += 1\n",
    "                        \n",
    "                    # if mutations are given, use those - otherwise make all mutations or wtonly\n",
    "                    if mutations_to_use == False:\n",
    "                        add_on_array = make_mutations(gene_name + '_block' + str(i+1),\n",
    "                                           gene[oligo_mutagenic_window[0]:oligo_mutagenic_window[1]],\n",
    "                                           region_flanks=[Seq(primer_F) + \\\n",
    "                                                          bsaI_seqplusone + \\\n",
    "                                                          gene_capped[oligo_breaks[0]:(oligo_mutagenic_window[0]+capping_length_F)] ,\n",
    "                                                          gene_capped[(oligo_mutagenic_window[1]+capping_length_F):(oligo_breaks[1]+4)] + \\\n",
    "                                                          bsaI_seqplusone.reverse_complement() + \\\n",
    "                                                          Seq(primer_R).reverse_complement()],\n",
    "                                           nt_start=oligo_mutagenic_window[0],\n",
    "                                           wt_only=wt_only,\n",
    "                                           synonymous=synonymous,\n",
    "                                           stops=stops,\n",
    "                                           all3ntdeletions=all3ntdeletions,\n",
    "                                           codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                                           aa_start=aa_start_gene)\n",
    "                    else:\n",
    "                        add_on_array = make_mutations(gene_name + '_block' + str(i+1),\n",
    "                                           gene[oligo_mutagenic_window[0]:oligo_mutagenic_window[1]],\n",
    "                                           region_flanks=[Seq(primer_F) + \\\n",
    "                                                          bsaI_seqplusone + \\\n",
    "                                                          gene_capped[oligo_breaks[0]:(oligo_mutagenic_window[0]+capping_length_F)] ,\n",
    "                                                          gene_capped[(oligo_mutagenic_window[1]+capping_length_F):(oligo_breaks[1]+4)] + \\\n",
    "                                                          bsaI_seqplusone.reverse_complement() + \\\n",
    "                                                          Seq(primer_R).reverse_complement()],\n",
    "                                           nt_start=oligo_mutagenic_window[0],\n",
    "                                           mutation_list=mutations_to_use[(gene_name,i+1)],\n",
    "                                           codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                                           aa_start=aa_start_gene)\n",
    "                    oligo_array.update(add_on_array)\n",
    "                    amp_primer_dict.update({(gene_name,i+1): (name_primer_F,primer_F,name_primer_R,primer_R,validated)})\n",
    "                    breakpoint_dict.update({(gene_name,i+1): oligo_mutagenic_window})\n",
    "                    \n",
    "    #Check that max oligo is less than the max oligo length\n",
    "    if sum([len(s)>max_oligo_size for s in oligo_array.values()]) == 0:\n",
    "        print('All oligos are below the maximum 250bp!')\n",
    "    else:\n",
    "        print('Some oligos are TOO BIG!')\n",
    "        \n",
    "    #Check for nonspecific amplification\n",
    "    wt_oligos = {tuple([key.split('_')[0],\n",
    "                        int((key.split('_block')[1]).split('_')[0])]\n",
    "                      ):oligo_array[key] \\\n",
    "                     for key in oligo_array.keys() if 'WT' in key}\n",
    "    nonspecific_primers = post_qc(amp_primer_dict, \n",
    "                                  wt_oligos,\n",
    "                                  primer_set, \n",
    "                                  melt_temp_threshold=qc_melt_temp_threshold,\n",
    "                                  check_all_primers=check_all_primers)\n",
    "    \n",
    "    # Check that all mutagenic windows overlap\n",
    "    breakpoint_df = pd.DataFrame.from_dict(breakpoint_dict, orient='index', columns=['Mutagenesis Start','Mutagenesis End'])\n",
    "    breakpoint_df.index = pd.MultiIndex.from_tuples(breakpoint_dict.keys())\n",
    "    breakpoint_df = breakpoint_df.reset_index().rename(columns={'level_0':'Gene',\n",
    "                                                                'level_1':'Block'})\n",
    "    if (mutations_to_use == False) and (all_blocks == True):\n",
    "        missed_counter = 0\n",
    "        for r,gene_group_breakpoints in breakpoint_df.groupby('Gene'):\n",
    "            for k,row in gene_group_breakpoints.iterrows():\n",
    "                # look at whether the current row start is later than the last row end\n",
    "                if row['Block'] > 1:\n",
    "                    if row['Mutagenesis Start'] > end:\n",
    "                        missed_counter += 1\n",
    "                        print('Mutagenic window missed at ' + str(r) + ' block ' + str(row['Block']))\n",
    "                start,end = row['Mutagenesis Start'],row['Mutagenesis End']\n",
    "        if missed_counter == 0:\n",
    "            print('All mutagenic windows overlap!')\n",
    "        else:\n",
    "            print(str(missed_counter) + ' number of times the mutagenic window does not close!')\n",
    "                \n",
    "    #Remove any oligos with additional BsaI sites or paqcI sites\n",
    "    bad_oligos = []\n",
    "    for name,oligo in oligo_array.items():\n",
    "        #check for paqcI sites\n",
    "        paqcI_F = sum([True for kmer in build_kmers(oligo, len(paqcI_seq)) if kmer==paqcI_seq])\n",
    "        paqcI_R = sum([True for kmer in build_kmers(oligo.reverse_complement(), len(paqcI_seq)) if kmer==paqcI_seq])\n",
    "        #check that oligo is block 1 if it contains a paqcI site in the forward orientation \n",
    "        if paqcI_F > 0:\n",
    "            if ('block1' not in name) | (paqcI_F > 1):\n",
    "                bad_oligos.append(name)\n",
    "        #check that the oligo is block final if it contains a paqcI site in the reverse orientation\n",
    "        if paqcI_R > 0:\n",
    "            if ('block'+str(num_blocks[name.split('_')[0]]) not in name) | (paqcI_R > 1):\n",
    "                bad_oligos.append(name)\n",
    "        #check for more than one BsaI site\n",
    "        bsaI_F = sum([True for kmer in build_kmers(oligo, len(bsaI_seq)) if kmer==bsaI_seq])\n",
    "        bsaI_R = sum([True for kmer in build_kmers(oligo.reverse_complement(), len(bsaI_seq)) if kmer==bsaI_seq])\n",
    "        if (bsaI_F != 1) | (bsaI_R != 1):\n",
    "            bad_oligos.append(name)\n",
    "    bad_oligos=np.unique(bad_oligos)\n",
    "    for oligo_name in bad_oligos:\n",
    "        del oligo_array[oligo_name]\n",
    "    print(str(len(bad_oligos)) + ' oligos deleted due to errant restriction sites.')\n",
    "    \n",
    "    #Remove any duplicate oligos\n",
    "    new_dict = {}\n",
    "    seen_values = set()\n",
    "    counter=0\n",
    "    for key, value in oligo_array.items():\n",
    "        if value not in seen_values:\n",
    "            new_dict[key] = value\n",
    "            seen_values.add(value)\n",
    "        else:\n",
    "            counter += 1\n",
    "    print(str(counter) + ' oligos removed due to duplication.')\n",
    "    oligo_array = new_dict\n",
    "    del new_dict\n",
    "    \n",
    "    #write oligo array to file\n",
    "    with open(oligo_file, 'w') as f:\n",
    "        for key in oligo_array.keys():\n",
    "            f.write(\"%s,%s\\n\"%(key,oligo_array[key]))\n",
    "    f.close()\n",
    "            \n",
    "    #write primers to file\n",
    "    if find_pcr_primers:\n",
    "        primer_order_sheet = []\n",
    "        for key in amp_primers.keys():\n",
    "            primer_order_sheet.append(key + '\\t' + \\\n",
    "                     str(amp_primers[key]) + \\\n",
    "                     '\\t' + '25nm' + '\\t' + 'STD\\n')\n",
    "        print(*primer_order_sheet)\n",
    "        with open(primer_file, 'w') as f:\n",
    "            for line in primer_order_sheet:\n",
    "                f.write(line)\n",
    "        f.close()\n",
    "    \n",
    "    #write amplification primer key to file\n",
    "    amp_primer_key = ['Gene' + '\\t' + 'Block' + '\\t' + \\\n",
    "                      'Forward Primer Well' + '\\t' + 'Forward Primer' + '\\t' + \\\n",
    "                      'Reverse Primer Well' + '\\t' + 'Reverse Primer' + '\\t' + 'Validated' + '\\n']\n",
    "    for key in amp_primer_dict.keys():\n",
    "        genename, geneblock = key[0], str(key[1])\n",
    "        name_primer_F, primer_F, name_primer_R, primer_R, validated = amp_primer_dict[key]\n",
    "        amp_primer_key.append(genename + '\\t' + geneblock + '\\t' + \\\n",
    "                 name_primer_F + '\\t' + primer_F + '\\t' + \\\n",
    "                 name_primer_R + '\\t' + primer_R + '\\t' + str(validated) + '\\n')\n",
    "    print(*amp_primer_key)\n",
    "    with open(amp_primer_key_file, 'w') as f:\n",
    "        for line in amp_primer_key:\n",
    "            f.write(line)\n",
    "    f.close()\n",
    "    \n",
    "    #write breakpoint dict to file\n",
    "    breakpoint_df.to_csv(breakpoint_file, sep='\\t')\n",
    "    \n",
    "    #write gblocks to file\n",
    "    gblock_order_sheet = []\n",
    "    gblock_large_order_sheet = []\n",
    "    for key in gblocks.keys():\n",
    "        # pad gblock if it is not 300bp for Twist\n",
    "        if len(gblocks[key]) < gblock_min_size:\n",
    "            gblocks[key] = Seq(randomsequencepad[0:(gblock_min_size-len(gblocks[key]))]) + gblocks[key]\n",
    "        if len(gblocks[key]) < gblock_large_threshold:\n",
    "            gblock_order_sheet.append(key + '\\t' + \\\n",
    "                     str(gblocks[key]) + '\\n')\n",
    "        else:\n",
    "            gblock_large_order_sheet.append(key + '\\t' + \\\n",
    "                     str(gblocks[key]) + '\\n')\n",
    "    print(*gblock_order_sheet)\n",
    "    print(*gblock_large_order_sheet)\n",
    "    with open(gbl_file, 'w') as f:\n",
    "        for line in gblock_order_sheet:\n",
    "            f.write(line)\n",
    "    f.close()\n",
    "    with open(gbl_large_file, 'w') as f:\n",
    "        for line in gblock_large_order_sheet:\n",
    "            f.write(line)\n",
    "    f.close()\n",
    "    \n",
    "    return oligo_array,amp_primers,gblocks,amp_primer_dict,breakpoint_df\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6327f571-6e0f-45b0-8f23-dd4cdc94a00c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions for internal superblock sets\n",
    "def post_qc(amp_primer_set, wt_oligos, primer_set, melt_temp_threshold = 35, check_all_primers=True):\n",
    "    print(\"Running QC for primer specificity on WT oligos\")\n",
    "    f_primer_map = {}\n",
    "    r_primer_map = {}\n",
    "    # invert the primer to subpool map\n",
    "    for k, v in amp_primer_set.items():\n",
    "        f_primer_map[v[1]] = f_primer_map.get(v[1], []) + [k]\n",
    "        r_primer_map[v[3]] = r_primer_map.get(v[3], []) + [k]\n",
    "    \n",
    "    # initialize list of nonspecific problems\n",
    "    nonspecific = {}\n",
    "    \n",
    "    # add unused primers if check_all_primers\n",
    "    if check_all_primers:\n",
    "        all_f_primers = np.unique(primer_set['Forward Primer'])\n",
    "        all_r_primers = np.unique(primer_set['Reverse Primer'])\n",
    "        for f_primer in all_f_primers:\n",
    "            if f_primer not in f_primer_map.keys():\n",
    "                f_primer_map[f_primer] = []\n",
    "        for r_primer in all_r_primers:\n",
    "            if r_primer not in r_primer_map.keys():\n",
    "                r_primer_map[r_primer] = []\n",
    "        \n",
    "    for f_primer, subpools_used in f_primer_map.items():\n",
    "    # iterate over every barcode primer pair and match to each oligo to check for nonspecific amplification\n",
    "        anneal_locs = []\n",
    "        for subpoolcheck, fragmentcheck in wt_oligos.items():  # iterate over every WT oligo\n",
    "            if (subpoolcheck not in subpools_used):  # ignore designed annealing (same name)\n",
    "                if check_nonspecific(f_primer, fragmentcheck, Tm_rem = melt_temp_threshold, verbose=False) > 0: #use high Tm_rem\n",
    "                    anneal_locs.append(subpoolcheck)\n",
    "        if anneal_locs:\n",
    "            nonspecific.update({f_primer:[a[0] + '_block' + str(a[1]+1) for a in anneal_locs]})\n",
    "    for r_primer, subpools_used in r_primer_map.items():\n",
    "    # iterate over every barcode primer pair and match to each oligo to check for nonspecific amplification\n",
    "        anneal_locs = []\n",
    "        for subpoolcheck, fragmentcheck in wt_oligos.items():  # iterate over every WT oligo\n",
    "            if (subpoolcheck not in subpools_used):  # ignore designed annealing (same name)\n",
    "                if check_nonspecific(r_primer, fragmentcheck, Tm_rem = melt_temp_threshold, verbose=False) > 0: #use high Tm_rem\n",
    "                    anneal_locs.append(subpoolcheck)\n",
    "        if anneal_locs:\n",
    "            nonspecific.update({r_primer:[a[0] + '_block' + str(a[1]+1) for a in anneal_locs]})\n",
    "    if nonspecific:\n",
    "        print(\"Nonspecific Primers: (Manually removing primer sequence recommended)\")\n",
    "        print(nonspecific)\n",
    "    else:\n",
    "        print(\"No non-specific primers detected\")\n",
    "        \n",
    "    return nonspecific\n",
    "\n",
    "def build_kmers(sequence, \n",
    "                ksize):\n",
    "    kmers = []\n",
    "    n_kmers = len(sequence) - ksize + 1\n",
    "\n",
    "    for i in range(n_kmers):\n",
    "        kmer = sequence[i:i + ksize]\n",
    "        kmers.append(kmer)\n",
    "\n",
    "    return kmers\n",
    "\n",
    "def compute_overlaps(breakpoints, \n",
    "                     inclusion_array, \n",
    "                     gene):\n",
    "    \n",
    "    overlaps = [[gene[val:val+4].reverse_complement(), gene[val:val+4]] for val in breakpoints]\n",
    "    counter = 0\n",
    "    for val in inclusion_array:\n",
    "        if val == -1:\n",
    "            (overlaps[counter][1],overlaps[counter+1][0]) = (overlaps[counter+1][0],overlaps[counter][1])\n",
    "            counter += 1\n",
    "        elif val == 0:\n",
    "            overlaps[counter][1] = overlaps[counter+1][1]\n",
    "            del overlaps[counter+1]\n",
    "        \n",
    "    return overlaps\n",
    "\n",
    "def score_breakpoints(gene, breakpoint_pair, empirical, overhang_blacklist=None):\n",
    "    \"\"\"\n",
    "    empirical: a DataFrame indexed by 4-mers, columns are reverse complements\n",
    "    overhang_blacklist: (optional) list of bad overhangs to penalize\n",
    "    \"\"\"\n",
    "    fidelity_score = 1.0\n",
    "    for bp in breakpoint_pair:\n",
    "        kmer_seq   = gene[bp:bp+4]\n",
    "        kmer       = str(kmer_seq)\n",
    "        rev_kmer   = str(kmer_seq.reverse_complement())\n",
    "        try:\n",
    "            fidelity_score *= empirical.at[kmer, rev_kmer]\n",
    "        except KeyError:\n",
    "            fidelity_score = 0.0\n",
    "            break\n",
    "    return fidelity_score\n",
    "\n",
    "    \n",
    "def optimize_breakpoints(gene, \n",
    "                         breakpoint_pair, \n",
    "                         indices_to_shift, \n",
    "                         indices_of_array,\n",
    "                         slack, \n",
    "                         empirical=bsaI_empirical, \n",
    "                         overhang_blacklist=overhang_blacklist):\n",
    "    \n",
    "    #compute all enrichments\n",
    "    shifts = list(range(-slack,slack+1))\n",
    "    if (len(indices_to_shift) > 2) | (len(indices_to_shift) < 1):\n",
    "        print('Error -- too many or too few breakpoints!')\n",
    "        optimum_breakpoint = breakpoint_pair\n",
    "        optimum_score = 0\n",
    "    elif (len(indices_to_shift) == 1): #external pair\n",
    "        scores = [0]*len(shifts)\n",
    "        for i,shift in enumerate(shifts):\n",
    "            scores[i] = score_breakpoints(gene, breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shift] + breakpoint_pair[(indices_to_shift[0]+1):], \n",
    "                                          empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "        \n",
    "        optimum_shift = np.argmax(scores)\n",
    "        optimum_breakpoint = breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shifts[optimum_shift]] + breakpoint_pair[(indices_to_shift[0]+1):]\n",
    "        optimum_score = scores[optimum_shift]\n",
    "        optimum_length = optimum_breakpoint[indices_of_array[1]] - optimum_breakpoint[indices_of_array[0]]\n",
    "            \n",
    "            \n",
    "    else: #internal pair\n",
    "        indices_to_shift = sorted(indices_to_shift)\n",
    "        scores = np.zeros((len(shifts),len(shifts)))\n",
    "        for i,shift1 in enumerate(shifts):\n",
    "            for j,shift2 in enumerate(shifts):\n",
    "                scores[i,j] = score_breakpoints(gene, breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shift1] + \\\n",
    "                                                            breakpoint_pair[(indices_to_shift[0]+1):indices_to_shift[1]] + \\\n",
    "                                                            [breakpoint_pair[indices_to_shift[1]]+shift2] + breakpoint_pair[(indices_to_shift[1]+1):], \n",
    "                                              empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "                \n",
    "        optimum_shift = np.unravel_index(np.argmax(scores,axis=None), scores.shape)\n",
    "        optimum_breakpoint = breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shifts[optimum_shift[0]]] + \\\n",
    "                                            breakpoint_pair[(indices_to_shift[0]+1):indices_to_shift[1]] + \\\n",
    "                                            [breakpoint_pair[indices_to_shift[1]]+shifts[optimum_shift[1]]] + breakpoint_pair[(indices_to_shift[1]+1):]\n",
    "        optimum_score = scores[optimum_shift]\n",
    "        optimum_length = optimum_breakpoint[indices_of_array[1]] - optimum_breakpoint[indices_of_array[0]] + 4\n",
    "    \n",
    "    return optimum_breakpoint, optimum_score, optimum_length\n",
    "\n",
    "def optimize_gene(gene, \n",
    "                  max_tile_size,\n",
    "                  first_last_block_reduction,\n",
    "                  block_size_range=block_size_range, \n",
    "                  slack=slack, \n",
    "                  empirical=bsaI_empirical, \n",
    "                  overhang_blacklist=overhang_blacklist): \n",
    "    \n",
    "    #setup initial inputs to optimization\n",
    "    gene_size = len(gene)\n",
    "    protein_size = len(gene.translate())\n",
    "        \n",
    "    #exclude gene if it is too big\n",
    "    if protein_size > 1000:\n",
    "        print('Protein size too big!')\n",
    "        \n",
    "    #divide genes between 500 and 1000aa into two blocks\n",
    "    elif protein_size > 620:\n",
    "        print('Protein size too big! Will add two superblock (620aa+ proteins) soon.')\n",
    "        \n",
    "    else:\n",
    "        #gene is one superblock \n",
    "        #print('Protein is one superblock.')\n",
    "        block_size = block_size_range[0] + np.argmin(\n",
    "            [abs((gene_size+2*first_last_block_reduction)/(i+block_size_range[0]) - \\\n",
    "                     round((gene_size+2*first_last_block_reduction)/(i+block_size_range[0]))) \\\n",
    "                 for i in range(0, block_size_range[1]-block_size_range[0])])\n",
    "        \n",
    "        \n",
    "        # now, set initial breakpoints\n",
    "        fragment_number = int((gene_size+2*first_last_block_reduction)/block_size)\n",
    "        \n",
    "        # if any of the tiles are too big?\n",
    "        tile_lengths = [1000]\n",
    "        while max(tile_lengths) > (max_tile_size-2*slack):\n",
    "            fragment_number = fragment_number + 1\n",
    "            first_breakpoint = 0\n",
    "            last_breakpoint = gene_size-4\n",
    "            step = (last_breakpoint - first_breakpoint + 2*first_last_block_reduction)/fragment_number\n",
    "            evenly_spaced_floats = [first_breakpoint] + [step * i - first_last_block_reduction for i in range(1,fragment_number)] + [last_breakpoint]        \n",
    "            initial_breakpoints = [[first_breakpoint,int(evenly_spaced_floats[1])+slack+2,last_breakpoint]] + \\\n",
    "                                    [[0,int(evenly_spaced_floats[i])-2-slack,int(evenly_spaced_floats[i+1])+slack+2,last_breakpoint] for i in range(1,len(evenly_spaced_floats)-2)] + \\\n",
    "                                    [[0,int(evenly_spaced_floats[-2])-2-slack,last_breakpoint]]\n",
    "            tile_lengths = [int(evenly_spaced_floats[1])+slack+2-first_breakpoint+first_last_block_reduction+4] + \\\n",
    "                                    [int(evenly_spaced_floats[i+1])-int(evenly_spaced_floats[i])+2*(slack+2)+4 for i in range(1,len(evenly_spaced_floats)-2)] + \\\n",
    "                                    [last_breakpoint+4-int(evenly_spaced_floats[-2])+slack+2+first_last_block_reduction]\n",
    "            \n",
    "\n",
    "        #optimize each breakpoint\n",
    "        optimum_breakpoints = []\n",
    "        optimum_scores = []\n",
    "        optimum_lengths = []\n",
    "        oligo_array_indices = []\n",
    "        for k,breakpoint in enumerate(initial_breakpoints):\n",
    "            if len(breakpoint) == 3:\n",
    "                indices_of_array = [0, 1] if k==0 else [1, 2]\n",
    "                optimum_breakpoint, optimum_score, optimum_length = optimize_breakpoints(gene, breakpoint, [1], indices_of_array,\n",
    "                                                            slack, empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "                optimum_breakpoints.append(optimum_breakpoint)\n",
    "                optimum_scores.append(optimum_score)\n",
    "                optimum_lengths.append(optimum_length)\n",
    "                oligo_array_indices.append(indices_of_array)\n",
    "            else:\n",
    "                indices_of_array = [1, 2]\n",
    "                optimum_breakpoint, optimum_score, optimum_length = optimize_breakpoints(gene, breakpoint, [1, 2], indices_of_array,\n",
    "                                                            slack, empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "                optimum_breakpoints.append(optimum_breakpoint)\n",
    "                optimum_scores.append(optimum_score)\n",
    "                optimum_lengths.append(optimum_length)\n",
    "                oligo_array_indices.append(indices_of_array)\n",
    "    \n",
    "    optimum_overlaps = [[str(gene[t:(t+4)]) for t in s] for s in optimum_breakpoints]\n",
    "    if all([s >= 0.95 for s in optimum_scores]):\n",
    "        print('All regions are high fidelity!')\n",
    "    elif all([s >= 0.9 for s in optimum_scores]):\n",
    "        print('Some regions are medium fidelity.')\n",
    "    else:\n",
    "        print('Some regions are low fidelity. Look closer')\n",
    "        \n",
    "    return optimum_breakpoints, optimum_overlaps, optimum_scores, optimum_lengths, oligo_array_indices\n",
    "\n",
    "def generate_primer(DNA_seq,\n",
    "                     Fwd=True,\n",
    "                     extendtoCG=False,\n",
    "                     smallest_primer_size=16,\n",
    "                     largest_primer_size=30,\n",
    "                     Tm=55):\n",
    "    \n",
    "    #Setup melting temperature arrays\n",
    "    melt_temp_array = np.zeros(largest_primer_size-smallest_primer_size+1)\n",
    "    \n",
    "    if Fwd:\n",
    "        DNA_seq_touse = DNA_seq\n",
    "    else:\n",
    "        DNA_seq_touse = DNA_seq.reverse_complement()\n",
    "            \n",
    "    #Make melting temperature arrays\n",
    "    primer_length = 0\n",
    "    for i in range(smallest_primer_size,largest_primer_size+1):\n",
    "        melt_temp_array[i-smallest_primer_size] = mt.Tm_NN(DNA_seq_touse[0:i])\n",
    "        \n",
    "        #Pick F primer when Tm is first >F_Tm\n",
    "        if (melt_temp_array[i-smallest_primer_size] >= Tm) & (primer_length==0):\n",
    "            primer_length = i\n",
    "    \n",
    "    #If Tm isnt high enough after max bases, just set primer length to be max and hope it works\n",
    "    if (primer_length == 0):\n",
    "        primer_length = largest_primer_size\n",
    "        \n",
    "    if extendtoCG:\n",
    "        while ((DNA_seq_touse[primer_length-1] == 'A') | (DNA_seq_touse[primer_length-1] == 'T')) & \\\n",
    "                    (primer_length < largest_primer_size):\n",
    "            primer_length += 1\n",
    "    \n",
    "    return DNA_seq_touse[0:primer_length]\n",
    "\n",
    "def make_mutations(region_name,\n",
    "                       region,\n",
    "                       region_flanks=[Seq(''),Seq('')],\n",
    "                       nt_start=0, #zero-indexed!\n",
    "                       wt_only=False,\n",
    "                       synonymous=True,\n",
    "                       stops='TAA',\n",
    "                       all3ntdeletions=True,\n",
    "                       mutation_list=False,\n",
    "                       codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                       aa_start=0,\n",
    "                       aa_stop=None) :\n",
    "                       \n",
    "    oligo_array = {}\n",
    "    #Check that region has size divisible by three\n",
    "    if (len(region)/3 != len(region)//3) | (nt_start/3 != nt_start//3):\n",
    "        print('Region is not translatable!')\n",
    "        \n",
    "    else:\n",
    "        #add wt seq to oligo array\n",
    "        oligo_name = region_name + '_WT'\n",
    "        wt_seq = \\\n",
    "            region_flanks[0] + region + region_flanks[1]\n",
    "        oligo_array[oligo_name] = wt_seq\n",
    "        \n",
    "        if not wt_only:\n",
    "            \n",
    "            # see if a mutation list was given\n",
    "            if mutation_list == False:\n",
    "                    \n",
    "                #loop over amino acids\n",
    "                for j in range(0,len(region),3):\n",
    "\n",
    "                    #add all missense variants\n",
    "                    aa = region[j:(j+3)].translate()\n",
    "                    for aa_to in codons_ranked_by_usage.keys():\n",
    "                        if aa_to != aa:\n",
    "                            oligo_name = region_name + '_' + str(aa) + str((nt_start+j)//3+1+aa_start) + str(aa_to)\n",
    "                            seq_to_append = \\\n",
    "                                region_flanks[0] + \\\n",
    "                                region[0:j] + Seq(codons_ranked_by_usage[aa_to][0]) + \\\n",
    "                                region[(j+3):] + \\\n",
    "                                region_flanks[1]\n",
    "                            oligo_array[oligo_name] = seq_to_append\n",
    "\n",
    "                    #add synonymous variant if True and if possible, \n",
    "                    # using the most common codon that is NOT the codon in the gene\n",
    "                    if synonymous:\n",
    "                        if len(codons_ranked_by_usage[aa]) > 1:\n",
    "                            oligo_name = region_name + '_' + str(aa) + str((nt_start+j)//3+1+aa_start) + str(aa)\n",
    "                            possible_codons = codons_ranked_by_usage[aa].copy()\n",
    "                            possible_codons.remove(region[j:(j+3)])\n",
    "                            seq_to_append = \\\n",
    "                                region_flanks[0] + \\\n",
    "                                region[0:j] + Seq(possible_codons[0]) + \\\n",
    "                                region[(j+3):] + \\\n",
    "                                region_flanks[1]\n",
    "                            oligo_array[oligo_name] = seq_to_append\n",
    "\n",
    "                    #add stops if true\n",
    "                    if stops:\n",
    "                        oligo_name = region_name + '_' + str(aa) + str((nt_start+j)//3+1+aa_start) + 'X'\n",
    "                        seq_to_append = \\\n",
    "                            region_flanks[0] + \\\n",
    "                            region[0:j] + Seq(stops) + \\\n",
    "                            region[(j+3):] + \\\n",
    "                            region_flanks[1]\n",
    "                        oligo_array[oligo_name] = seq_to_append\n",
    "\n",
    "                    #add all 3nt deletions if True\n",
    "                    if all3ntdeletions:\n",
    "                        for k in range(0,3):\n",
    "                            if j+k+3 <= len(region):\n",
    "                                oligo_name = region_name + '_' + 'del' + str(nt_start+j+k+1+3*aa_start)\n",
    "                                seq_to_append = \\\n",
    "                                    region_flanks[0] + \\\n",
    "                                    region[0:(j+k)] + \\\n",
    "                                    region[(j+k+3):] + \\\n",
    "                                    region_flanks[1]\n",
    "                                oligo_array[oligo_name] = seq_to_append\n",
    "                                \n",
    "            else:\n",
    "                \n",
    "                #loop over mutation list\n",
    "                for i in range(len(mutation_list)):\n",
    "                        \n",
    "                    #iterate over every single aa change\n",
    "                    oligo_name = region_name + '_' + 'variant' + str(i+1)\n",
    "                    seq_to_append = region\n",
    "                    for k,v in enumerate(mutation_list[i]):\n",
    "                        aa_from = v[0]\n",
    "                        aa_to = v[-1]\n",
    "                        pos = int(v[1:-1])\n",
    "                        j=3*(pos-(nt_start//3+1))\n",
    "                        aa = region[j:(j+3)].translate()\n",
    "                        if aa != aa_from:\n",
    "                            print('Check mutation list!')\n",
    "                        else:\n",
    "                            seq_to_append = \\\n",
    "                                seq_to_append[0:j] + \\\n",
    "                                Seq(codons_ranked_by_usage[aa_to][0]) + \\\n",
    "                                seq_to_append[(j+3):]\n",
    "\n",
    "                    # append oligo to array\n",
    "                    seq_to_append = region_flanks[0] + \\\n",
    "                                    seq_to_append + \\\n",
    "                                    region_flanks[1]\n",
    "                    oligo_array[oligo_name] = seq_to_append\n",
    "        \n",
    "    return oligo_array\n",
    "\n",
    "\n",
    "def write_oligo_library(genes,\n",
    "                        oligo_file='./oligo_test.csv',\n",
    "                        primer_file='./primer_test.tsv',\n",
    "                        gbl_file='./gbl_test.tsv',\n",
    "                        gbl_large_file='./gbl_test_large.tsv',\n",
    "                        amp_primer_key_file='./amp_primer_key.tsv',\n",
    "                        breakpoint_file='./breakpoints.tsv',\n",
    "                        primer_set=orthogonal_primers_touse,\n",
    "                        codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                        block_size_range=block_size_range, \n",
    "                        max_oligo_size=max_oligo_size,\n",
    "                        slack=slack, \n",
    "                        empirical=bsaI_empirical, \n",
    "                        overhang_blacklist=overhang_blacklist,\n",
    "                        validated_primer_set=False,\n",
    "                        aa_start=False,\n",
    "                        aa_stop=False,\n",
    "                        wt_only=False,\n",
    "                        synonymous=True,\n",
    "                        stops='TAA',\n",
    "                        all3ntdeletions=True,\n",
    "                        mutations_to_use=False,\n",
    "                        find_pcr_primers=True,\n",
    "                        smallest_primer_size=16,\n",
    "                        largest_primer_size=30,\n",
    "                        Tm=55,\n",
    "                        extendtoCG=True,\n",
    "                        bsaI_firstoverlap='ACGA',\n",
    "                        bsaI_lastoverlap='AATG',\n",
    "                        all_blocks=True,\n",
    "                        blocks_to_include=False,\n",
    "                        tile_boundaries=False,\n",
    "                        paqcIcapF=True,\n",
    "                        paqcIcapR=True,\n",
    "                        check_all_primers=True,\n",
    "                        qc_melt_temp_threshold=32,\n",
    "                        gblock_min_size=300,\n",
    "                        gblock_large_threshold=1000,\n",
    "                        randomsequencepad=randomsequencepad):\n",
    "    \n",
    "    #Split up primer set into F and R primers, cannot do more than 82 sublibraries\n",
    "    oligo_primer_counter = 0\n",
    "    oligo_array = {}\n",
    "    amp_primers = {}\n",
    "    gblocks = {}\n",
    "    num_blocks = {}\n",
    "    amp_primer_dict = {}\n",
    "    breakpoint_dict = {}\n",
    "    \n",
    "    #Convert genes to Seq and genes to list\n",
    "    gene_names = list(genes.keys())\n",
    "    genes = [Seq(genes[gene_name].upper()) for gene_name in gene_names]\n",
    "    \n",
    "    #PaqCI and BsaI site sequences\n",
    "    paqcI_seq = Seq('CACCTGC')\n",
    "    paqcI_seqplusfour = Seq('CACCTGCCTAG')\n",
    "    bsaI_seq = Seq('GGTCTC')\n",
    "    bsaI_seqplusone = Seq('GGTCTCT')\n",
    "    pcr_capseq = Seq('GGCTAC') + bsaI_seqplusone\n",
    "    gbl_capseq_F = Seq('CCGCGTGATTACGAGTCG') + pcr_capseq\n",
    "    gbl_capseq_R = Seq('GGGTTAGCAAGTGGCAGCCT') + pcr_capseq\n",
    "    \n",
    "    # set max size of a tile\n",
    "    primer_len = len(primer_set['Forward Primer'][0])\n",
    "    max_tile_size = max_oligo_size - 2*primer_len - 2*len(bsaI_seqplusone)\n",
    "    first_last_block_reduction = len(paqcI_seqplusfour)\n",
    "    \n",
    "    # iterate over genes\n",
    "    for r,gene in enumerate(genes):\n",
    "        \n",
    "        gene_name = gene_names[r]\n",
    "        print('Processing gene ' + str(r+1) + ': ' + gene_name)\n",
    "        \n",
    "        # amino acid to start at\n",
    "        if aa_start and gene_name in aa_start:\n",
    "            aa_start_gene = aa_start[gene_name] - 1\n",
    "        else:\n",
    "            aa_start_gene = 0\n",
    "\n",
    "        if aa_stop and gene_name in aa_stop:\n",
    "            aa_stop_gene = aa_stop[gene_name]\n",
    "        else:\n",
    "            aa_stop_gene = None\n",
    "        \n",
    "        #exclude if gene size is not divisible by three\n",
    "        if len(gene)/3 != len(gene)//3:\n",
    "            print('Gene length is not divisible by 3!')\n",
    "    \n",
    "        #exclude if there is a paqcI site in the gene\n",
    "        elif any([True for kmer in build_kmers(gene, len(paqcI_seq)) if kmer==paqcI_seq]) | \\\n",
    "            any([True for kmer in build_kmers(gene.reverse_complement(), len(paqcI_seq)) if kmer==paqcI_seq]):\n",
    "            print('Gene has PaqCI site!')\n",
    "        \n",
    "        #exclude if there is a BsaI site in the gene\n",
    "        elif any([True for kmer in build_kmers(gene, len(bsaI_seq)) if kmer==bsaI_seq]) | \\\n",
    "            any([True for kmer in build_kmers(gene.reverse_complement(), len(bsaI_seq)) if kmer==bsaI_seq]):\n",
    "            print('Gene has BsaI site!')\n",
    "            \n",
    "        else:\n",
    "            print('Gene has no PaqCI or BsaI sites! Performing GoldenGate optimization...')\n",
    "            \n",
    "            #cap gene with BsaI breakpoints and possible PaqCI sites \n",
    "            if paqcIcapF & paqcIcapR:\n",
    "                gene_capped = bsaI_firstoverlap + paqcI_seqplusfour + \\\n",
    "                        gene + paqcI_seqplusfour.reverse_complement() + bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap + paqcI_seqplusfour)\n",
    "                capping_length_R = len(paqcI_seqplusfour.reverse_complement() + bsaI_lastoverlap)\n",
    "            elif paqcIcapF:\n",
    "                gene_capped = bsaI_firstoverlap + paqcI_seqplusfour + \\\n",
    "                        gene + bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap + paqcI_seqplusfour)\n",
    "                capping_length_R = len(bsaI_lastoverlap)\n",
    "            elif paqcIcapR:\n",
    "                gene_capped = bsaI_firstoverlap + \\\n",
    "                        gene + paqcI_seqplusfour.reverse_complement() + bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap)\n",
    "                capping_length_R = len(paqcI_seqplusfour.reverse_complement() + bsaI_lastoverlap)\n",
    "            else:\n",
    "                gene_capped = bsaI_firstoverlap + gene + bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap)\n",
    "                capping_length_R = len(bsaI_lastoverlap)\n",
    "            \n",
    "            #Optimize gene if tile boundaries are not given\n",
    "            if tile_boundaries is False:\n",
    "                optimum_breakpoints, optimum_overlaps, optimum_scores, optimum_lengths, oligo_array_indices = \\\n",
    "                optimize_gene(gene_capped, \n",
    "                          max_tile_size=max_tile_size,\n",
    "                          first_last_block_reduction=first_last_block_reduction,\n",
    "                          block_size_range=block_size_range, \n",
    "                          slack=slack, \n",
    "                          empirical=bsaI_empirical, \n",
    "                          overhang_blacklist=overhang_blacklist)\n",
    "                pprint.pprint({'Optimum Breakpoints': optimum_breakpoints, \n",
    "                       'Optimum Overlaps': optimum_overlaps, \n",
    "                       'Optimum Scores': optimum_scores})\n",
    "                num_blocks[gene_name] = len(optimum_breakpoints)\n",
    "            else:\n",
    "                optimum_breakpoints = tile_boundaries[gene_name]\n",
    "                if len(optimum_breakpoints[0])==3:\n",
    "                    if len(optimum_breakpoints)>1:\n",
    "                        #multiple tiles, including one at beginning of gene\n",
    "                        oligo_array_indices = [[0,1]] + [[1,2]]*(len(optimum_breakpoints)-1)\n",
    "                    else:\n",
    "                        # one tile\n",
    "                        if ((optimum_breakpoints[1]-optimum_breakpoints[0]) > (optimum_breakpoints[2]-optimum_breakpoints[1])):\n",
    "                            # at end of gene\n",
    "                            oligo_array_indices = [[1,2]]\n",
    "                        else:\n",
    "                            # at beginning of gene\n",
    "                            oligo_array_indices = [[0,1]]\n",
    "                else:\n",
    "                    # multiple tiles, starting in the middle\n",
    "                    oligo_array_indices = [[1,2]]*(len(optimum_breakpoints))\n",
    "                num_blocks[gene_name] = len(optimum_breakpoints)\n",
    "                \n",
    "            \n",
    "            #add primers for gene_F and gene_R that are repeated constantly throughout the PCRs\n",
    "            #note: should probably prevalidate these primers!\n",
    "            if find_pcr_primers:\n",
    "                F_primer = generate_primer(gene,\n",
    "                                           Fwd=True,\n",
    "                                           extendtoCG=extendtoCG,\n",
    "                                           smallest_primer_size=smallest_primer_size,\n",
    "                                           largest_primer_size=largest_primer_size,\n",
    "                                           Tm=Tm)\n",
    "                F_primer = pcr_capseq + bsaI_firstoverlap + paqcI_seqplusfour + F_primer\n",
    "                amp_primers[gene_name+'_gene'+'_ampF'] = F_primer\n",
    "                R_primer = generate_primer(gene,\n",
    "                                           Fwd=False,\n",
    "                                           extendtoCG=extendtoCG,\n",
    "                                           smallest_primer_size=smallest_primer_size,\n",
    "                                           largest_primer_size=largest_primer_size,\n",
    "                                           Tm=Tm)\n",
    "                R_primer = pcr_capseq + Seq(bsaI_lastoverlap).reverse_complement() + \\\n",
    "                            paqcI_seqplusfour + R_primer\n",
    "                amp_primers[gene_name+'_gene'+'_ampR'] = R_primer\n",
    "            \n",
    "            #make oligos, primers, gblocks for each block\n",
    "            for i,breakpoint in enumerate(optimum_breakpoints):\n",
    "                \n",
    "                #find indices of breakpoint that correspond to oligo vs need to be PCRed/gblock\n",
    "                pcr_indices = [[j,j+1] for j in range(len(breakpoint)-1)]\n",
    "                pcr_indices.remove(oligo_array_indices[i])\n",
    "                \n",
    "                #find mutagenic window of oligo\n",
    "                oligo_breaks = [breakpoint[j] for j in oligo_array_indices[i]]\n",
    "                oligo_mutagenic_window = [int(3*np.ceil(max(oligo_breaks[0]+4-capping_length_F,3)/3)), int(3*np.floor(min(oligo_breaks[1]-capping_length_F,len(gene)-1)/3))]\n",
    "                \n",
    "                #subset the right block if needed\n",
    "                if (all_blocks == True) | ((i+1) in blocks_to_include[r] if blocks_to_include != False else True): #subset on allowed blocks\n",
    "                \n",
    "                    #add pcr primers and gblocks, one segment at a time\n",
    "                    for k,pcr_index in enumerate(pcr_indices):\n",
    "                        piece_name = gene_name + '_block' + str(i+1) + '_s' + str(k+1)\n",
    "                        pcr_breaks = [breakpoint[j] for j in pcr_index]\n",
    "                                            \n",
    "                        #get pcr primers\n",
    "                        if find_pcr_primers:\n",
    "                            if pcr_breaks[0] == breakpoint[0]: #Fragment beginning at gene start \n",
    "                                R_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=False,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                R_primer = pcr_capseq + R_primer\n",
    "                                amp_primers[piece_name+'_ampR'] = R_primer\n",
    "                            elif pcr_breaks[1] == breakpoint[-1]: #Fragment ending at gene end\n",
    "                                F_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=True,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                F_primer = pcr_capseq + F_primer\n",
    "                                amp_primers[piece_name+'_ampF'] = F_primer\n",
    "                            else:\n",
    "                                F_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=True,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                F_primer = pcr_capseq + F_primer\n",
    "                                R_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=False,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                R_primer = pcr_capseq + R_primer\n",
    "                                amp_primers[piece_name+'_ampF'] = F_primer\n",
    "                                amp_primers[piece_name+'_ampR'] = R_primer\n",
    "\n",
    "                        #make gblocks\n",
    "                        gbl = gbl_capseq_F + gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)] + gbl_capseq_R.reverse_complement()\n",
    "                        gblocks[piece_name] = gbl\n",
    "                        \n",
    "                    #add oligos to oligo array, checking first for validated primers if they are given\n",
    "                    validated=False\n",
    "                    if validated_primer_set is not False:\n",
    "                        validated_combo = validated_primer_set.query('Gene == @gene_name & Block == (@i+1)')\n",
    "                        if not validated_combo.empty:\n",
    "                            validated=True\n",
    "                            name_primer_F, primer_F, name_primer_R, primer_R = \\\n",
    "                                validated_combo[['Forward Name',\n",
    "                                                  'Forward Primer',\n",
    "                                                  'Reverse Name',\n",
    "                                                  'Reverse Primer']].values[0]\n",
    "                        else:\n",
    "                            name_primer_F, primer_F, name_primer_R, primer_R = \\\n",
    "                                primer_set.iloc[oligo_primer_counter,][['Forward Name',\n",
    "                                                                      'Forward Primer',\n",
    "                                                                      'Reverse Name',\n",
    "                                                                      'Reverse Primer']]\n",
    "                            oligo_primer_counter += 1\n",
    "                    else:\n",
    "                        name_primer_F, primer_F, name_primer_R, primer_R = \\\n",
    "                                primer_set.iloc[oligo_primer_counter,][['Forward Name',\n",
    "                                                                      'Forward Primer',\n",
    "                                                                      'Reverse Name',\n",
    "                                                                      'Reverse Primer']]\n",
    "                        oligo_primer_counter += 1\n",
    "                        \n",
    "                    # if mutations are given, use those - otherwise make all mutations or wtonly\n",
    "                    if mutations_to_use == False:\n",
    "                        add_on_array = make_mutations(gene_name + '_block' + str(i+1),\n",
    "                                           gene[oligo_mutagenic_window[0]:oligo_mutagenic_window[1]],\n",
    "                                           region_flanks=[Seq(primer_F) + \\\n",
    "                                                          bsaI_seqplusone + \\\n",
    "                                                          gene_capped[oligo_breaks[0]:(oligo_mutagenic_window[0]+capping_length_F)] ,\n",
    "                                                          gene_capped[(oligo_mutagenic_window[1]+capping_length_F):(oligo_breaks[1]+4)] + \\\n",
    "                                                          bsaI_seqplusone.reverse_complement() + \\\n",
    "                                                          Seq(primer_R).reverse_complement()],\n",
    "                                           nt_start=oligo_mutagenic_window[0],\n",
    "                                           wt_only=wt_only,\n",
    "                                           synonymous=synonymous,\n",
    "                                           stops=stops,\n",
    "                                           all3ntdeletions=all3ntdeletions,\n",
    "                                           codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                                           aa_start=aa_start_gene)\n",
    "                    else:\n",
    "                        add_on_array = make_mutations(gene_name + '_block' + str(i+1),\n",
    "                                           gene[oligo_mutagenic_window[0]:oligo_mutagenic_window[1]],\n",
    "                                           region_flanks=[Seq(primer_F) + \\\n",
    "                                                          bsaI_seqplusone + \\\n",
    "                                                          gene_capped[oligo_breaks[0]:(oligo_mutagenic_window[0]+capping_length_F)] ,\n",
    "                                                          gene_capped[(oligo_mutagenic_window[1]+capping_length_F):(oligo_breaks[1]+4)] + \\\n",
    "                                                          bsaI_seqplusone.reverse_complement() + \\\n",
    "                                                          Seq(primer_R).reverse_complement()],\n",
    "                                           nt_start=oligo_mutagenic_window[0],\n",
    "                                           mutation_list=mutations_to_use[(gene_name,i+1)],\n",
    "                                           codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                                           aa_start=aa_start_gene,\n",
    "                                           aa_stop=aa_stop_gene)\n",
    "                    oligo_array.update(add_on_array)\n",
    "                    amp_primer_dict.update({(gene_name,i+1): (name_primer_F,primer_F,name_primer_R,primer_R,validated)})\n",
    "                    breakpoint_dict.update({(gene_name,i+1): oligo_mutagenic_window})\n",
    "                    \n",
    "    #Check that max oligo is less than the max oligo length\n",
    "    if sum([len(s)>max_oligo_size for s in oligo_array.values()]) == 0:\n",
    "        print('All oligos are below the maximum 250bp!')\n",
    "    else:\n",
    "        print('Some oligos are TOO BIG!')\n",
    "        \n",
    "    #Check for nonspecific amplification\n",
    "    wt_oligos = {tuple([key.split('_')[0],\n",
    "                        int((key.split('_block')[1]).split('_')[0])]\n",
    "                      ):oligo_array[key] \\\n",
    "                     for key in oligo_array.keys() if 'WT' in key}\n",
    "    nonspecific_primers = post_qc(amp_primer_dict, \n",
    "                                  wt_oligos,\n",
    "                                  primer_set, \n",
    "                                  melt_temp_threshold=qc_melt_temp_threshold,\n",
    "                                  check_all_primers=check_all_primers)\n",
    "    \n",
    "    # Check that all mutagenic windows overlap\n",
    "    breakpoint_df = pd.DataFrame.from_dict(breakpoint_dict, orient='index', columns=['Mutagenesis Start','Mutagenesis End'])\n",
    "    breakpoint_df.index = pd.MultiIndex.from_tuples(breakpoint_dict.keys())\n",
    "    breakpoint_df = breakpoint_df.reset_index().rename(columns={'level_0':'Gene',\n",
    "                                                                'level_1':'Block'})\n",
    "    if (mutations_to_use == False) and (all_blocks == True):\n",
    "        missed_counter = 0\n",
    "        for r,gene_group_breakpoints in breakpoint_df.groupby('Gene'):\n",
    "            for k,row in gene_group_breakpoints.iterrows():\n",
    "                # look at whether the current row start is later than the last row end\n",
    "                if row['Block'] > 1:\n",
    "                    if row['Mutagenesis Start'] > end:\n",
    "                        missed_counter += 1\n",
    "                        print('Mutagenic window missed at ' + str(r) + ' block ' + str(row['Block']))\n",
    "                start,end = row['Mutagenesis Start'],row['Mutagenesis End']\n",
    "        if missed_counter == 0:\n",
    "            print('All mutagenic windows overlap!')\n",
    "        else:\n",
    "            print(str(missed_counter) + ' number of times the mutagenic window does not close!')\n",
    "                \n",
    "    #Remove any oligos with additional BsaI sites or PaqCI sites\n",
    "    bad_oligos = []\n",
    "    for name,oligo in oligo_array.items():\n",
    "        #check for PaqCI sites\n",
    "        paqcI_F = sum([True for kmer in build_kmers(oligo, len(paqcI_seq)) if kmer==paqcI_seq])\n",
    "        paqcI_R = sum([True for kmer in build_kmers(oligo.reverse_complement(), len(paqcI_seq)) if kmer==paqcI_seq])\n",
    "        #check that oligo is block 1 if it contains a paqcI site in the forward orientation \n",
    "        if paqcI_F > 0:\n",
    "            if ('block1' not in name) | (paqcI_F > 1):\n",
    "                bad_oligos.append(name)\n",
    "        #check that the oligo is block final if it contains a paqcI site in the reverse orientation\n",
    "        if paqcI_R > 0:\n",
    "            if ('block'+str(num_blocks[name.split('_')[0]]) not in name) | (paqcI_R > 1):\n",
    "                bad_oligos.append(name)\n",
    "        #check for more than one BsaI site\n",
    "        bsaI_F = sum([True for kmer in build_kmers(oligo, len(bsaI_seq)) if kmer==bsaI_seq])\n",
    "        bsaI_R = sum([True for kmer in build_kmers(oligo.reverse_complement(), len(bsaI_seq)) if kmer==bsaI_seq])\n",
    "        if (bsaI_F != 1) | (bsaI_R != 1):\n",
    "            bad_oligos.append(name)\n",
    "    bad_oligos=np.unique(bad_oligos)\n",
    "    for oligo_name in bad_oligos:\n",
    "        del oligo_array[oligo_name]\n",
    "    print(str(len(bad_oligos)) + ' oligos deleted due to errant restriction sites.')\n",
    "    \n",
    "    #Remove any duplicate oligos\n",
    "    new_dict = {}\n",
    "    seen_values = set()\n",
    "    counter=0\n",
    "    for key, value in oligo_array.items():\n",
    "        if value not in seen_values:\n",
    "            new_dict[key] = value\n",
    "            seen_values.add(value)\n",
    "        else:\n",
    "            counter += 1\n",
    "    print(str(counter) + ' oligos removed due to duplication.')\n",
    "    oligo_array = new_dict\n",
    "    del new_dict\n",
    "    \n",
    "    #write oligo array to file\n",
    "    with open(oligo_file, 'w') as f:\n",
    "        for key in oligo_array.keys():\n",
    "            f.write(\"%s,%s\\n\"%(key,oligo_array[key]))\n",
    "    f.close()\n",
    "            \n",
    "    #write primers to file\n",
    "    if find_pcr_primers:\n",
    "        primer_order_sheet = []\n",
    "        for key in amp_primers.keys():\n",
    "            primer_order_sheet.append(key + '\\t' + \\\n",
    "                     str(amp_primers[key]) + \\\n",
    "                     '\\t' + '25nm' + '\\t' + 'STD\\n')\n",
    "        print(*primer_order_sheet)\n",
    "        with open(primer_file, 'w') as f:\n",
    "            for line in primer_order_sheet:\n",
    "                f.write(line)\n",
    "        f.close()\n",
    "    \n",
    "    #write amplification primer key to file\n",
    "    amp_primer_key = ['Gene' + '\\t' + 'Block' + '\\t' + \\\n",
    "                      'Forward Primer Well' + '\\t' + 'Forward Primer' + '\\t' + \\\n",
    "                      'Reverse Primer Well' + '\\t' + 'Reverse Primer' + '\\t' + 'Validated' + '\\n']\n",
    "    for key in amp_primer_dict.keys():\n",
    "        genename, geneblock = key[0], str(key[1])\n",
    "        name_primer_F, primer_F, name_primer_R, primer_R, validated = amp_primer_dict[key]\n",
    "        amp_primer_key.append(genename + '\\t' + geneblock + '\\t' + \\\n",
    "                 name_primer_F + '\\t' + primer_F + '\\t' + \\\n",
    "                 name_primer_R + '\\t' + primer_R + '\\t' + str(validated) + '\\n')\n",
    "    print(*amp_primer_key)\n",
    "    with open(amp_primer_key_file, 'w') as f:\n",
    "        for line in amp_primer_key:\n",
    "            f.write(line)\n",
    "    f.close()\n",
    "    \n",
    "    #write breakpoint dict to file\n",
    "    breakpoint_df.to_csv(breakpoint_file, sep='\\t')\n",
    "    \n",
    "    #write gblocks to file\n",
    "    gblock_order_sheet = []\n",
    "    gblock_large_order_sheet = []\n",
    "    for key in gblocks.keys():\n",
    "        # pad gblock if it is not 300bp for Twist\n",
    "        if len(gblocks[key]) < gblock_min_size:\n",
    "            gblocks[key] = Seq(randomsequencepad[0:(gblock_min_size-len(gblocks[key]))]) + gblocks[key]\n",
    "        if len(gblocks[key]) < gblock_large_threshold:\n",
    "            gblock_order_sheet.append(key + '\\t' + \\\n",
    "                     str(gblocks[key]) + '\\n')\n",
    "        else:\n",
    "            gblock_large_order_sheet.append(key + '\\t' + \\\n",
    "                     str(gblocks[key]) + '\\n')\n",
    "    print(*gblock_order_sheet)\n",
    "    print(*gblock_large_order_sheet)\n",
    "    with open(gbl_file, 'w') as f:\n",
    "        for line in gblock_order_sheet:\n",
    "            f.write(line)\n",
    "    f.close()\n",
    "    with open(gbl_large_file, 'w') as f:\n",
    "        for line in gblock_large_order_sheet:\n",
    "            f.write(line)\n",
    "    f.close()\n",
    "    \n",
    "    return oligo_array,amp_primers,gblocks,amp_primer_dict,breakpoint_df\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "8da2eda8-3393-4a96-ac93-d9556e5b544a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#redefine functions for cterm tagging (again) 20250613\n",
    "def post_qc(amp_primer_set, wt_oligos, primer_set, melt_temp_threshold = 35, check_all_primers=True):\n",
    "    print(\"Running QC for primer specificity on WT oligos\")\n",
    "    f_primer_map = {}\n",
    "    r_primer_map = {}\n",
    "    # invert the primer to subpool map\n",
    "    for k, v in amp_primer_set.items():\n",
    "        f_primer_map[v[1]] = f_primer_map.get(v[1], []) + [k]\n",
    "        r_primer_map[v[3]] = r_primer_map.get(v[3], []) + [k]\n",
    "    \n",
    "    # initialize list of nonspecific problems\n",
    "    nonspecific = {}\n",
    "    \n",
    "    # add unused primers if check_all_primers\n",
    "    if check_all_primers:\n",
    "        all_f_primers = np.unique(primer_set['Forward Primer'])\n",
    "        all_r_primers = np.unique(primer_set['Reverse Primer'])\n",
    "        for f_primer in all_f_primers:\n",
    "            if f_primer not in f_primer_map.keys():\n",
    "                f_primer_map[f_primer] = []\n",
    "        for r_primer in all_r_primers:\n",
    "            if r_primer not in r_primer_map.keys():\n",
    "                r_primer_map[r_primer] = []\n",
    "        \n",
    "    for f_primer, subpools_used in f_primer_map.items():\n",
    "    # iterate over every barcode primer pair and match to each oligo to check for nonspecific amplification\n",
    "        anneal_locs = []\n",
    "        for subpoolcheck, fragmentcheck in wt_oligos.items():  # iterate over every WT oligo\n",
    "            if (subpoolcheck not in subpools_used):  # ignore designed annealing (same name)\n",
    "                if check_nonspecific(f_primer, fragmentcheck, Tm_rem = melt_temp_threshold, verbose=False) > 0: #use high Tm_rem\n",
    "                    anneal_locs.append(subpoolcheck)\n",
    "        if anneal_locs:\n",
    "            nonspecific.update({f_primer:[a[0] + '_block' + str(a[1]+1) for a in anneal_locs]})\n",
    "    for r_primer, subpools_used in r_primer_map.items():\n",
    "    # iterate over every barcode primer pair and match to each oligo to check for nonspecific amplification\n",
    "        anneal_locs = []\n",
    "        for subpoolcheck, fragmentcheck in wt_oligos.items():  # iterate over every WT oligo\n",
    "            if (subpoolcheck not in subpools_used):  # ignore designed annealing (same name)\n",
    "                if check_nonspecific(r_primer, fragmentcheck, Tm_rem = melt_temp_threshold, verbose=False) > 0: #use high Tm_rem\n",
    "                    anneal_locs.append(subpoolcheck)\n",
    "        if anneal_locs:\n",
    "            nonspecific.update({r_primer:[a[0] + '_block' + str(a[1]+1) for a in anneal_locs]})\n",
    "    if nonspecific:\n",
    "        print(\"Nonspecific Primers: (Manually removing primer sequence recommended)\")\n",
    "        print(nonspecific)\n",
    "    else:\n",
    "        print(\"No non-specific primers detected\")\n",
    "        \n",
    "    return nonspecific\n",
    "\n",
    "def build_kmers(sequence, \n",
    "                ksize):\n",
    "    kmers = []\n",
    "    n_kmers = len(sequence) - ksize + 1\n",
    "\n",
    "    for i in range(n_kmers):\n",
    "        kmer = sequence[i:i + ksize]\n",
    "        kmers.append(kmer)\n",
    "\n",
    "    return kmers\n",
    "\n",
    "def compute_overlaps(breakpoints, \n",
    "                     inclusion_array, \n",
    "                     gene):\n",
    "    \n",
    "    overlaps = [[gene[val:val+4].reverse_complement(), gene[val:val+4]] for val in breakpoints]\n",
    "    counter = 0\n",
    "    for val in inclusion_array:\n",
    "        if val == -1:\n",
    "            (overlaps[counter][1],overlaps[counter+1][0]) = (overlaps[counter+1][0],overlaps[counter][1])\n",
    "            counter += 1\n",
    "        elif val == 0:\n",
    "            overlaps[counter][1] = overlaps[counter+1][1]\n",
    "            del overlaps[counter+1]\n",
    "        \n",
    "    return overlaps\n",
    "\n",
    "def score_breakpoints(gene, \n",
    "                      breakpoint_pair, \n",
    "                      empirical, \n",
    "                      overhang_blacklist=overhang_blacklist):\n",
    "    \n",
    "    #subset empirical matrix by the set of all overlaps\n",
    "    all_overlaps = []\n",
    "    for breakpoint in breakpoint_pair:\n",
    "        all_overlaps.append(gene[breakpoint:(breakpoint+4)])\n",
    "        all_overlaps.append(gene[breakpoint:(breakpoint+4)].reverse_complement())\n",
    "    all_overlaps = [str(o) for o in all_overlaps]\n",
    "    if (len(np.unique(all_overlaps)) == len(all_overlaps)) & (len(set(all_overlaps).intersection(set(overhang_blacklist))) == 0):\n",
    "        empirical_subset = empirical.loc[all_overlaps,all_overlaps]\n",
    "\n",
    "        #compute fidelity score\n",
    "        empirical_subset = empirical_subset/empirical_subset.sum(axis=1)\n",
    "        fidelity_score = 1\n",
    "        for breakpoint in breakpoint_pair:\n",
    "            fidelity_score = fidelity_score * \\\n",
    "                empirical_subset.loc[str(gene[breakpoint:(breakpoint+4)]),\n",
    "                                     str(gene[breakpoint:(breakpoint+4)].reverse_complement())]\n",
    "    \n",
    "    else:\n",
    "        fidelity_score = 0\n",
    "    \n",
    "    return fidelity_score\n",
    "    \n",
    "def optimize_breakpoints(gene, \n",
    "                         breakpoint_pair, \n",
    "                         indices_to_shift, \n",
    "                         indices_of_array,\n",
    "                         slack, \n",
    "                         empirical=bsaI_empirical, \n",
    "                         overhang_blacklist=overhang_blacklist):\n",
    "    \n",
    "    #compute all enrichments\n",
    "    shifts = list(range(-slack,slack+1))\n",
    "    if (len(indices_to_shift) > 2) | (len(indices_to_shift) < 1):\n",
    "        print('Error -- too many or too few breakpoints!')\n",
    "        optimum_breakpoint = breakpoint_pair\n",
    "        optimum_score = 0\n",
    "    elif (len(indices_to_shift) == 1): #external pair\n",
    "        scores = [0]*len(shifts)\n",
    "        for i,shift in enumerate(shifts):\n",
    "            scores[i] = score_breakpoints(gene, breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shift] + breakpoint_pair[(indices_to_shift[0]+1):], \n",
    "                                          empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "        \n",
    "        optimum_shift = np.argmax(scores)\n",
    "        optimum_breakpoint = breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shifts[optimum_shift]] + breakpoint_pair[(indices_to_shift[0]+1):]\n",
    "        optimum_score = scores[optimum_shift]\n",
    "        optimum_length = optimum_breakpoint[indices_of_array[1]] - optimum_breakpoint[indices_of_array[0]]\n",
    "            \n",
    "            \n",
    "    else: #internal pair\n",
    "        indices_to_shift = sorted(indices_to_shift)\n",
    "        scores = np.zeros((len(shifts),len(shifts)))\n",
    "        for i,shift1 in enumerate(shifts):\n",
    "            for j,shift2 in enumerate(shifts):\n",
    "                scores[i,j] = score_breakpoints(gene, breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shift1] + \\\n",
    "                                                            breakpoint_pair[(indices_to_shift[0]+1):indices_to_shift[1]] + \\\n",
    "                                                            [breakpoint_pair[indices_to_shift[1]]+shift2] + breakpoint_pair[(indices_to_shift[1]+1):], \n",
    "                                              empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "                \n",
    "        optimum_shift = np.unravel_index(np.argmax(scores,axis=None), scores.shape)\n",
    "        optimum_breakpoint = breakpoint_pair[0:indices_to_shift[0]] + [breakpoint_pair[indices_to_shift[0]]+shifts[optimum_shift[0]]] + \\\n",
    "                                            breakpoint_pair[(indices_to_shift[0]+1):indices_to_shift[1]] + \\\n",
    "                                            [breakpoint_pair[indices_to_shift[1]]+shifts[optimum_shift[1]]] + breakpoint_pair[(indices_to_shift[1]+1):]\n",
    "        optimum_score = scores[optimum_shift]\n",
    "        optimum_length = optimum_breakpoint[indices_of_array[1]] - optimum_breakpoint[indices_of_array[0]] + 4\n",
    "    \n",
    "    return optimum_breakpoint, optimum_score, optimum_length\n",
    "\n",
    "def optimize_gene(gene, \n",
    "                  max_tile_size,\n",
    "                  first_last_block_reduction,\n",
    "                  block_size_range=block_size_range, \n",
    "                  slack=slack, \n",
    "                  empirical=bsaI_empirical, \n",
    "                  overhang_blacklist=overhang_blacklist): \n",
    "    \n",
    "    #setup initial inputs to optimization\n",
    "    gene_size = len(gene)\n",
    "    protein_size = len(gene.translate())\n",
    "        \n",
    "    #exclude gene if it is too big\n",
    "    if protein_size > 1000:\n",
    "        print('Protein size too big!')\n",
    "        \n",
    "    #divide genes between 500 and 1000aa into two blocks\n",
    "    elif protein_size > 620:\n",
    "        print('Protein size too big! Will add two superblock (620aa+ proteins) soon.')\n",
    "        \n",
    "    else:\n",
    "        #gene is one superblock \n",
    "        #print('Protein is one superblock.')\n",
    "        block_size = block_size_range[0] + np.argmin(\n",
    "            [abs((gene_size+2*first_last_block_reduction)/(i+block_size_range[0]) - \\\n",
    "                     round((gene_size+2*first_last_block_reduction)/(i+block_size_range[0]))) \\\n",
    "                 for i in range(0, block_size_range[1]-block_size_range[0])])\n",
    "        \n",
    "        \n",
    "        # now, set initial breakpoints\n",
    "        fragment_number = int((gene_size+2*first_last_block_reduction)/block_size)\n",
    "        \n",
    "        # if any of the tiles are too big?\n",
    "        tile_lengths = [1000]\n",
    "        while max(tile_lengths) > (max_tile_size-2*slack):\n",
    "            fragment_number = fragment_number + 1\n",
    "            first_breakpoint = 0\n",
    "            last_breakpoint = gene_size-4\n",
    "            step = (last_breakpoint - first_breakpoint + 2*first_last_block_reduction)/fragment_number\n",
    "            evenly_spaced_floats = [first_breakpoint] + [step * i - first_last_block_reduction for i in range(1,fragment_number)] + [last_breakpoint]        \n",
    "            initial_breakpoints = [[first_breakpoint,int(evenly_spaced_floats[1])+slack+2,last_breakpoint]] + \\\n",
    "                                    [[0,int(evenly_spaced_floats[i])-2-slack,int(evenly_spaced_floats[i+1])+slack+2,last_breakpoint] for i in range(1,len(evenly_spaced_floats)-2)] + \\\n",
    "                                    [[0,int(evenly_spaced_floats[-2])-2-slack,last_breakpoint]]\n",
    "            tile_lengths = [int(evenly_spaced_floats[1])+slack+2-first_breakpoint+first_last_block_reduction+4] + \\\n",
    "                                    [int(evenly_spaced_floats[i+1])-int(evenly_spaced_floats[i])+2*(slack+2)+4 for i in range(1,len(evenly_spaced_floats)-2)] + \\\n",
    "                                    [last_breakpoint+4-int(evenly_spaced_floats[-2])+slack+2+first_last_block_reduction]\n",
    "            \n",
    "\n",
    "        #optimize each breakpoint\n",
    "        optimum_breakpoints = []\n",
    "        optimum_scores = []\n",
    "        optimum_lengths = []\n",
    "        oligo_array_indices = []\n",
    "        for k,breakpoint in enumerate(initial_breakpoints):\n",
    "            if len(breakpoint) == 3:\n",
    "                indices_of_array = [0, 1] if k==0 else [1, 2]\n",
    "                optimum_breakpoint, optimum_score, optimum_length = optimize_breakpoints(gene, breakpoint, [1], indices_of_array,\n",
    "                                                            slack, empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "                optimum_breakpoints.append(optimum_breakpoint)\n",
    "                optimum_scores.append(optimum_score)\n",
    "                optimum_lengths.append(optimum_length)\n",
    "                oligo_array_indices.append(indices_of_array)\n",
    "            else:\n",
    "                indices_of_array = [1, 2]\n",
    "                optimum_breakpoint, optimum_score, optimum_length = optimize_breakpoints(gene, breakpoint, [1, 2], indices_of_array,\n",
    "                                                            slack, empirical=bsaI_empirical, overhang_blacklist=overhang_blacklist)\n",
    "                optimum_breakpoints.append(optimum_breakpoint)\n",
    "                optimum_scores.append(optimum_score)\n",
    "                optimum_lengths.append(optimum_length)\n",
    "                oligo_array_indices.append(indices_of_array)\n",
    "    \n",
    "    optimum_overlaps = [[str(gene[t:(t+4)]) for t in s] for s in optimum_breakpoints]\n",
    "    if all([s >= 0.95 for s in optimum_scores]):\n",
    "        print('All regions are high fidelity!')\n",
    "    elif all([s >= 0.9 for s in optimum_scores]):\n",
    "        print('Some regions are medium fidelity.')\n",
    "    else:\n",
    "        print('Some regions are low fidelity. Look closer')\n",
    "        \n",
    "    return optimum_breakpoints, optimum_overlaps, optimum_scores, optimum_lengths, oligo_array_indices\n",
    "\n",
    "def generate_primer(DNA_seq,\n",
    "                     Fwd=True,\n",
    "                     extendtoCG=False,\n",
    "                     smallest_primer_size=16,\n",
    "                     largest_primer_size=30,\n",
    "                     Tm=55):\n",
    "    \n",
    "    #Setup melting temperature arrays\n",
    "    melt_temp_array = np.zeros(largest_primer_size-smallest_primer_size+1)\n",
    "    \n",
    "    if Fwd:\n",
    "        DNA_seq_touse = DNA_seq\n",
    "    else:\n",
    "        DNA_seq_touse = DNA_seq.reverse_complement()\n",
    "            \n",
    "    #Make melting temperature arrays\n",
    "    primer_length = 0\n",
    "    for i in range(smallest_primer_size,largest_primer_size+1):\n",
    "        melt_temp_array[i-smallest_primer_size] = mt.Tm_NN(DNA_seq_touse[0:i])\n",
    "        \n",
    "        #Pick F primer when Tm is first >F_Tm\n",
    "        if (melt_temp_array[i-smallest_primer_size] >= Tm) & (primer_length==0):\n",
    "            primer_length = i\n",
    "    \n",
    "    #If Tm isnt high enough after max bases, just set primer length to be max and hope it works\n",
    "    if (primer_length == 0):\n",
    "        primer_length = largest_primer_size\n",
    "        \n",
    "    if extendtoCG:\n",
    "        while ((DNA_seq_touse[primer_length-1] == 'A') | (DNA_seq_touse[primer_length-1] == 'T')) & \\\n",
    "                    (primer_length < largest_primer_size):\n",
    "            primer_length += 1\n",
    "    \n",
    "    return DNA_seq_touse[0:primer_length]\n",
    "\n",
    "def make_mutations(region_name,\n",
    "                       region,\n",
    "                       region_flanks=[Seq(''),Seq('')],\n",
    "                       nt_start=0, #zero-indexed!\n",
    "                       wt_only=False,\n",
    "                       synonymous=True,\n",
    "                       stops='TAA',\n",
    "                       all3ntdeletions=True,\n",
    "                       mutation_list=False,\n",
    "                       codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                       aa_start=0):\n",
    "                       \n",
    "    oligo_array = {}\n",
    "    #Check that region has size divisible by three\n",
    "    if (len(region)/3 != len(region)//3) | (nt_start/3 != nt_start//3):\n",
    "        print('Region is not translatable!')\n",
    "        \n",
    "    else:\n",
    "        #add wt seq to oligo array\n",
    "        oligo_name = region_name + '_WT'\n",
    "        wt_seq = \\\n",
    "            region_flanks[0] + region + region_flanks[1]\n",
    "        oligo_array[oligo_name] = wt_seq\n",
    "        \n",
    "        if not wt_only:\n",
    "            \n",
    "            # see if a mutation list was given\n",
    "            if mutation_list == False:\n",
    "                    \n",
    "                #loop over amino acids\n",
    "                for j in range(0,len(region),3):\n",
    "\n",
    "                    #add all missense variants\n",
    "                    aa = region[j:(j+3)].translate()\n",
    "                    for aa_to in codons_ranked_by_usage.keys():\n",
    "                        if aa_to != aa:\n",
    "                            oligo_name = region_name + '_' + str(aa) + str((nt_start+j)//3+1+aa_start) + str(aa_to)\n",
    "                            seq_to_append = \\\n",
    "                                region_flanks[0] + \\\n",
    "                                region[0:j] + Seq(codons_ranked_by_usage[aa_to][0]) + \\\n",
    "                                region[(j+3):] + \\\n",
    "                                region_flanks[1]\n",
    "                            oligo_array[oligo_name] = seq_to_append\n",
    "\n",
    "                    #add synonymous variant if True and if possible, \n",
    "                    # using the most common codon that is NOT the codon in the gene\n",
    "                    if synonymous:\n",
    "                        if len(codons_ranked_by_usage[aa]) > 1:\n",
    "                            oligo_name = region_name + '_' + str(aa) + str((nt_start+j)//3+1+aa_start) + str(aa)\n",
    "                            possible_codons = codons_ranked_by_usage[aa].copy()\n",
    "                            possible_codons.remove(region[j:(j+3)])\n",
    "                            seq_to_append = \\\n",
    "                                region_flanks[0] + \\\n",
    "                                region[0:j] + Seq(possible_codons[0]) + \\\n",
    "                                region[(j+3):] + \\\n",
    "                                region_flanks[1]\n",
    "                            oligo_array[oligo_name] = seq_to_append\n",
    "\n",
    "                    #add stops if true\n",
    "                    if stops:\n",
    "                        oligo_name = region_name + '_' + str(aa) + str((nt_start+j)//3+1+aa_start) + 'X'\n",
    "                        seq_to_append = \\\n",
    "                            region_flanks[0] + \\\n",
    "                            region[0:j] + Seq(stops) + \\\n",
    "                            region[(j+3):] + \\\n",
    "                            region_flanks[1]\n",
    "                        oligo_array[oligo_name] = seq_to_append\n",
    "\n",
    "                    #add all 3nt deletions if True\n",
    "                    if all3ntdeletions:\n",
    "                        for k in range(0,3):\n",
    "                            if j+k+3 <= len(region):\n",
    "                                oligo_name = region_name + '_' + 'del' + str(nt_start+j+k+1+3*aa_start)\n",
    "                                seq_to_append = \\\n",
    "                                    region_flanks[0] + \\\n",
    "                                    region[0:(j+k)] + \\\n",
    "                                    region[(j+k+3):] + \\\n",
    "                                    region_flanks[1]\n",
    "                                oligo_array[oligo_name] = seq_to_append\n",
    "                                \n",
    "            else:\n",
    "                \n",
    "                #loop over mutation list\n",
    "                for i in range(len(mutation_list)):\n",
    "                        \n",
    "                    #iterate over every single aa change\n",
    "                    oligo_name = region_name + '_' + 'variant' + str(i+1)\n",
    "                    seq_to_append = region\n",
    "                    for k,v in enumerate(mutation_list[i]):\n",
    "                        aa_from = v[0]\n",
    "                        aa_to = v[-1]\n",
    "                        pos = int(v[1:-1])\n",
    "                        j=3*(pos-(nt_start//3+1))\n",
    "                        aa = region[j:(j+3)].translate()\n",
    "                        if aa != aa_from:\n",
    "                            print('Check mutation list!')\n",
    "                        else:\n",
    "                            seq_to_append = \\\n",
    "                                seq_to_append[0:j] + \\\n",
    "                                Seq(codons_ranked_by_usage[aa_to][0]) + \\\n",
    "                                seq_to_append[(j+3):]\n",
    "\n",
    "                    # append oligo to array\n",
    "                    seq_to_append = region_flanks[0] + \\\n",
    "                                    seq_to_append + \\\n",
    "                                    region_flanks[1]\n",
    "                    oligo_array[oligo_name] = seq_to_append\n",
    "        \n",
    "    return oligo_array\n",
    "\n",
    "\n",
    "def write_oligo_library(genes,\n",
    "                        oligo_file='./oligo_test.csv',\n",
    "                        primer_file='./primer_test.tsv',\n",
    "                        gbl_file='./gbl_test.tsv',\n",
    "                        gbl_large_file='./gbl_test_large.tsv',\n",
    "                        amp_primer_key_file='./amp_primer_key.tsv',\n",
    "                        breakpoint_file='./breakpoints.tsv',\n",
    "                        primer_set=orthogonal_primers_touse,\n",
    "                        codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                        block_size_range=block_size_range, \n",
    "                        max_oligo_size=max_oligo_size,\n",
    "                        slack=slack, \n",
    "                        empirical=bsaI_empirical, \n",
    "                        overhang_blacklist=overhang_blacklist,\n",
    "                        validated_primer_set=False,\n",
    "                        aa_start=False,\n",
    "                        wt_only=False,\n",
    "                        synonymous=True,\n",
    "                        stops='TAA',\n",
    "                        all3ntdeletions=True,\n",
    "                        mutations_to_use=False,\n",
    "                        find_pcr_primers=True,\n",
    "                        smallest_primer_size=16,\n",
    "                        largest_primer_size=30,\n",
    "                        Tm=55,\n",
    "                        extendtoCG=True,\n",
    "                        bsaI_firstoverlap='CGTC',\n",
    "                        bsaI_lastoverlap='GCAT',\n",
    "                        all_blocks=True,\n",
    "                        blocks_to_include=False,\n",
    "                        tile_boundaries=False,\n",
    "                        paqcIcapF=True,\n",
    "                        paqcIcapR=True,\n",
    "                        check_all_primers=True,\n",
    "                        qc_melt_temp_threshold=32,\n",
    "                        gblock_min_size=300,\n",
    "                        gblock_large_threshold=1000,\n",
    "                        randomsequencepad=randomsequencepad):\n",
    "    \n",
    "    #Split up primer set into F and R primers, cannot do more than 82 sublibraries\n",
    "    oligo_primer_counter = 0\n",
    "    oligo_array = {}\n",
    "    amp_primers = {}\n",
    "    gblocks = {}\n",
    "    num_blocks = {}\n",
    "    amp_primer_dict = {}\n",
    "    breakpoint_dict = {}\n",
    "    \n",
    "    #Convert genes to Seq and genes to list\n",
    "    gene_names = list(genes.keys())\n",
    "    genes = [Seq(genes[gene_name].upper()) for gene_name in gene_names]\n",
    "    \n",
    "    #PaqCI and BsaI site sequences and overhangs for paqcI\n",
    "    paqcI_seq = Seq('CACCTGC')\n",
    "    paqcI_overhang_nterm_ctag = Seq('CCACC')\n",
    "    paqcI_overhang_cterm_ctag = Seq('ACGGG')\n",
    "    paqcI_overhang_nterm_ntag = Seq('TGGC')\n",
    "    paqcI_overhang_cterm_ntag = Seq('TAGG')\n",
    "    paqcI_seqplusfour = Seq('CACCTGCCTAG')\n",
    "    bsaI_seq = Seq('GGTCTC')\n",
    "    bsaI_seqplusone = Seq('GGTCTCT')\n",
    "    pcr_capseq = Seq('GGCTAC') + bsaI_seqplusone\n",
    "    gbl_capseq_F = Seq('CCGCGTGATTACGAGTCG') + pcr_capseq\n",
    "    gbl_capseq_R = Seq('GGGTTAGCAAGTGGCAGCCT') + pcr_capseq\n",
    "    \n",
    "    # set max size of a tile\n",
    "    primer_len = len(primer_set['Forward Primer'][0])\n",
    "    max_tile_size = max_oligo_size - 2*primer_len - 2*len(bsaI_seqplusone)\n",
    "    first_last_block_reduction = len(paqcI_seqplusfour)\n",
    "    \n",
    "    # iterate over genes\n",
    "    for r,gene in enumerate(genes):\n",
    "        \n",
    "        gene_name = gene_names[r]\n",
    "        print('Processing gene ' + str(r+1) + ': ' + gene_name)\n",
    "        \n",
    "        # amino acid to start at\n",
    "        if aa_start != False:\n",
    "            if gene_name in aa_start.keys():\n",
    "                aa_start_gene=aa_start[gene_name]-1\n",
    "            else:\n",
    "                aa_start_gene=0\n",
    "        else:\n",
    "            aa_start_gene=0\n",
    "        \n",
    "        #exclude if gene size is not divisible by three\n",
    "        if len(gene)/3 != len(gene)//3:\n",
    "            print('Gene length is not divisible by 3!')\n",
    "    \n",
    "        #exclude if there is a paqcI site in the gene\n",
    "        elif any([True for kmer in build_kmers(gene, len(paqcI_seq)) if kmer==paqcI_seq]) | \\\n",
    "            any([True for kmer in build_kmers(gene.reverse_complement(), len(paqcI_seq)) if kmer==paqcI_seq]):\n",
    "            print('Gene has paqcI site!')\n",
    "        \n",
    "        #exclude if there is a BsaI site in the gene\n",
    "        elif any([True for kmer in build_kmers(gene, len(bsaI_seq)) if kmer==bsaI_seq]) | \\\n",
    "            any([True for kmer in build_kmers(gene.reverse_complement(), len(bsaI_seq)) if kmer==bsaI_seq]):\n",
    "            print('Gene has BsaI site!')\n",
    "            \n",
    "        else:\n",
    "            print('Gene has no paqcI or BsaI sites! Performing GoldenGate optimization...')\n",
    "            \n",
    "            #cap gene with BsaI breakpoints and possible paqcI sites \n",
    "            if paqcIcapF & paqcIcapR:\n",
    "                gene_capped = bsaI_firstoverlap + paqcI_seqplusfour +  paqcI_overhang_nterm_ctag  +\\\n",
    "                        gene + paqcI_overhang_cterm_ctag + paqcI_seqplusfour.reverse_complement() + \\\n",
    "                        bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap + paqcI_seqplusfour + paqcI_overhang_nterm_ctag)\n",
    "                capping_length_R = len(paqcI_overhang_cterm_ctag + paqcI_seqplusfour.reverse_complement() + \\\n",
    "                                   bsaI_lastoverlap)\n",
    "            elif paqcIcapF:\n",
    "                gene_capped = bsaI_firstoverlap + paqcI_seqplusfour + paqcI_overhang_nterm_ctag + \\\n",
    "                        gene + bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap + paqcI_seqplusfour + paqcI_overhang_nterm_ctag)\n",
    "                capping_length_R = len(bsaI_lastoverlap)\n",
    "            elif paqcIcapR:\n",
    "                gene_capped = bsaI_firstoverlap + \\\n",
    "                        gene + paqcI_overhang_cterm_ctag + paqcI_seqplusfour.reverse_complement() + bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap)\n",
    "                capping_length_R = len(paqcI_overhang_cterm_ctag + paqcI_seqplusfour.reverse_complement() + \\\n",
    "                                   bsaI_lastoverlap)\n",
    "            else:\n",
    "                gene_capped = bsaI_firstoverlap + gene + bsaI_lastoverlap\n",
    "                capping_length_F = len(bsaI_firstoverlap)\n",
    "                capping_length_R = len(bsaI_lastoverlap)\n",
    "            \n",
    "            #Optimize gene if tile boundaries are not given\n",
    "            if tile_boundaries is False:\n",
    "                optimum_breakpoints, optimum_overlaps, optimum_scores, optimum_lengths, oligo_array_indices = \\\n",
    "                optimize_gene(gene_capped, \n",
    "                          max_tile_size=max_tile_size,\n",
    "                          first_last_block_reduction=first_last_block_reduction,\n",
    "                          block_size_range=block_size_range, \n",
    "                          slack=slack, \n",
    "                          empirical=bsaI_empirical, \n",
    "                          overhang_blacklist=overhang_blacklist)\n",
    "                pprint.pprint({'Optimum Breakpoints': optimum_breakpoints, \n",
    "                       'Optimum Overlaps': optimum_overlaps, \n",
    "                       'Optimum Scores': optimum_scores})\n",
    "                num_blocks[gene_name] = len(optimum_breakpoints)\n",
    "            else:\n",
    "                optimum_breakpoints = tile_boundaries[gene_name]\n",
    "                if len(optimum_breakpoints[0])==3:\n",
    "                    if len(optimum_breakpoints)>1:\n",
    "                        #multiple tiles, including one at beginning of gene\n",
    "                        oligo_array_indices = [[0,1]] + [[1,2]]*(len(optimum_breakpoints)-1)\n",
    "                    else:\n",
    "                        # one tile\n",
    "                        if ((optimum_breakpoints[1]-optimum_breakpoints[0]) > (optimum_breakpoints[2]-optimum_breakpoints[1])):\n",
    "                            # at end of gene\n",
    "                            oligo_array_indices = [[1,2]]\n",
    "                        else:\n",
    "                            # at beginning of gene\n",
    "                            oligo_array_indices = [[0,1]]\n",
    "                else:\n",
    "                    # multiple tiles, starting in the middle\n",
    "                    oligo_array_indices = [[1,2]]*(len(optimum_breakpoints))\n",
    "                num_blocks[gene_name] = len(optimum_breakpoints)\n",
    "                \n",
    "            \n",
    "            #add primers for gene_F and gene_R that are repeated constantly throughout the PCRs\n",
    "            #note: should probably prevalidate these primers!\n",
    "            if find_pcr_primers:\n",
    "                F_primer = generate_primer(gene,\n",
    "                                           Fwd=True,\n",
    "                                           extendtoCG=extendtoCG,\n",
    "                                           smallest_primer_size=smallest_primer_size,\n",
    "                                           largest_primer_size=largest_primer_size,\n",
    "                                           Tm=Tm)\n",
    "                F_primer = pcr_capseq + bsaI_firstoverlap + paqcI_seqplusfour + F_primer\n",
    "                amp_primers[gene_name+'_gene'+'_ampF'] = F_primer\n",
    "                R_primer = generate_primer(gene,\n",
    "                                           Fwd=False,\n",
    "                                           extendtoCG=extendtoCG,\n",
    "                                           smallest_primer_size=smallest_primer_size,\n",
    "                                           largest_primer_size=largest_primer_size,\n",
    "                                           Tm=Tm)\n",
    "                R_primer = pcr_capseq + Seq(bsaI_lastoverlap).reverse_complement() + \\\n",
    "                            paqcI_seqplusfour + R_primer\n",
    "                amp_primers[gene_name+'_gene'+'_ampR'] = R_primer\n",
    "            \n",
    "            #make oligos, primers, gblocks for each block\n",
    "            for i,breakpoint in enumerate(optimum_breakpoints):\n",
    "                \n",
    "                #find indices of breakpoint that correspond to oligo vs need to be PCRed/gblock\n",
    "                pcr_indices = [[j,j+1] for j in range(len(breakpoint)-1)]\n",
    "                pcr_indices.remove(oligo_array_indices[i])\n",
    "                \n",
    "                #find mutagenic window of oligo\n",
    "                oligo_breaks = [breakpoint[j] for j in oligo_array_indices[i]]\n",
    "                oligo_mutagenic_window = [int(3*np.ceil(max(oligo_breaks[0]+4-capping_length_F,3)/3)), int(3*np.floor(min(oligo_breaks[1]-capping_length_F,len(gene)-1)/3))]\n",
    "                \n",
    "                #subset the right block if needed\n",
    "                if (all_blocks == True) | ((i+1) in blocks_to_include[r] if blocks_to_include != False else True): #subset on allowed blocks\n",
    "                \n",
    "                    #add pcr primers and gblocks, one segment at a time\n",
    "                    for k,pcr_index in enumerate(pcr_indices):\n",
    "                        piece_name = gene_name + '_block' + str(i+1) + '_s' + str(k+1)\n",
    "                        pcr_breaks = [breakpoint[j] for j in pcr_index]\n",
    "                                            \n",
    "                        #get pcr primers\n",
    "                        if find_pcr_primers:\n",
    "                            if pcr_breaks[0] == breakpoint[0]: #Fragment beginning at gene start \n",
    "                                R_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=False,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                R_primer = pcr_capseq + R_primer\n",
    "                                amp_primers[piece_name+'_ampR'] = R_primer\n",
    "                            elif pcr_breaks[1] == breakpoint[-1]: #Fragment ending at gene end\n",
    "                                F_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=True,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                F_primer = pcr_capseq + F_primer\n",
    "                                amp_primers[piece_name+'_ampF'] = F_primer\n",
    "                            else:\n",
    "                                F_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=True,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                F_primer = pcr_capseq + F_primer\n",
    "                                R_primer = generate_primer(gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)],\n",
    "                                                           Fwd=False,\n",
    "                                                           extendtoCG=extendtoCG,\n",
    "                                                           smallest_primer_size=smallest_primer_size,\n",
    "                                                           largest_primer_size=largest_primer_size,\n",
    "                                                           Tm=Tm)\n",
    "                                R_primer = pcr_capseq + R_primer\n",
    "                                amp_primers[piece_name+'_ampF'] = F_primer\n",
    "                                amp_primers[piece_name+'_ampR'] = R_primer\n",
    "\n",
    "                        #make gblocks\n",
    "                        gbl = gbl_capseq_F + gene_capped[pcr_breaks[0]:(pcr_breaks[1]+4)] + gbl_capseq_R.reverse_complement()\n",
    "                        gblocks[piece_name] = gbl\n",
    "                        \n",
    "                    #add oligos to oligo array, checking first for validated primers if they are given\n",
    "                    validated=False\n",
    "                    if validated_primer_set is not False:\n",
    "                        validated_combo = validated_primer_set.query('Gene == @gene_name & Block == (@i+1)')\n",
    "                        if not validated_combo.empty:\n",
    "                            validated=True\n",
    "                            name_primer_F, primer_F, name_primer_R, primer_R = \\\n",
    "                                validated_combo[['Forward Name',\n",
    "                                                  'Forward Primer',\n",
    "                                                  'Reverse Name',\n",
    "                                                  'Reverse Primer']].values[0]\n",
    "                        else:\n",
    "                            name_primer_F, primer_F, name_primer_R, primer_R = \\\n",
    "                                primer_set.iloc[oligo_primer_counter,][['Forward Name',\n",
    "                                                                      'Forward Primer',\n",
    "                                                                      'Reverse Name',\n",
    "                                                                      'Reverse Primer']]\n",
    "                            oligo_primer_counter += 1\n",
    "                    else:\n",
    "                        name_primer_F, primer_F, name_primer_R, primer_R = \\\n",
    "                                primer_set.iloc[oligo_primer_counter,][['Forward Name',\n",
    "                                                                      'Forward Primer',\n",
    "                                                                      'Reverse Name',\n",
    "                                                                      'Reverse Primer']]\n",
    "                        oligo_primer_counter += 1\n",
    "                        \n",
    "                    # if mutations are given, use those - otherwise make all mutations or wtonly\n",
    "                    if mutations_to_use == False:\n",
    "                        add_on_array = make_mutations(gene_name + '_block' + str(i+1),\n",
    "                                           gene[oligo_mutagenic_window[0]:oligo_mutagenic_window[1]],\n",
    "                                           region_flanks=[Seq(primer_F) + \\\n",
    "                                                          bsaI_seqplusone + \\\n",
    "                                                          gene_capped[oligo_breaks[0]:(oligo_mutagenic_window[0]+capping_length_F)] ,\n",
    "                                                          gene_capped[(oligo_mutagenic_window[1]+capping_length_F):(oligo_breaks[1]+4)] + \\\n",
    "                                                          bsaI_seqplusone.reverse_complement() + \\\n",
    "                                                          Seq(primer_R).reverse_complement()],\n",
    "                                           nt_start=oligo_mutagenic_window[0],\n",
    "                                           wt_only=wt_only,\n",
    "                                           synonymous=synonymous,\n",
    "                                           stops=stops,\n",
    "                                           all3ntdeletions=all3ntdeletions,\n",
    "                                           codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                                           aa_start=aa_start_gene)\n",
    "                    else:\n",
    "                        add_on_array = make_mutations(gene_name + '_block' + str(i+1),\n",
    "                                           gene[oligo_mutagenic_window[0]:oligo_mutagenic_window[1]],\n",
    "                                           region_flanks=[Seq(primer_F) + \\\n",
    "                                                          bsaI_seqplusone + \\\n",
    "                                                          gene_capped[oligo_breaks[0]:(oligo_mutagenic_window[0]+capping_length_F)] ,\n",
    "                                                          gene_capped[(oligo_mutagenic_window[1]+capping_length_F):(oligo_breaks[1]+4)] + \\\n",
    "                                                          bsaI_seqplusone.reverse_complement() + \\\n",
    "                                                          Seq(primer_R).reverse_complement()],\n",
    "                                           nt_start=oligo_mutagenic_window[0],\n",
    "                                           mutation_list=mutations_to_use[(gene_name,i+1)],\n",
    "                                           codons_ranked_by_usage=codons_ranked_by_usage,\n",
    "                                           aa_start=aa_start_gene)\n",
    "                    oligo_array.update(add_on_array)\n",
    "                    amp_primer_dict.update({(gene_name,i+1): (name_primer_F,primer_F,name_primer_R,primer_R,validated)})\n",
    "                    breakpoint_dict.update({(gene_name,i+1): oligo_mutagenic_window})\n",
    "                    \n",
    "    #Check that max oligo is less than the max oligo length\n",
    "    if sum([len(s)>max_oligo_size for s in oligo_array.values()]) == 0:\n",
    "        print('All oligos are below the maximum 250bp!')\n",
    "    else:\n",
    "        print('Some oligos are TOO BIG!')\n",
    "        \n",
    "    #Check for nonspecific amplification\n",
    "    wt_oligos = {tuple([key.split('_')[0],\n",
    "                        int((key.split('_block')[1]).split('_')[0])]\n",
    "                      ):oligo_array[key] \\\n",
    "                     for key in oligo_array.keys() if 'WT' in key}\n",
    "    nonspecific_primers = post_qc(amp_primer_dict, \n",
    "                                  wt_oligos,\n",
    "                                  primer_set, \n",
    "                                  melt_temp_threshold=qc_melt_temp_threshold,\n",
    "                                  check_all_primers=check_all_primers)\n",
    "    \n",
    "    # Check that all mutagenic windows overlap\n",
    "    breakpoint_df = pd.DataFrame.from_dict(breakpoint_dict, orient='index', columns=['Mutagenesis Start','Mutagenesis End'])\n",
    "    breakpoint_df.index = pd.MultiIndex.from_tuples(breakpoint_dict.keys())\n",
    "    breakpoint_df = breakpoint_df.reset_index().rename(columns={'level_0':'Gene',\n",
    "                                                                'level_1':'Block'})\n",
    "    if (mutations_to_use == False) and (all_blocks == True):\n",
    "        missed_counter = 0\n",
    "        for r,gene_group_breakpoints in breakpoint_df.groupby('Gene'):\n",
    "            for k,row in gene_group_breakpoints.iterrows():\n",
    "                # look at whether the current row start is later than the last row end\n",
    "                if row['Block'] > 1:\n",
    "                    if row['Mutagenesis Start'] > end:\n",
    "                        missed_counter += 1\n",
    "                        print('Mutagenic window missed at ' + str(r) + ' block ' + str(row['Block']))\n",
    "                start,end = row['Mutagenesis Start'],row['Mutagenesis End']\n",
    "        if missed_counter == 0:\n",
    "            print('All mutagenic windows overlap!')\n",
    "        else:\n",
    "            print(str(missed_counter) + ' number of times the mutagenic window does not close!')\n",
    "                \n",
    "    #Remove any oligos with additional BsaI sites or paqcI sites\n",
    "    bad_oligos = []\n",
    "    for name,oligo in oligo_array.items():\n",
    "        #check for paqcI sites\n",
    "        paqcI_F = sum([True for kmer in build_kmers(oligo, len(paqcI_seq)) if kmer==paqcI_seq])\n",
    "        paqcI_R = sum([True for kmer in build_kmers(oligo.reverse_complement(), len(paqcI_seq)) if kmer==paqcI_seq])\n",
    "        #check that oligo is block 1 if it contains a paqcI site in the forward orientation \n",
    "        if paqcI_F > 0:\n",
    "            if ('block1' not in name) | (paqcI_F > 1):\n",
    "                bad_oligos.append(name)\n",
    "        #check that the oligo is block final if it contains a paqcI site in the reverse orientation\n",
    "        if paqcI_R > 0:\n",
    "            if ('block'+str(num_blocks[name.split('_')[0]]) not in name) | (paqcI_R > 1):\n",
    "                bad_oligos.append(name)\n",
    "        #check for more than one BsaI site\n",
    "        bsaI_F = sum([True for kmer in build_kmers(oligo, len(bsaI_seq)) if kmer==bsaI_seq])\n",
    "        bsaI_R = sum([True for kmer in build_kmers(oligo.reverse_complement(), len(bsaI_seq)) if kmer==bsaI_seq])\n",
    "        if (bsaI_F != 1) | (bsaI_R != 1):\n",
    "            bad_oligos.append(name)\n",
    "    bad_oligos=np.unique(bad_oligos)\n",
    "    for oligo_name in bad_oligos:\n",
    "        del oligo_array[oligo_name]\n",
    "    print(str(len(bad_oligos)) + ' oligos deleted due to errant restriction sites.')\n",
    "    \n",
    "    #Remove any duplicate oligos\n",
    "    new_dict = {}\n",
    "    seen_values = set()\n",
    "    counter=0\n",
    "    for key, value in oligo_array.items():\n",
    "        if value not in seen_values:\n",
    "            new_dict[key] = value\n",
    "            seen_values.add(value)\n",
    "        else:\n",
    "            counter += 1\n",
    "    print(str(counter) + ' oligos removed due to duplication.')\n",
    "    oligo_array = new_dict\n",
    "    del new_dict\n",
    "    \n",
    "    #write oligo array to file\n",
    "    with open(oligo_file, 'w') as f:\n",
    "        for key in oligo_array.keys():\n",
    "            f.write(\"%s,%s\\n\"%(key,oligo_array[key]))\n",
    "    f.close()\n",
    "            \n",
    "    #write primers to file\n",
    "    if find_pcr_primers:\n",
    "        primer_order_sheet = []\n",
    "        for key in amp_primers.keys():\n",
    "            primer_order_sheet.append(key + '\\t' + \\\n",
    "                     str(amp_primers[key]) + \\\n",
    "                     '\\t' + '25nm' + '\\t' + 'STD\\n')\n",
    "        print(*primer_order_sheet)\n",
    "        with open(primer_file, 'w') as f:\n",
    "            for line in primer_order_sheet:\n",
    "                f.write(line)\n",
    "        f.close()\n",
    "    \n",
    "    #write amplification primer key to file\n",
    "    amp_primer_key = ['Gene' + '\\t' + 'Block' + '\\t' + \\\n",
    "                      'Forward Primer Well' + '\\t' + 'Forward Primer' + '\\t' + \\\n",
    "                      'Reverse Primer Well' + '\\t' + 'Reverse Primer' + '\\t' + 'Validated' + '\\n']\n",
    "    for key in amp_primer_dict.keys():\n",
    "        genename, geneblock = key[0], str(key[1])\n",
    "        name_primer_F, primer_F, name_primer_R, primer_R, validated = amp_primer_dict[key]\n",
    "        amp_primer_key.append(genename + '\\t' + geneblock + '\\t' + \\\n",
    "                 name_primer_F + '\\t' + primer_F + '\\t' + \\\n",
    "                 name_primer_R + '\\t' + primer_R + '\\t' + str(validated) + '\\n')\n",
    "    print(*amp_primer_key)\n",
    "    with open(amp_primer_key_file, 'w') as f:\n",
    "        for line in amp_primer_key:\n",
    "            f.write(line)\n",
    "    f.close()\n",
    "    \n",
    "    #write breakpoint dict to file\n",
    "    breakpoint_df.to_csv(breakpoint_file, sep='\\t')\n",
    "    \n",
    "    #write gblocks to file\n",
    "    gblock_order_sheet = []\n",
    "    gblock_large_order_sheet = []\n",
    "    for key in gblocks.keys():\n",
    "        # pad gblock if it is not 300bp for Twist\n",
    "        if len(gblocks[key]) < gblock_min_size:\n",
    "            gblocks[key] = Seq(randomsequencepad[0:(gblock_min_size-len(gblocks[key]))]) + gblocks[key]\n",
    "        if len(gblocks[key]) < gblock_large_threshold:\n",
    "            gblock_order_sheet.append(key + '\\t' + \\\n",
    "                     str(gblocks[key]) + '\\n')\n",
    "        else:\n",
    "            gblock_large_order_sheet.append(key + '\\t' + \\\n",
    "                     str(gblocks[key]) + '\\n')\n",
    "    print(*gblock_order_sheet)\n",
    "    print(*gblock_large_order_sheet)\n",
    "    with open(gbl_file, 'w') as f:\n",
    "        for line in gblock_order_sheet:\n",
    "            f.write(line)\n",
    "    f.close()\n",
    "    with open(gbl_large_file, 'w') as f:\n",
    "        for line in gblock_large_order_sheet:\n",
    "            f.write(line)\n",
    "    f.close()\n",
    "    \n",
    "    return oligo_array,amp_primers,gblocks,amp_primer_dict,breakpoint_df\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "40d436bc-73f8-44c7-a393-eae591c2fca6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing gene 1: FRED1\n",
      "Gene has no paqcI or BsaI sites! Performing GoldenGate optimization...\n",
      "All regions are high fidelity!\n",
      "{'Optimum Breakpoints': [[0, 152, 1693],\n",
      "                         [0, 134, 308, 1693],\n",
      "                         [0, 290, 468, 1693],\n",
      "                         [0, 444, 624, 1693],\n",
      "                         [0, 609, 778, 1693],\n",
      "                         [0, 764, 932, 1693],\n",
      "                         [0, 917, 1089, 1693],\n",
      "                         [0, 1076, 1238, 1693],\n",
      "                         [0, 1225, 1401, 1693],\n",
      "                         [0, 1386, 1553, 1693],\n",
      "                         [0, 1543, 1693]],\n",
      " 'Optimum Overlaps': [['ACGA', 'TGGA', 'TTTC'],\n",
      "                      ['ACGA', 'ATCT', 'CTGA', 'TTTC'],\n",
      "                      ['ACGA', 'TAGG', 'TGGA', 'TTTC'],\n",
      "                      ['ACGA', 'AGAA', 'TCTG', 'TTTC'],\n",
      "                      ['ACGA', 'ATGT', 'AATG', 'TTTC'],\n",
      "                      ['ACGA', 'AAAT', 'TGGA', 'TTTC'],\n",
      "                      ['ACGA', 'ATCG', 'AAAT', 'TTTC'],\n",
      "                      ['ACGA', 'AACT', 'CTGA', 'TTTC'],\n",
      "                      ['ACGA', 'CATT', 'TCAG', 'TTTC'],\n",
      "                      ['ACGA', 'GACT', 'AAGG', 'TTTC'],\n",
      "                      ['ACGA', 'AGAT', 'TTTC']],\n",
      " 'Optimum Scores': [np.float64(0.9785946393033325),\n",
      "                    np.float64(0.9608187290412055),\n",
      "                    np.float64(0.9575330839430957),\n",
      "                    np.float64(0.9612362130311706),\n",
      "                    np.float64(0.9589565359642697),\n",
      "                    np.float64(0.9597553070083468),\n",
      "                    np.float64(0.9584058487071176),\n",
      "                    np.float64(0.9590421766916686),\n",
      "                    np.float64(0.9604421822688576),\n",
      "                    np.float64(0.9590137925936293),\n",
      "                    np.float64(0.9786293836240394)]}\n",
      "Processing gene 2: VWF1\n",
      "Gene has no paqcI or BsaI sites! Performing GoldenGate optimization...\n",
      "All regions are high fidelity!\n",
      "{'Optimum Breakpoints': [[0, 152, 1720],\n",
      "                         [0, 135, 313, 1720],\n",
      "                         [0, 299, 469, 1720],\n",
      "                         [0, 452, 627, 1720],\n",
      "                         [0, 610, 791, 1720],\n",
      "                         [0, 770, 945, 1720],\n",
      "                         [0, 937, 1102, 1720],\n",
      "                         [0, 1092, 1266, 1720],\n",
      "                         [0, 1244, 1416, 1720],\n",
      "                         [0, 1407, 1577, 1720],\n",
      "                         [0, 1562, 1720]],\n",
      " 'Optimum Overlaps': [['ACGA', 'CCTT', 'TTTC'],\n",
      "                      ['ACGA', 'AAGT', 'AATG', 'TTTC'],\n",
      "                      ['ACGA', 'TCCA', 'TCAG', 'TTTC'],\n",
      "                      ['ACGA', 'ACTT', 'ATCT', 'TTTC'],\n",
      "                      ['ACGA', 'CAGT', 'CTGG', 'TTTC'],\n",
      "                      ['ACGA', 'AGAA', 'CAGG', 'TTTC'],\n",
      "                      ['ACGA', 'CCTT', 'TCCC', 'TTTC'],\n",
      "                      ['ACGA', 'TCCC', 'GGAT', 'TTTC'],\n",
      "                      ['ACGA', 'TCTG', 'GGAT', 'TTTC'],\n",
      "                      ['ACGA', 'AGTT', 'CCTG', 'TTTC'],\n",
      "                      ['ACGA', 'TCTA', 'TTTC']],\n",
      " 'Optimum Scores': [np.float64(0.978500402763188),\n",
      "                    np.float64(0.9598580659594683),\n",
      "                    np.float64(0.9607711185264781),\n",
      "                    np.float64(0.960234383725724),\n",
      "                    np.float64(0.9568128787901111),\n",
      "                    np.float64(0.9594905367555729),\n",
      "                    np.float64(0.9579038819138755),\n",
      "                    np.float64(0.9567861891756589),\n",
      "                    np.float64(0.9585704355531822),\n",
      "                    np.float64(0.9563520163082417),\n",
      "                    np.float64(0.9781382098394528)]}\n",
      "All oligos are below the maximum 250bp!\n",
      "Running QC for primer specificity on WT oligos\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/lib/python3.13/site-packages/Bio/Seq.py:2879: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "/opt/miniconda3/lib/python3.13/site-packages/Bio/Seq.py:2879: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n",
      "/opt/miniconda3/lib/python3.13/site-packages/primer3/bindings.py:305: UserWarning: Function deprecated please use \"calc_heterodimer\" instead\n",
      "  return THERMO_ANALYSIS.calcHeterodimer(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No non-specific primers detected\n",
      "All mutagenic windows overlap!\n",
      "229 oligos deleted due to errant restriction sites.\n",
      "911 oligos removed due to duplication.\n",
      "FRED1_gene_ampF\tGGCTACGGTCTCTACGACACCTGCCTAGATGCAAATAGAGCTCTCTACCTGCTTC\t25nm\tSTD\n",
      " FRED1_gene_ampR\tGGCTACGGTCTCTGAAACACCTGCCTAGAGAGTAATAGCGGGTCAGGCAC\t25nm\tSTD\n",
      " FRED1_block1_s1_ampF\tGGCTACGGTCTCTTGGACGCAAGATTTCCTCCTAGAG\t25nm\tSTD\n",
      " FRED1_block2_s1_ampR\tGGCTACGGTCTCTAGATCACTTTGCATATAGTCCCATGACAG\t25nm\tSTD\n",
      " FRED1_block2_s2_ampF\tGGCTACGGTCTCTCTGAGGTTTATGATACAGTGGTCATTACAC\t25nm\tSTD\n",
      " FRED1_block3_s1_ampR\tGGCTACGGTCTCTCCTAGCAGACCCATCCAGGG\t25nm\tSTD\n",
      " FRED1_block3_s2_ampF\tGGCTACGGTCTCTTGGAAGCCATACATATGTCTGGCAG\t25nm\tSTD\n",
      " FRED1_block4_s1_ampR\tGGCTACGGTCTCTTTCTTTCTCCCTTTGACTGGTCTGATC\t25nm\tSTD\n",
      " FRED1_block4_s2_ampF\tGGCTACGGTCTCTTCTGGCCAAGGAAAAGACACAGAC\t25nm\tSTD\n",
      " FRED1_block5_s1_ampR\tGGCTACGGTCTCTACATACTAGTAGGGCTCCAATGAGG\t25nm\tSTD\n",
      " FRED1_block5_s2_ampF\tGGCTACGGTCTCTAATGGTTATGTAAACAGATCTCTGCCAG\t25nm\tSTD\n",
      " FRED1_block6_s1_ampR\tGGCTACGGTCTCTATTTTAGGCCAGGCCCGAGC\t25nm\tSTD\n",
      " FRED1_block6_s2_ampF\tGGCTACGGTCTCTTGGAAATCTCGCCAATAACTTTCCTTAC\t25nm\tSTD\n",
      " FRED1_block7_s1_ampR\tGGCTACGGTCTCTCGATGGTTCCTCACAAGAAATGTGTG\t25nm\tSTD\n",
      " FRED1_block7_s2_ampF\tGGCTACGGTCTCTAAATAATGAAGAAGCGGAAGACTATGATGA\t25nm\tSTD\n",
      " FRED1_block8_s1_ampR\tGGCTACGGTCTCTAGTTGGGGTTCCTCTGGACAG\t25nm\tSTD\n",
      " FRED1_block8_s2_ampF\tGGCTACGGTCTCTCTGAAGAGGAGGACTGGGACTATG\t25nm\tSTD\n",
      " FRED1_block9_s1_ampR\tGGCTACGGTCTCTAATGTACCCAAGTTTTAGGATGCTTCTTG\t25nm\tSTD\n",
      " FRED1_block9_s2_ampF\tGGCTACGGTCTCTTCAGCATGAATCAGGAATCTTGGGAC\t25nm\tSTD\n",
      " FRED1_block10_s1_ampR\tGGCTACGGTCTCTAGTCTTAAAGGTTTCATCTGTGTATGCC\t25nm\tSTD\n",
      " FRED1_block10_s2_ampF\tGGCTACGGTCTCTAAGGTGTAAAACATTTGAAGGATTTTCCAA\t25nm\tSTD\n",
      " FRED1_block11_s1_ampR\tGGCTACGGTCTCTATCTCCTTGAATACAAAGGACGGACATC\t25nm\tSTD\n",
      " VWF1_gene_ampF\tGGCTACGGTCTCTACGACACCTGCCTAGATGATTCCTGCCAGATTTGCCG\t25nm\tSTD\n",
      " VWF1_gene_ampR\tGGCTACGGTCTCTGAAACACCTGCCTAGCAGGTCCTGGCAGTCCCC\t25nm\tSTD\n",
      " VWF1_block1_s1_ampF\tGGCTACGGTCTCTCCTTTGATGGGAGCATGTACAGC\t25nm\tSTD\n",
      " VWF1_block2_s1_ampR\tGGCTACGGTCTCTACTTCCGAAAAGGCTGCATCG\t25nm\tSTD\n",
      " VWF1_block2_s2_ampF\tGGCTACGGTCTCTAATGGTACCGTGACACAGGGG\t25nm\tSTD\n",
      " VWF1_block3_s1_ampR\tGGCTACGGTCTCTTGGATGTCAAAAAATTCCCCAAGATACAC\t25nm\tSTD\n",
      " VWF1_block3_s2_ampF\tGGCTACGGTCTCTTCAGACAGATACTTCAACAAGACCTGC\t25nm\tSTD\n",
      " VWF1_block4_s1_ampR\tGGCTACGGTCTCTAAGTTGCCGCTGCCATCG\t25nm\tSTD\n",
      " VWF1_block4_s2_ampF\tGGCTACGGTCTCTATCTCCTCCCAGCAGCTCATG\t25nm\tSTD\n",
      " VWF1_block5_s1_ampR\tGGCTACGGTCTCTACTGTTCTCCACTGCTCAGAGC\t25nm\tSTD\n",
      " VWF1_block5_s2_ampF\tGGCTACGGTCTCTCTGGGGGGCTGGAGTGC\t25nm\tSTD\n",
      " VWF1_block6_s1_ampR\tGGCTACGGTCTCTTTCTCACACAGGGCCACAAAAG\t25nm\tSTD\n",
      " VWF1_block6_s2_ampF\tGGCTACGGTCTCTCAGGACCTGCCAGAGCCTG\t25nm\tSTD\n",
      " VWF1_block7_s1_ampR\tGGCTACGGTCTCTAAGGGGACACACACTGCCTATAC\t25nm\tSTD\n",
      " VWF1_block7_s2_ampF\tGGCTACGGTCTCTTCCCTCTCTCGAGACTGCAAC\t25nm\tSTD\n",
      " VWF1_block8_s1_ampR\tGGCTACGGTCTCTGGGAGGGTAGCGCTTTCCG\t25nm\tSTD\n",
      " VWF1_block8_s2_ampF\tGGCTACGGTCTCTGGATTGCCAGGACCACTCCTTC\t25nm\tSTD\n",
      " VWF1_block9_s1_ampR\tGGCTACGGTCTCTCAGATCCCACTGAAGGTGAAGTATCTG\t25nm\tSTD\n",
      " VWF1_block9_s2_ampF\tGGCTACGGTCTCTGGATGGCCAGGACGTCCAG\t25nm\tSTD\n",
      " VWF1_block10_s1_ampR\tGGCTACGGTCTCTAACTCCTGCCCCATGCTTCAG\t25nm\tSTD\n",
      " VWF1_block10_s2_ampF\tGGCTACGGTCTCTCCTGCGGCCTGTGTGGG\t25nm\tSTD\n",
      " VWF1_block11_s1_ampR\tGGCTACGGTCTCTTAGACGGGGGACAGCTTCAC\t25nm\tSTD\n",
      "\n",
      "Gene\tBlock\tForward Primer Well\tForward Primer\tReverse Primer Well\tReverse Primer\tValidated\n",
      " FRED1\t1\tG8\tGTACTCAGAGATTGCCGGAG\tH1\tAGATAGTCACGCACAAGACC\tFalse\n",
      " FRED1\t2\tB10\tTAGTGACCTAATGCCATGGG\tD4\tATCCAGGAGGTCTAGGAACC\tFalse\n",
      " FRED1\t3\tF10\tCTATAGAATCCGGGCTGGTC\tH3\tCCACGAGATAAGAGGATGGC\tFalse\n",
      " FRED1\t4\tD6\tCGATAGAACGACCAGGTAGC\tA8\tAAACCGGAGCCATACAGTAC\tFalse\n",
      " FRED1\t5\tF6\tAGTCCCAAGTTCAGACGTAC\tA3\tTCAAAGGAGCACGAACCTAC\tFalse\n",
      " FRED1\t6\tB8\tCTAATATCCCTGAGCGACGG\tE8\tGAAAAGTCCCAATGAGTGCC\tFalse\n",
      " FRED1\t7\tG10\tACGATGGGGACATAGAACAC\tE5\tAGGTGATCTGACGAATGTCC\tFalse\n",
      " FRED1\t8\tC2\tACGAGATGATGCACCGATAG\tF6\tACAGGGTATATTGAGTGCCC\tFalse\n",
      " FRED1\t9\tD8\tTCAAGGTCCGTTATGGAACC\tA11\tAAGAATTACTGACCCCTCGG\tFalse\n",
      " FRED1\t10\tF7\tACATAAGCGATCCCAAGGTC\tB7\tATTCTCACGACGCAAGATGG\tFalse\n",
      " FRED1\t11\tA6\tAGCTATAAGAATTGCCGGGC\tA6\tACGTATGGGGAACACTACAC\tFalse\n",
      " VWF1\t1\tH6\tGGAAACAATAACCATCGGCG\tC1\tATGCTAGCTGGAACTATCGG\tFalse\n",
      " VWF1\t2\tA2\tTATCAATCCGGAACCAGTGC\tF1\tTACTGATAATTCGGACGCCC\tFalse\n",
      " VWF1\t3\tE8\tACTAGATTAGCAAGGCACCC\tC7\tTCGAGACAAGAACGATTCCC\tFalse\n",
      " VWF1\t4\tA10\tTTGACAGACAATCCGTAGGC\tD5\tAGTGGTCTGTAAACCGTACC\tFalse\n",
      " VWF1\t5\tB1\tCCAGAGCTTAGGGGACATAC\tD8\tAGGACACCAGACCAATGAAG\tFalse\n",
      " VWF1\t6\tF9\tTAAGGGACGATGCTTAACCC\tA4\tTAAGGATTCATCAGGTGCGC\tFalse\n",
      " VWF1\t7\tD5\tAGCTATGATCCCGGTGTAAC\tB6\tATCACAACAAAGGACGGGTC\tFalse\n",
      " VWF1\t8\tF5\tATAGATCATGTCGGCAGTCG\tD1\tACGTGAAACTGTATCGAGCC\tFalse\n",
      " VWF1\t9\tG7\tATCAGACAACACAGAGGCTG\tE9\tACAGGAAGCAAGGTATACGC\tFalse\n",
      " VWF1\t10\tE7\tTAGCCAGGCAAAAGAGATCC\tA1\tGTTGCATCTAAGCCAAGTGC\tFalse\n",
      " VWF1\t11\tB4\tAATGTCCTTAGGCAGTCGTC\tB8\tAGTTGCAGTATCTAACCCGC\tFalse\n",
      "\n",
      "FRED1_block2_s1\tACGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGTAATACCATGGTCCCTGAAATTCGGCTCAGTTACTTCGAGCGTAATGTCTCAAATCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGCAAATAGAGCTCTCTACCTGCTTCTTTCTGTGCCTTTTGCGATTCTGCTTTAGTGCCACCAGAAGATACTACCTGGGTGCAGTGGAACTGTCATGGGACTATATGCAAAGTGATCTAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block3_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGCAAATAGAGCTCTCTACCTGCTTCTTTCTGTGCCTTTTGCGATTCTGCTTTAGTGCCACCAGAAGATACTACCTGGGTGCAGTGGAACTGTCATGGGACTATATGCAAAGTGATCTCGGTGAGCTGCCTGTGGACGCAAGATTTCCTCCTAGAGTGCCAAAATCTTTTCCATTCAACACCTCAGTCGTGTACAAAAAGACTCTGTTTGTAGAATTCACGGATCACCTTTTCAACATCGCTAAGCCAAGGCCACCCTGGATGGGTCTGCTAGGAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block4_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGCAAATAGAGCTCTCTACCTGCTTCTTTCTGTGCCTTTTGCGATTCTGCTTTAGTGCCACCAGAAGATACTACCTGGGTGCAGTGGAACTGTCATGGGACTATATGCAAAGTGATCTCGGTGAGCTGCCTGTGGACGCAAGATTTCCTCCTAGAGTGCCAAAATCTTTTCCATTCAACACCTCAGTCGTGTACAAAAAGACTCTGTTTGTAGAATTCACGGATCACCTTTTCAACATCGCTAAGCCAAGGCCACCCTGGATGGGTCTGCTAGGTCCTACCATCCAGGCTGAGGTTTATGATACAGTGGTCATTACACTTAAGAACATGGCTTCCCATCCTGTCAGTCTTCATGCTGTTGGTGTATCCTACTGGAAAGCTTCTGAGGGAGCTGAATATGATGATCAGACCAGTCAAAGGGAGAAAGAAAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block5_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGCAAATAGAGCTCTCTACCTGCTTCTTTCTGTGCCTTTTGCGATTCTGCTTTAGTGCCACCAGAAGATACTACCTGGGTGCAGTGGAACTGTCATGGGACTATATGCAAAGTGATCTCGGTGAGCTGCCTGTGGACGCAAGATTTCCTCCTAGAGTGCCAAAATCTTTTCCATTCAACACCTCAGTCGTGTACAAAAAGACTCTGTTTGTAGAATTCACGGATCACCTTTTCAACATCGCTAAGCCAAGGCCACCCTGGATGGGTCTGCTAGGTCCTACCATCCAGGCTGAGGTTTATGATACAGTGGTCATTACACTTAAGAACATGGCTTCCCATCCTGTCAGTCTTCATGCTGTTGGTGTATCCTACTGGAAAGCTTCTGAGGGAGCTGAATATGATGATCAGACCAGTCAAAGGGAGAAAGAAGATGATAAAGTCTTCCCTGGTGGAAGCCATACATATGTCTGGCAGGTCCTGAAAGAGAATGGTCCAATGGCCTCTGACCCACTGTGCCTTACCTACTCATATCTTTCTCATGTGGACCTGGTAAAAGACTTGAATTCAGGCCTCATTGGAGCCCTACTAGTATGTAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block5_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTAATGGTTATGTAAACAGATCTCTGCCAGGTCTGATTGGATGCCACAGGAAATCAGTCTATTGGCATGTGATTGGAATGGGCACCACTCCTGAAGTGCACTCAATATTCCTCGAAGGTCACACATTTCTTGTGAGGAACCATCGCCAGGCGTCCTTGGAAATCTCGCCAATAACTTTCCTTACTGCTCAAACACTCTTGATGGACCTTGGACAGTTTCTACTGTTTTGTCATATCTCTTCCCACCAACATGATGGCATGGAAGCTTATGTCAAAGTAGACAGCTGTCCAGAGGAACCCCAACTACGAATGAAAAATAATGAAGAAGCGGAAGACTATGATGATGATCTTACTGATTCTGAAATGGATGTGGTCAGGTTTGATGATGACAACTCTCCTTCCTTTATCCAAATTCGCTCAGTTGCCAAGAAGCATCCTAAAACTTGGGTACATTACATTGCTGCTGAAGAGGAGGACTGGGACTATGCTCCCTTAGTCCTCGCCCCCGATGACAGAAGTTATAAAAGTCAATATTTGAACAATGGCCCTCAGCGGATTGGTAGGAAGTACAAAAAAGTCCGATTTATGGCATACACAGATGAAACCTTTAAGACTCGTGAAGCTATTCAGCATGAATCAGGAATCTTGGGACCTTTACTTTATGGGGAAGTTGGAGACACACTGTTGATTATATTTAAGAATCAAGCAAGCAGACCATATAACATCTACCCTCACGGAATCACTGATGTCCGTCCTTTGTATTCAAGGAGATTACCAAAAGGTGTAAAACATTTGAAGGATTTTCCAATTCTGCCAGGAGAAATATTCAAATATAAATGGACAGTGACTGTAGAAGATGGGCCAACTAAATCAGATCCTCGGTGCCTGACCCGCTATTACTCTTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block6_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGCAAATAGAGCTCTCTACCTGCTTCTTTCTGTGCCTTTTGCGATTCTGCTTTAGTGCCACCAGAAGATACTACCTGGGTGCAGTGGAACTGTCATGGGACTATATGCAAAGTGATCTCGGTGAGCTGCCTGTGGACGCAAGATTTCCTCCTAGAGTGCCAAAATCTTTTCCATTCAACACCTCAGTCGTGTACAAAAAGACTCTGTTTGTAGAATTCACGGATCACCTTTTCAACATCGCTAAGCCAAGGCCACCCTGGATGGGTCTGCTAGGTCCTACCATCCAGGCTGAGGTTTATGATACAGTGGTCATTACACTTAAGAACATGGCTTCCCATCCTGTCAGTCTTCATGCTGTTGGTGTATCCTACTGGAAAGCTTCTGAGGGAGCTGAATATGATGATCAGACCAGTCAAAGGGAGAAAGAAGATGATAAAGTCTTCCCTGGTGGAAGCCATACATATGTCTGGCAGGTCCTGAAAGAGAATGGTCCAATGGCCTCTGACCCACTGTGCCTTACCTACTCATATCTTTCTCATGTGGACCTGGTAAAAGACTTGAATTCAGGCCTCATTGGAGCCCTACTAGTATGTAGAGAAGGGAGTCTGGCCAAGGAAAAGACACAGACCTTGCACAAATTTATACTACTTTTTGCTGTATTTGATGAAGGGAAAAGTTGGCACTCAGAAACAAAGAACTCCTTGATGCAGGATAGGGATGCTGCATCTGCTCGGGCCTGGCCTAAAATAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block6_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTTGGAAATCTCGCCAATAACTTTCCTTACTGCTCAAACACTCTTGATGGACCTTGGACAGTTTCTACTGTTTTGTCATATCTCTTCCCACCAACATGATGGCATGGAAGCTTATGTCAAAGTAGACAGCTGTCCAGAGGAACCCCAACTACGAATGAAAAATAATGAAGAAGCGGAAGACTATGATGATGATCTTACTGATTCTGAAATGGATGTGGTCAGGTTTGATGATGACAACTCTCCTTCCTTTATCCAAATTCGCTCAGTTGCCAAGAAGCATCCTAAAACTTGGGTACATTACATTGCTGCTGAAGAGGAGGACTGGGACTATGCTCCCTTAGTCCTCGCCCCCGATGACAGAAGTTATAAAAGTCAATATTTGAACAATGGCCCTCAGCGGATTGGTAGGAAGTACAAAAAAGTCCGATTTATGGCATACACAGATGAAACCTTTAAGACTCGTGAAGCTATTCAGCATGAATCAGGAATCTTGGGACCTTTACTTTATGGGGAAGTTGGAGACACACTGTTGATTATATTTAAGAATCAAGCAAGCAGACCATATAACATCTACCCTCACGGAATCACTGATGTCCGTCCTTTGTATTCAAGGAGATTACCAAAAGGTGTAAAACATTTGAAGGATTTTCCAATTCTGCCAGGAGAAATATTCAAATATAAATGGACAGTGACTGTAGAAGATGGGCCAACTAAATCAGATCCTCGGTGCCTGACCCGCTATTACTCTTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block7_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGCAAATAGAGCTCTCTACCTGCTTCTTTCTGTGCCTTTTGCGATTCTGCTTTAGTGCCACCAGAAGATACTACCTGGGTGCAGTGGAACTGTCATGGGACTATATGCAAAGTGATCTCGGTGAGCTGCCTGTGGACGCAAGATTTCCTCCTAGAGTGCCAAAATCTTTTCCATTCAACACCTCAGTCGTGTACAAAAAGACTCTGTTTGTAGAATTCACGGATCACCTTTTCAACATCGCTAAGCCAAGGCCACCCTGGATGGGTCTGCTAGGTCCTACCATCCAGGCTGAGGTTTATGATACAGTGGTCATTACACTTAAGAACATGGCTTCCCATCCTGTCAGTCTTCATGCTGTTGGTGTATCCTACTGGAAAGCTTCTGAGGGAGCTGAATATGATGATCAGACCAGTCAAAGGGAGAAAGAAGATGATAAAGTCTTCCCTGGTGGAAGCCATACATATGTCTGGCAGGTCCTGAAAGAGAATGGTCCAATGGCCTCTGACCCACTGTGCCTTACCTACTCATATCTTTCTCATGTGGACCTGGTAAAAGACTTGAATTCAGGCCTCATTGGAGCCCTACTAGTATGTAGAGAAGGGAGTCTGGCCAAGGAAAAGACACAGACCTTGCACAAATTTATACTACTTTTTGCTGTATTTGATGAAGGGAAAAGTTGGCACTCAGAAACAAAGAACTCCTTGATGCAGGATAGGGATGCTGCATCTGCTCGGGCCTGGCCTAAAATGCACACAGTCAATGGTTATGTAAACAGATCTCTGCCAGGTCTGATTGGATGCCACAGGAAATCAGTCTATTGGCATGTGATTGGAATGGGCACCACTCCTGAAGTGCACTCAATATTCCTCGAAGGTCACACATTTCTTGTGAGGAACCATCGAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block7_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTAAATAATGAAGAAGCGGAAGACTATGATGATGATCTTACTGATTCTGAAATGGATGTGGTCAGGTTTGATGATGACAACTCTCCTTCCTTTATCCAAATTCGCTCAGTTGCCAAGAAGCATCCTAAAACTTGGGTACATTACATTGCTGCTGAAGAGGAGGACTGGGACTATGCTCCCTTAGTCCTCGCCCCCGATGACAGAAGTTATAAAAGTCAATATTTGAACAATGGCCCTCAGCGGATTGGTAGGAAGTACAAAAAAGTCCGATTTATGGCATACACAGATGAAACCTTTAAGACTCGTGAAGCTATTCAGCATGAATCAGGAATCTTGGGACCTTTACTTTATGGGGAAGTTGGAGACACACTGTTGATTATATTTAAGAATCAAGCAAGCAGACCATATAACATCTACCCTCACGGAATCACTGATGTCCGTCCTTTGTATTCAAGGAGATTACCAAAAGGTGTAAAACATTTGAAGGATTTTCCAATTCTGCCAGGAGAAATATTCAAATATAAATGGACAGTGACTGTAGAAGATGGGCCAACTAAATCAGATCCTCGGTGCCTGACCCGCTATTACTCTTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block8_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTCTGAAGAGGAGGACTGGGACTATGCTCCCTTAGTCCTCGCCCCCGATGACAGAAGTTATAAAAGTCAATATTTGAACAATGGCCCTCAGCGGATTGGTAGGAAGTACAAAAAAGTCCGATTTATGGCATACACAGATGAAACCTTTAAGACTCGTGAAGCTATTCAGCATGAATCAGGAATCTTGGGACCTTTACTTTATGGGGAAGTTGGAGACACACTGTTGATTATATTTAAGAATCAAGCAAGCAGACCATATAACATCTACCCTCACGGAATCACTGATGTCCGTCCTTTGTATTCAAGGAGATTACCAAAAGGTGTAAAACATTTGAAGGATTTTCCAATTCTGCCAGGAGAAATATTCAAATATAAATGGACAGTGACTGTAGAAGATGGGCCAACTAAATCAGATCCTCGGTGCCTGACCCGCTATTACTCTTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block9_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTTCAGCATGAATCAGGAATCTTGGGACCTTTACTTTATGGGGAAGTTGGAGACACACTGTTGATTATATTTAAGAATCAAGCAAGCAGACCATATAACATCTACCCTCACGGAATCACTGATGTCCGTCCTTTGTATTCAAGGAGATTACCAAAAGGTGTAAAACATTTGAAGGATTTTCCAATTCTGCCAGGAGAAATATTCAAATATAAATGGACAGTGACTGTAGAAGATGGGCCAACTAAATCAGATCCTCGGTGCCTGACCCGCTATTACTCTTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block10_s2\tACGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGTAATACCATGGTCCCTGAAATTCGGCTCAGTTACTTCGAGCGTAATGTCCCGCGTGATTACGAGTCGGGCTACGGTCTCTAAGGTGTAAAACATTTGAAGGATTTTCCAATTCTGCCAGGAGAAATATTCAAATATAAATGGACAGTGACTGTAGAAGATGGGCCAACTAAATCAGATCCTCGGTGCCTGACCCGCTATTACTCTTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block2_s1\tACGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGTAATACCATGGTCCCTGAAATTCGGCTCAGTTACTTCGAGCGTAATGTCTCAAACCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block3_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTGACTTCGTCAACACCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCAAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block4_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTGACTTCGTCAACACCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCATTTGTTTGTCAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block5_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTGACTTCGTCAACACCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCATTTGTTTGTCAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTTCAAGTCCTGCTGTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block5_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTGCGCCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCGGCACCTCCCTCTCTCGAGACTGCAACACATGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGCCAGTACCTGCTGGCCCGGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTGCCATGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTATGCCGGGAAGACCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTGTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block6_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTGACTTCGTCAACACCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCATTTGTTTGTCAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTTCAAGTCCTGCTGTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTGGTGTGAACGGGCATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block6_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCGGCACCTCCCTCTCTCGAGACTGCAACACATGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGCCAGTACCTGCTGGCCCGGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTGCCATGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTATGCCGGGAAGACCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTGTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block7_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTTCCCTCTCTCGAGACTGCAACACATGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGCCAGTACCTGCTGGCCCGGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTGCCATGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTATGCCGGGAAGACCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTGTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block8_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTGCCATGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTATGCCGGGAAGACCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTGTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block9_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTATGCCGGGAAGACCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTGTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block10_s2\tACGCCGCCACGTGTTCGTTAACTGTTGATTGGTGGCACATAAGTAATACCATGGTCCCTGAAATTCGGCTCAGTTACTTCGAGCGTAATCCGCGTGATTACGAGTCGGGCTACGGTCTCTCCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTGTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      "\n",
      "FRED1_block1_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTTGGACGCAAGATTTCCTCCTAGAGTGCCAAAATCTTTTCCATTCAACACCTCAGTCGTGTACAAAAAGACTCTGTTTGTAGAATTCACGGATCACCTTTTCAACATCGCTAAGCCAAGGCCACCCTGGATGGGTCTGCTAGGTCCTACCATCCAGGCTGAGGTTTATGATACAGTGGTCATTACACTTAAGAACATGGCTTCCCATCCTGTCAGTCTTCATGCTGTTGGTGTATCCTACTGGAAAGCTTCTGAGGGAGCTGAATATGATGATCAGACCAGTCAAAGGGAGAAAGAAGATGATAAAGTCTTCCCTGGTGGAAGCCATACATATGTCTGGCAGGTCCTGAAAGAGAATGGTCCAATGGCCTCTGACCCACTGTGCCTTACCTACTCATATCTTTCTCATGTGGACCTGGTAAAAGACTTGAATTCAGGCCTCATTGGAGCCCTACTAGTATGTAGAGAAGGGAGTCTGGCCAAGGAAAAGACACAGACCTTGCACAAATTTATACTACTTTTTGCTGTATTTGATGAAGGGAAAAGTTGGCACTCAGAAACAAAGAACTCCTTGATGCAGGATAGGGATGCTGCATCTGCTCGGGCCTGGCCTAAAATGCACACAGTCAATGGTTATGTAAACAGATCTCTGCCAGGTCTGATTGGATGCCACAGGAAATCAGTCTATTGGCATGTGATTGGAATGGGCACCACTCCTGAAGTGCACTCAATATTCCTCGAAGGTCACACATTTCTTGTGAGGAACCATCGCCAGGCGTCCTTGGAAATCTCGCCAATAACTTTCCTTACTGCTCAAACACTCTTGATGGACCTTGGACAGTTTCTACTGTTTTGTCATATCTCTTCCCACCAACATGATGGCATGGAAGCTTATGTCAAAGTAGACAGCTGTCCAGAGGAACCCCAACTACGAATGAAAAATAATGAAGAAGCGGAAGACTATGATGATGATCTTACTGATTCTGAAATGGATGTGGTCAGGTTTGATGATGACAACTCTCCTTCCTTTATCCAAATTCGCTCAGTTGCCAAGAAGCATCCTAAAACTTGGGTACATTACATTGCTGCTGAAGAGGAGGACTGGGACTATGCTCCCTTAGTCCTCGCCCCCGATGACAGAAGTTATAAAAGTCAATATTTGAACAATGGCCCTCAGCGGATTGGTAGGAAGTACAAAAAAGTCCGATTTATGGCATACACAGATGAAACCTTTAAGACTCGTGAAGCTATTCAGCATGAATCAGGAATCTTGGGACCTTTACTTTATGGGGAAGTTGGAGACACACTGTTGATTATATTTAAGAATCAAGCAAGCAGACCATATAACATCTACCCTCACGGAATCACTGATGTCCGTCCTTTGTATTCAAGGAGATTACCAAAAGGTGTAAAACATTTGAAGGATTTTCCAATTCTGCCAGGAGAAATATTCAAATATAAATGGACAGTGACTGTAGAAGATGGGCCAACTAAATCAGATCCTCGGTGCCTGACCCGCTATTACTCTTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block2_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTCTGAGGTTTATGATACAGTGGTCATTACACTTAAGAACATGGCTTCCCATCCTGTCAGTCTTCATGCTGTTGGTGTATCCTACTGGAAAGCTTCTGAGGGAGCTGAATATGATGATCAGACCAGTCAAAGGGAGAAAGAAGATGATAAAGTCTTCCCTGGTGGAAGCCATACATATGTCTGGCAGGTCCTGAAAGAGAATGGTCCAATGGCCTCTGACCCACTGTGCCTTACCTACTCATATCTTTCTCATGTGGACCTGGTAAAAGACTTGAATTCAGGCCTCATTGGAGCCCTACTAGTATGTAGAGAAGGGAGTCTGGCCAAGGAAAAGACACAGACCTTGCACAAATTTATACTACTTTTTGCTGTATTTGATGAAGGGAAAAGTTGGCACTCAGAAACAAAGAACTCCTTGATGCAGGATAGGGATGCTGCATCTGCTCGGGCCTGGCCTAAAATGCACACAGTCAATGGTTATGTAAACAGATCTCTGCCAGGTCTGATTGGATGCCACAGGAAATCAGTCTATTGGCATGTGATTGGAATGGGCACCACTCCTGAAGTGCACTCAATATTCCTCGAAGGTCACACATTTCTTGTGAGGAACCATCGCCAGGCGTCCTTGGAAATCTCGCCAATAACTTTCCTTACTGCTCAAACACTCTTGATGGACCTTGGACAGTTTCTACTGTTTTGTCATATCTCTTCCCACCAACATGATGGCATGGAAGCTTATGTCAAAGTAGACAGCTGTCCAGAGGAACCCCAACTACGAATGAAAAATAATGAAGAAGCGGAAGACTATGATGATGATCTTACTGATTCTGAAATGGATGTGGTCAGGTTTGATGATGACAACTCTCCTTCCTTTATCCAAATTCGCTCAGTTGCCAAGAAGCATCCTAAAACTTGGGTACATTACATTGCTGCTGAAGAGGAGGACTGGGACTATGCTCCCTTAGTCCTCGCCCCCGATGACAGAAGTTATAAAAGTCAATATTTGAACAATGGCCCTCAGCGGATTGGTAGGAAGTACAAAAAAGTCCGATTTATGGCATACACAGATGAAACCTTTAAGACTCGTGAAGCTATTCAGCATGAATCAGGAATCTTGGGACCTTTACTTTATGGGGAAGTTGGAGACACACTGTTGATTATATTTAAGAATCAAGCAAGCAGACCATATAACATCTACCCTCACGGAATCACTGATGTCCGTCCTTTGTATTCAAGGAGATTACCAAAAGGTGTAAAACATTTGAAGGATTTTCCAATTCTGCCAGGAGAAATATTCAAATATAAATGGACAGTGACTGTAGAAGATGGGCCAACTAAATCAGATCCTCGGTGCCTGACCCGCTATTACTCTTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block3_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTTGGAAGCCATACATATGTCTGGCAGGTCCTGAAAGAGAATGGTCCAATGGCCTCTGACCCACTGTGCCTTACCTACTCATATCTTTCTCATGTGGACCTGGTAAAAGACTTGAATTCAGGCCTCATTGGAGCCCTACTAGTATGTAGAGAAGGGAGTCTGGCCAAGGAAAAGACACAGACCTTGCACAAATTTATACTACTTTTTGCTGTATTTGATGAAGGGAAAAGTTGGCACTCAGAAACAAAGAACTCCTTGATGCAGGATAGGGATGCTGCATCTGCTCGGGCCTGGCCTAAAATGCACACAGTCAATGGTTATGTAAACAGATCTCTGCCAGGTCTGATTGGATGCCACAGGAAATCAGTCTATTGGCATGTGATTGGAATGGGCACCACTCCTGAAGTGCACTCAATATTCCTCGAAGGTCACACATTTCTTGTGAGGAACCATCGCCAGGCGTCCTTGGAAATCTCGCCAATAACTTTCCTTACTGCTCAAACACTCTTGATGGACCTTGGACAGTTTCTACTGTTTTGTCATATCTCTTCCCACCAACATGATGGCATGGAAGCTTATGTCAAAGTAGACAGCTGTCCAGAGGAACCCCAACTACGAATGAAAAATAATGAAGAAGCGGAAGACTATGATGATGATCTTACTGATTCTGAAATGGATGTGGTCAGGTTTGATGATGACAACTCTCCTTCCTTTATCCAAATTCGCTCAGTTGCCAAGAAGCATCCTAAAACTTGGGTACATTACATTGCTGCTGAAGAGGAGGACTGGGACTATGCTCCCTTAGTCCTCGCCCCCGATGACAGAAGTTATAAAAGTCAATATTTGAACAATGGCCCTCAGCGGATTGGTAGGAAGTACAAAAAAGTCCGATTTATGGCATACACAGATGAAACCTTTAAGACTCGTGAAGCTATTCAGCATGAATCAGGAATCTTGGGACCTTTACTTTATGGGGAAGTTGGAGACACACTGTTGATTATATTTAAGAATCAAGCAAGCAGACCATATAACATCTACCCTCACGGAATCACTGATGTCCGTCCTTTGTATTCAAGGAGATTACCAAAAGGTGTAAAACATTTGAAGGATTTTCCAATTCTGCCAGGAGAAATATTCAAATATAAATGGACAGTGACTGTAGAAGATGGGCCAACTAAATCAGATCCTCGGTGCCTGACCCGCTATTACTCTTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block4_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTTCTGGCCAAGGAAAAGACACAGACCTTGCACAAATTTATACTACTTTTTGCTGTATTTGATGAAGGGAAAAGTTGGCACTCAGAAACAAAGAACTCCTTGATGCAGGATAGGGATGCTGCATCTGCTCGGGCCTGGCCTAAAATGCACACAGTCAATGGTTATGTAAACAGATCTCTGCCAGGTCTGATTGGATGCCACAGGAAATCAGTCTATTGGCATGTGATTGGAATGGGCACCACTCCTGAAGTGCACTCAATATTCCTCGAAGGTCACACATTTCTTGTGAGGAACCATCGCCAGGCGTCCTTGGAAATCTCGCCAATAACTTTCCTTACTGCTCAAACACTCTTGATGGACCTTGGACAGTTTCTACTGTTTTGTCATATCTCTTCCCACCAACATGATGGCATGGAAGCTTATGTCAAAGTAGACAGCTGTCCAGAGGAACCCCAACTACGAATGAAAAATAATGAAGAAGCGGAAGACTATGATGATGATCTTACTGATTCTGAAATGGATGTGGTCAGGTTTGATGATGACAACTCTCCTTCCTTTATCCAAATTCGCTCAGTTGCCAAGAAGCATCCTAAAACTTGGGTACATTACATTGCTGCTGAAGAGGAGGACTGGGACTATGCTCCCTTAGTCCTCGCCCCCGATGACAGAAGTTATAAAAGTCAATATTTGAACAATGGCCCTCAGCGGATTGGTAGGAAGTACAAAAAAGTCCGATTTATGGCATACACAGATGAAACCTTTAAGACTCGTGAAGCTATTCAGCATGAATCAGGAATCTTGGGACCTTTACTTTATGGGGAAGTTGGAGACACACTGTTGATTATATTTAAGAATCAAGCAAGCAGACCATATAACATCTACCCTCACGGAATCACTGATGTCCGTCCTTTGTATTCAAGGAGATTACCAAAAGGTGTAAAACATTTGAAGGATTTTCCAATTCTGCCAGGAGAAATATTCAAATATAAATGGACAGTGACTGTAGAAGATGGGCCAACTAAATCAGATCCTCGGTGCCTGACCCGCTATTACTCTTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block8_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGCAAATAGAGCTCTCTACCTGCTTCTTTCTGTGCCTTTTGCGATTCTGCTTTAGTGCCACCAGAAGATACTACCTGGGTGCAGTGGAACTGTCATGGGACTATATGCAAAGTGATCTCGGTGAGCTGCCTGTGGACGCAAGATTTCCTCCTAGAGTGCCAAAATCTTTTCCATTCAACACCTCAGTCGTGTACAAAAAGACTCTGTTTGTAGAATTCACGGATCACCTTTTCAACATCGCTAAGCCAAGGCCACCCTGGATGGGTCTGCTAGGTCCTACCATCCAGGCTGAGGTTTATGATACAGTGGTCATTACACTTAAGAACATGGCTTCCCATCCTGTCAGTCTTCATGCTGTTGGTGTATCCTACTGGAAAGCTTCTGAGGGAGCTGAATATGATGATCAGACCAGTCAAAGGGAGAAAGAAGATGATAAAGTCTTCCCTGGTGGAAGCCATACATATGTCTGGCAGGTCCTGAAAGAGAATGGTCCAATGGCCTCTGACCCACTGTGCCTTACCTACTCATATCTTTCTCATGTGGACCTGGTAAAAGACTTGAATTCAGGCCTCATTGGAGCCCTACTAGTATGTAGAGAAGGGAGTCTGGCCAAGGAAAAGACACAGACCTTGCACAAATTTATACTACTTTTTGCTGTATTTGATGAAGGGAAAAGTTGGCACTCAGAAACAAAGAACTCCTTGATGCAGGATAGGGATGCTGCATCTGCTCGGGCCTGGCCTAAAATGCACACAGTCAATGGTTATGTAAACAGATCTCTGCCAGGTCTGATTGGATGCCACAGGAAATCAGTCTATTGGCATGTGATTGGAATGGGCACCACTCCTGAAGTGCACTCAATATTCCTCGAAGGTCACACATTTCTTGTGAGGAACCATCGCCAGGCGTCCTTGGAAATCTCGCCAATAACTTTCCTTACTGCTCAAACACTCTTGATGGACCTTGGACAGTTTCTACTGTTTTGTCATATCTCTTCCCACCAACATGATGGCATGGAAGCTTATGTCAAAGTAGACAGCTGTCCAGAGGAACCCCAACTAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block9_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGCAAATAGAGCTCTCTACCTGCTTCTTTCTGTGCCTTTTGCGATTCTGCTTTAGTGCCACCAGAAGATACTACCTGGGTGCAGTGGAACTGTCATGGGACTATATGCAAAGTGATCTCGGTGAGCTGCCTGTGGACGCAAGATTTCCTCCTAGAGTGCCAAAATCTTTTCCATTCAACACCTCAGTCGTGTACAAAAAGACTCTGTTTGTAGAATTCACGGATCACCTTTTCAACATCGCTAAGCCAAGGCCACCCTGGATGGGTCTGCTAGGTCCTACCATCCAGGCTGAGGTTTATGATACAGTGGTCATTACACTTAAGAACATGGCTTCCCATCCTGTCAGTCTTCATGCTGTTGGTGTATCCTACTGGAAAGCTTCTGAGGGAGCTGAATATGATGATCAGACCAGTCAAAGGGAGAAAGAAGATGATAAAGTCTTCCCTGGTGGAAGCCATACATATGTCTGGCAGGTCCTGAAAGAGAATGGTCCAATGGCCTCTGACCCACTGTGCCTTACCTACTCATATCTTTCTCATGTGGACCTGGTAAAAGACTTGAATTCAGGCCTCATTGGAGCCCTACTAGTATGTAGAGAAGGGAGTCTGGCCAAGGAAAAGACACAGACCTTGCACAAATTTATACTACTTTTTGCTGTATTTGATGAAGGGAAAAGTTGGCACTCAGAAACAAAGAACTCCTTGATGCAGGATAGGGATGCTGCATCTGCTCGGGCCTGGCCTAAAATGCACACAGTCAATGGTTATGTAAACAGATCTCTGCCAGGTCTGATTGGATGCCACAGGAAATCAGTCTATTGGCATGTGATTGGAATGGGCACCACTCCTGAAGTGCACTCAATATTCCTCGAAGGTCACACATTTCTTGTGAGGAACCATCGCCAGGCGTCCTTGGAAATCTCGCCAATAACTTTCCTTACTGCTCAAACACTCTTGATGGACCTTGGACAGTTTCTACTGTTTTGTCATATCTCTTCCCACCAACATGATGGCATGGAAGCTTATGTCAAAGTAGACAGCTGTCCAGAGGAACCCCAACTACGAATGAAAAATAATGAAGAAGCGGAAGACTATGATGATGATCTTACTGATTCTGAAATGGATGTGGTCAGGTTTGATGATGACAACTCTCCTTCCTTTATCCAAATTCGCTCAGTTGCCAAGAAGCATCCTAAAACTTGGGTACATTAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block10_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGCAAATAGAGCTCTCTACCTGCTTCTTTCTGTGCCTTTTGCGATTCTGCTTTAGTGCCACCAGAAGATACTACCTGGGTGCAGTGGAACTGTCATGGGACTATATGCAAAGTGATCTCGGTGAGCTGCCTGTGGACGCAAGATTTCCTCCTAGAGTGCCAAAATCTTTTCCATTCAACACCTCAGTCGTGTACAAAAAGACTCTGTTTGTAGAATTCACGGATCACCTTTTCAACATCGCTAAGCCAAGGCCACCCTGGATGGGTCTGCTAGGTCCTACCATCCAGGCTGAGGTTTATGATACAGTGGTCATTACACTTAAGAACATGGCTTCCCATCCTGTCAGTCTTCATGCTGTTGGTGTATCCTACTGGAAAGCTTCTGAGGGAGCTGAATATGATGATCAGACCAGTCAAAGGGAGAAAGAAGATGATAAAGTCTTCCCTGGTGGAAGCCATACATATGTCTGGCAGGTCCTGAAAGAGAATGGTCCAATGGCCTCTGACCCACTGTGCCTTACCTACTCATATCTTTCTCATGTGGACCTGGTAAAAGACTTGAATTCAGGCCTCATTGGAGCCCTACTAGTATGTAGAGAAGGGAGTCTGGCCAAGGAAAAGACACAGACCTTGCACAAATTTATACTACTTTTTGCTGTATTTGATGAAGGGAAAAGTTGGCACTCAGAAACAAAGAACTCCTTGATGCAGGATAGGGATGCTGCATCTGCTCGGGCCTGGCCTAAAATGCACACAGTCAATGGTTATGTAAACAGATCTCTGCCAGGTCTGATTGGATGCCACAGGAAATCAGTCTATTGGCATGTGATTGGAATGGGCACCACTCCTGAAGTGCACTCAATATTCCTCGAAGGTCACACATTTCTTGTGAGGAACCATCGCCAGGCGTCCTTGGAAATCTCGCCAATAACTTTCCTTACTGCTCAAACACTCTTGATGGACCTTGGACAGTTTCTACTGTTTTGTCATATCTCTTCCCACCAACATGATGGCATGGAAGCTTATGTCAAAGTAGACAGCTGTCCAGAGGAACCCCAACTACGAATGAAAAATAATGAAGAAGCGGAAGACTATGATGATGATCTTACTGATTCTGAAATGGATGTGGTCAGGTTTGATGATGACAACTCTCCTTCCTTTATCCAAATTCGCTCAGTTGCCAAGAAGCATCCTAAAACTTGGGTACATTACATTGCTGCTGAAGAGGAGGACTGGGACTATGCTCCCTTAGTCCTCGCCCCCGATGACAGAAGTTATAAAAGTCAATATTTGAACAATGGCCCTCAGCGGATTGGTAGGAAGTACAAAAAAGTCCGATTTATGGCATACACAGATGAAACCTTTAAGACTAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " FRED1_block11_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGCAAATAGAGCTCTCTACCTGCTTCTTTCTGTGCCTTTTGCGATTCTGCTTTAGTGCCACCAGAAGATACTACCTGGGTGCAGTGGAACTGTCATGGGACTATATGCAAAGTGATCTCGGTGAGCTGCCTGTGGACGCAAGATTTCCTCCTAGAGTGCCAAAATCTTTTCCATTCAACACCTCAGTCGTGTACAAAAAGACTCTGTTTGTAGAATTCACGGATCACCTTTTCAACATCGCTAAGCCAAGGCCACCCTGGATGGGTCTGCTAGGTCCTACCATCCAGGCTGAGGTTTATGATACAGTGGTCATTACACTTAAGAACATGGCTTCCCATCCTGTCAGTCTTCATGCTGTTGGTGTATCCTACTGGAAAGCTTCTGAGGGAGCTGAATATGATGATCAGACCAGTCAAAGGGAGAAAGAAGATGATAAAGTCTTCCCTGGTGGAAGCCATACATATGTCTGGCAGGTCCTGAAAGAGAATGGTCCAATGGCCTCTGACCCACTGTGCCTTACCTACTCATATCTTTCTCATGTGGACCTGGTAAAAGACTTGAATTCAGGCCTCATTGGAGCCCTACTAGTATGTAGAGAAGGGAGTCTGGCCAAGGAAAAGACACAGACCTTGCACAAATTTATACTACTTTTTGCTGTATTTGATGAAGGGAAAAGTTGGCACTCAGAAACAAAGAACTCCTTGATGCAGGATAGGGATGCTGCATCTGCTCGGGCCTGGCCTAAAATGCACACAGTCAATGGTTATGTAAACAGATCTCTGCCAGGTCTGATTGGATGCCACAGGAAATCAGTCTATTGGCATGTGATTGGAATGGGCACCACTCCTGAAGTGCACTCAATATTCCTCGAAGGTCACACATTTCTTGTGAGGAACCATCGCCAGGCGTCCTTGGAAATCTCGCCAATAACTTTCCTTACTGCTCAAACACTCTTGATGGACCTTGGACAGTTTCTACTGTTTTGTCATATCTCTTCCCACCAACATGATGGCATGGAAGCTTATGTCAAAGTAGACAGCTGTCCAGAGGAACCCCAACTACGAATGAAAAATAATGAAGAAGCGGAAGACTATGATGATGATCTTACTGATTCTGAAATGGATGTGGTCAGGTTTGATGATGACAACTCTCCTTCCTTTATCCAAATTCGCTCAGTTGCCAAGAAGCATCCTAAAACTTGGGTACATTACATTGCTGCTGAAGAGGAGGACTGGGACTATGCTCCCTTAGTCCTCGCCCCCGATGACAGAAGTTATAAAAGTCAATATTTGAACAATGGCCCTCAGCGGATTGGTAGGAAGTACAAAAAAGTCCGATTTATGGCATACACAGATGAAACCTTTAAGACTCGTGAAGCTATTCAGCATGAATCAGGAATCTTGGGACCTTTACTTTATGGGGAAGTTGGAGACACACTGTTGATTATATTTAAGAATCAAGCAAGCAGACCATATAACATCTACCCTCACGGAATCACTGATGTCCGTCCTTTGTATTCAAGGAGATAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block1_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTCCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCATTTGTTTGTCAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTTCAAGTCCTGCTGTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTGGTGTGAACGGGCATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAGACTTTGTGTGAGTGTGCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTGCGCCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCGGCACCTCCCTCTCTCGAGACTGCAACACATGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGCCAGTACCTGCTGGCCCGGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTGCCATGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTATGCCGGGAAGACCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTGTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block2_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTTCAAGTCCTGCTGTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTGGTGTGAACGGGCATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAGACTTTGTGTGAGTGTGCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTGCGCCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCGGCACCTCCCTCTCTCGAGACTGCAACACATGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGCCAGTACCTGCTGGCCCGGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTGCCATGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTATGCCGGGAAGACCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTGTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block3_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTGGTGTGAACGGGCATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAGACTTTGTGTGAGTGTGCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTGCGCCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCGGCACCTCCCTCTCTCGAGACTGCAACACATGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGCCAGTACCTGCTGGCCCGGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTGCCATGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTATGCCGGGAAGACCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTGTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block4_s2\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAGACTTTGTGTGAGTGTGCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTGCGCCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCGGCACCTCCCTCTCTCGAGACTGCAACACATGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGCCAGTACCTGCTGGCCCGGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTGCCATGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTATGCCGGGAAGACCTGCGGCCTGTGTGGGAATTACAATGGCAACCAGGGCGACGACTTCCTTACCCCCTCTGGGCTGGCGGAGCCCCGGGTGGAGGACTTCGGGAACGCCTGGAAGCTGCACGGGGACTGCCAGGACCTGTAGGCTAGGCAGGTGTTTCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block7_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTGACTTCGTCAACACCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCATTTGTTTGTCAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTTCAAGTCCTGCTGTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTGGTGTGAACGGGCATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAGACTTTGTGTGAGTGTGCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block8_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTGACTTCGTCAACACCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCATTTGTTTGTCAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTTCAAGTCCTGCTGTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTGGTGTGAACGGGCATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAGACTTTGTGTGAGTGTGCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTGCGCCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block9_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTGACTTCGTCAACACCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCATTTGTTTGTCAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTTCAAGTCCTGCTGTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTGGTGTGAACGGGCATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAGACTTTGTGTGAGTGTGCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTGCGCCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCGGCACCTCCCTCTCTCGAGACTGCAACACATGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block10_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTGACTTCGTCAACACCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCATTTGTTTGTCAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTTCAAGTCCTGCTGTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTGGTGTGAACGGGCATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAGACTTTGTGTGAGTGTGCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTGCGCCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCGGCACCTCCCTCTCTCGAGACTGCAACACATGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGCCAGTACCTGCTGGCCCGGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      " VWF1_block11_s1\tCCGCGTGATTACGAGTCGGGCTACGGTCTCTACGACACCTGCCTAGTGGCATGATTCCTGCCAGATTTGCCGGGGTGCTGCTTGCTCTGGCCCTCATTTTGCCAGGGACCCTTTGTGCAGAAGGAACTCGCGGCAGGTCATCCACGGCCCGATGCAGCCTTTTCGGAAGTGACTTCGTCAACACCTTTGATGGGAGCATGTACAGCTTTGCGGGATACTGCAGTTACCTCCTGGCAGGGGGCTGCCAGAAACGCTCCTTCTCGATTATTGGGGACTTCCAGAATGGCAAGAGAGTGAGCCTCTCCGTGTATCTTGGGGAATTTTTTGACATCCATTTGTTTGTCAATGGTACCGTGACACAGGGGGACCAAAGAGTCTCCATGCCCTATGCCTCCAAAGGGCTGTATCTAGAAACTGAGGCTGGGTACTACAAGCTGTCCGGTGAGGCCTATGGCTTTGTGGCCAGGATCGATGGCAGCGGCAACTTTCAAGTCCTGCTGTCAGACAGATACTTCAACAAGACCTGCGGGCTGTGTGGCAACTTTAACATCTTTGCTGAAGATGACTTTATGACCCAAGAAGGGACCTTGACCTCGGACCCTTATGACTTTGCCAACTCATGGGCTCTGAGCAGTGGAGAACAGTGGTGTGAACGGGCATCTCCTCCCAGCAGCTCATGCAACATCTCCTCTGGGGAAATGCAGAAGGGCCTGTGGGAGCAGTGCCAGCTTCTGAAGAGCACCTCGGTGTTTGCCCGCTGCCACCCTCTGGTGGACCCCGAGCCTTTTGTGGCCCTGTGTGAGAAGACTTTGTGTGAGTGTGCTGGGGGGCTGGAGTGCGCCTGCCCTGCCCTCCTGGAGTACGCCCGGACCTGTGCCCAGGAGGGAATGGTGCTGTACGGCTGGACCGACCACAGCGCGTGCAGCCCAGTGTGCCCTGCTGGTATGGAGTATAGGCAGTGTGTGTCCCCTTGCGCCAGGACCTGCCAGAGCCTGCACATCAATGAAATGTGTCAGGAGCGATGCGTGGATGGCTGCAGCTGCCCTGAGGGACAGCTCCTGGATGAAGGCCTCTGCGTGGAGAGCACCGAGTGTCCCTGCGTGCATTCCGGAAAGCGCTACCCTCCCGGCACCTCCCTCTCTCGAGACTGCAACACATGCATTTGCCGAAACAGCCAGTGGATCTGCAGCAATGAAGAATGTCCAGGGGAGTGCCTTGTCACAGGTCAATCACACTTCAAGAGCTTTGACAACAGATACTTCACCTTCAGTGGGATCTGCCAGTACCTGCTGGCCCGGGATTGCCAGGACCACTCCTTCTCCATTGTCATTGAGACTGTCCAGTGTGCTGATGACCGCGACGCTGTGTGCACCCGCTCCGTCACCGTCCGGCTGCCTGGCCTGCACAACAGCCTTGTGAAACTGAAGCATGGGGCAGGAGTTGCCATGGATGGCCAGGACGTCCAGCTCCCCCTCCTGAAAGGTGACCTCCGCATCCAGCATACAGTGACGGCCTCCGTGCGCCTCAGCTACGGGGAGGACCTGCAGATGGACTGGGATGGCCGCGGGAGGCTGCTGGTGAAGCTGTCCCCCGTCTAAGAGACCGTAGCCAGGCTGCCACTTGCTAACCC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Make the 5' library\n",
    "oligo_array,amp_primers,gblocks,amp_primer_dict,breakpoint_df = write_oligo_library({\n",
    "                                            'FRED1':FRED_SB1,\n",
    "                                            'VWF1':VWF_SB1},\n",
    "                                          oligo_file='Johnsenlab/5p_oligos.csv',\n",
    "                                          primer_file='Johnsenlab/5p_primers.tsv',\n",
    "                                          gbl_file='Johnsenlab/5p_gblocks.tsv',\n",
    "                                          gbl_large_file='Johnsenlab/5p_gblocks_large.tsv',\n",
    "                                          amp_primer_key_file='Johnsenlab/5p_ampkey.tsv',\n",
    "                                          breakpoint_file='Johnsenlab/5p_breakpoints.tsv',\n",
    "                                          block_size_range=[148,190],\n",
    "                                          primer_set=orthogonal_primers_touse,\n",
    "                                          validated_primer_set=validated_primer_combos_filtered,\n",
    "                                          aa_start={                    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25f0736e-4105-48ec-bae8-2b0b22d46b51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subtract the primer pairs used in making the 5p libraries from the remaining pairs to make the list for the internal libraries\n",
    "\n",
    "#1. Read the two primer tables\n",
    "df_used = pd.read_csv(\n",
    "    \"Johnsenlab/5p_ampkey.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    usecols=[\"Forward Primer\", \"Reverse Primer\"]\n",
    ")\n",
    "\n",
    "df_all = orthogonal_primers_touse  # keep every column\n",
    "\n",
    "#2. Build helper columns to subracting \n",
    "df_used_norm = df_used.assign(\n",
    "    FWD=df_used[\"Forward Primer\"].str.strip().str.upper(),\n",
    "    REV=df_used[\"Reverse Primer\"].str.strip().str.upper()\n",
    ")\n",
    "\n",
    "df_all_norm = df_all.assign(\n",
    "    FWD=df_all[\"Forward Primer\"].str.strip().str.upper(),\n",
    "    REV=df_all[\"Reverse Primer\"].str.strip().str.upper()\n",
    ")\n",
    "\n",
    "#3. Identify pairs that are already used\n",
    "used_index = df_used_norm.set_index([\"FWD\", \"REV\"]).index\n",
    "mask       = ~df_all_norm.set_index([\"FWD\", \"REV\"]).index.isin(used_index)\n",
    "\n",
    "#4. Keep the rows (and *all* original columns) that remain\n",
    "orthogonal_primers_remaining = df_all.loc[mask].reset_index(drop=True)\n",
    "\n",
    "print(f\"{len(orthogonal_primers_remaining_iter5)} primer pairs remain.\")\n",
    "orthogonal_primers_remaining.head()\n",
    "orthogonal_primers_remaining.to_csv(\"orthogonal_primers_remaining.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afa46345-3d1f-473e-aef5-b9b2045d4b80",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Make make the internal libraries\n",
    "oligo_array,amp_primers,gblocks,amp_primer_dict,breakpoint_df = write_oligo_library({\n",
    "                                            'FRED2':FRED_SB2,\n",
    "                                            'VWF2':VWF_SB2,\n",
    "                                            'VWF3':VWF_SB3,\n",
    "                                            'VWF4':VWF_SB4},\n",
    "                                          oligo_file='Johnsenlab/int_oligos.csv',\n",
    "                                          primer_file='Johnsenlab/int_primers.tsv',\n",
    "                                          gbl_file='Johnsenlab/int_gblocks.tsv',\n",
    "                                          gbl_large_file='Johnsenlab/int_gblocks_large.tsv',\n",
    "                                          amp_primer_key_file='Johnsenlab/int_ampkey.tsv',\n",
    "                                          breakpoint_file='Johnsenlab/int_breakpoints.tsv',\n",
    "                                          block_size_range=[148,190],\n",
    "                                          primer_set=orthogonal_primers_remaining,\n",
    "                                          validated_primer_set=validated_primer_combos_filtered,\n",
    "                                          aa_start={                    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dccc5b7-d5e0-44b3-a683-1cd5cbde887c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subtract the primer pairs used in making the first two sets of libraries from the remaining pairs to make the list for the 3p libraries\n",
    "\n",
    "#1. Read the two primer tables\n",
    "df_used = pd.read_csv(\n",
    "    \"Johnsenlab/int_ampkey.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    usecols=[\"Forward Primer\", \"Reverse Primer\"]\n",
    ")\n",
    "\n",
    "df_all = orthogonal_primers_remaining  # keep every column\n",
    "\n",
    "#2. Build helper columns to subracting \n",
    "df_used_norm = df_used.assign(\n",
    "    FWD=df_used[\"Forward Primer\"].str.strip().str.upper(),\n",
    "    REV=df_used[\"Reverse Primer\"].str.strip().str.upper()\n",
    ")\n",
    "\n",
    "df_all_norm = df_all.assign(\n",
    "    FWD=df_all[\"Forward Primer\"].str.strip().str.upper(),\n",
    "    REV=df_all[\"Reverse Primer\"].str.strip().str.upper()\n",
    ")\n",
    "\n",
    "#3. Identify pairs that are already used\n",
    "used_index = df_used_norm.set_index([\"FWD\", \"REV\"]).index\n",
    "mask       = ~df_all_norm.set_index([\"FWD\", \"REV\"]).index.isin(used_index)\n",
    "\n",
    "#4. Keep the rows (and *all* original columns) that remain\n",
    "orthogonal_primers_remaining_iter2 = df_all.loc[mask].reset_index(drop=True)\n",
    "\n",
    "print(f\"{len(orthogonal_primers_remaining_iter5)} primer pairs remain.\")\n",
    "orthogonal_primers_remaining_iter2.head()\n",
    "orthogonal_primers_remaining_iter2.to_csv(\"orthogonal_primers_remaining_iter2.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1eb64bb5-a8b5-4940-a1c7-50790e1fb327",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Make the 3p libraries\n",
    "oligo_array,amp_primers,gblocks,amp_primer_dict,breakpoint_df = write_oligo_library({\n",
    "                                            'FRED3':FRED_SB3,\n",
    "                                            'VWF5':VWF_SB5},\n",
    "                                          oligo_file='Johnsenlab/3p_oligos.csv',\n",
    "                                          primer_file='Johnsenlab/3p_primers.tsv',\n",
    "                                          gbl_file='Johnsenlab/3p_gblocks.tsv',\n",
    "                                          gbl_large_file='Johnsenlab/3p_gblocks_large.tsv',\n",
    "                                          amp_primer_key_file='Johnsenlab/3p_ampkey.tsv',\n",
    "                                          breakpoint_file='Johnsenlab/3p_breakpoints.tsv',\n",
    "                                          block_size_range=[148,190],\n",
    "                                          primer_set=orthogonal_primers_remaining_iter2,\n",
    "                                          validated_primer_set=validated_primer_combos_filtered,\n",
    "                                          aa_start={                    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b087596a-516d-473e-9ae3-27639fa894af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Okay, I have a draft of the library, now I need to see if the overhangs conflict with the superblocks\n",
    "# First I need to pull the overhangs from the tiles using the breakpoint values\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "67b4fc64-6f21-4793-884b-1d8254f8a4a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = {\n",
    "           'FRED1':FRED_SB1,\n",
    "           'FRED2':FRED_SB2,\n",
    "           'FRED3':FRED_SB3,\n",
    "           'VWF1':VWF_SB1,\n",
    "           'VWF2':VWF_SB2,\n",
    "           'VWF3':VWF_SB3,\n",
    "           'VWF4':VWF_SB4,\n",
    "           'VWF5':VWF_SB5\n",
    "        }\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f12435a9-0bcc-438a-ba3d-0f803fc17092",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
